

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>API Documentation &mdash; MHCflurry 1.6.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Command-line reference" href="commandline_tools.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> MHCflurry
          

          
          </a>

          
            
            
              <div class="version">
                1.6.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction and setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_tutorial.html">Command-line tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="python_tutorial.html">Python library tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="commandline_tools.html">Command-line reference</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.allele_encoding">mhcflurry.allele_encoding module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.amino_acid">mhcflurry.amino_acid module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.calibrate_percentile_ranks_command">mhcflurry.calibrate_percentile_ranks_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.class1_affinity_predictor">mhcflurry.class1_affinity_predictor module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.class1_neural_network">mhcflurry.class1_neural_network module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.class1_presentation_predictor">mhcflurry.class1_presentation_predictor module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.class1_processing_neural_network">mhcflurry.class1_processing_neural_network module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.class1_processing_predictor">mhcflurry.class1_processing_predictor module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.cluster_parallelism">mhcflurry.cluster_parallelism module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.common">mhcflurry.common module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.custom_loss">mhcflurry.custom_loss module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.data_dependent_weights_initialization">mhcflurry.data_dependent_weights_initialization module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.downloads">mhcflurry.downloads module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.downloads_command">mhcflurry.downloads_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.encodable_sequences">mhcflurry.encodable_sequences module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.ensemble_centrality">mhcflurry.ensemble_centrality module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.fasta">mhcflurry.fasta module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.flanking_encoding">mhcflurry.flanking_encoding module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.hyperparameters">mhcflurry.hyperparameters module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.local_parallelism">mhcflurry.local_parallelism module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.percent_rank_transform">mhcflurry.percent_rank_transform module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.predict_command">mhcflurry.predict_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.predict_scan_command">mhcflurry.predict_scan_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.random_negative_peptides">mhcflurry.random_negative_peptides module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.regression_target">mhcflurry.regression_target module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.scoring">mhcflurry.scoring module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.select_allele_specific_models_command">mhcflurry.select_allele_specific_models_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.select_pan_allele_models_command">mhcflurry.select_pan_allele_models_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.select_processing_models_command">mhcflurry.select_processing_models_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.testing_utils">mhcflurry.testing_utils module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.train_allele_specific_models_command">mhcflurry.train_allele_specific_models_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.train_pan_allele_models_command">mhcflurry.train_pan_allele_models_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.train_presentation_models_command">mhcflurry.train_presentation_models_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.train_processing_models_command">mhcflurry.train_processing_models_command module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-mhcflurry.version">mhcflurry.version module</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MHCflurry</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>API Documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-mhcflurry">
<span id="id1"></span><span id="api-documentation"></span><h1>API Documentation<a class="headerlink" href="#module-mhcflurry" title="Permalink to this headline">¶</a></h1>
<p>Class I MHC ligand prediction package</p>
<dl class="class">
<dt id="mhcflurry.Class1AffinityPredictor">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.</code><code class="sig-name descname">Class1AffinityPredictor</code><span class="sig-paren">(</span><em class="sig-param">allele_to_allele_specific_models=None</em>, <em class="sig-param">class1_pan_allele_models=None</em>, <em class="sig-param">allele_to_sequence=None</em>, <em class="sig-param">manifest_df=None</em>, <em class="sig-param">allele_to_percent_rank_transform=None</em>, <em class="sig-param">metadata_dataframes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>High-level interface for peptide/MHC I binding affinity prediction.</p>
<p>This class manages low-level <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a> instances, each of which
wraps a single Keras network. The purpose of <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> is to
implement ensembles, handling of multiple alleles, and predictor loading and
saving. It also provides a place to keep track of metadata like prediction
histograms for percentile rank calibration.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>allele_to_allele_specific_models</strong><span class="classifier">dict of string -&gt; list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a></span></dt><dd><p>Ensemble of single-allele models to use for each allele.</p>
</dd>
<dt><strong>class1_pan_allele_models</strong><span class="classifier">list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a></span></dt><dd><p>Ensemble of pan-allele models.</p>
</dd>
<dt><strong>allele_to_sequence</strong><span class="classifier">dict of string -&gt; string</span></dt><dd><p>MHC allele name to fixed-length amino acid sequence (sometimes
referred to as the pseudosequence). Required only if
class1_pan_allele_models is specified.</p>
</dd>
<dt><strong>manifest_df</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>, optional</span></dt><dd><p>Must have columns: model_name, allele, config_json, model.
Only required if you want to update an existing serialization of a
Class1AffinityPredictor. Otherwise this dataframe will be generated
automatically based on the supplied models.</p>
</dd>
<dt><strong>allele_to_percent_rank_transform</strong><span class="classifier">dict of string -&gt; <code class="xref py py-obj docutils literal notranslate"><span class="pre">PercentRankTransform</span></code>, optional</span></dt><dd><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">PercentRankTransform</span></code> instances to use for each allele</p>
</dd>
<dt><strong>metadata_dataframes</strong><span class="classifier">dict of string -&gt; pandas.DataFrame, optional</span></dt><dd><p>Optional additional dataframes to write to the models dir when
save() is called. Useful for tracking provenance.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.manifest_df">
<em class="property">property </em><code class="sig-name descname">manifest_df</code><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.manifest_df" title="Permalink to this definition">¶</a></dt>
<dd><p>A pandas.DataFrame describing the models included in this predictor.</p>
<p>Based on:
- self.class1_pan_allele_models
- self.allele_to_allele_specific_models</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.clear_cache">
<code class="sig-name descname">clear_cache</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.clear_cache"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.clear_cache" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear values cached based on the neural networks in this predictor.</p>
<dl class="simple">
<dt>Users should call this after mutating any of the following:</dt><dd><ul class="simple">
<li><p>self.class1_pan_allele_models</p></li>
<li><p>self.allele_to_allele_specific_models</p></li>
<li><p>self.allele_to_sequence</p></li>
</ul>
</dd>
</dl>
<p>Methods that mutate these instance variables will call this method on
their own if needed.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.neural_networks">
<em class="property">property </em><code class="sig-name descname">neural_networks</code><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.neural_networks" title="Permalink to this definition">¶</a></dt>
<dd><p>List of the neural networks in the ensemble.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">predictors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge the ensembles of two or more <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> instances.</p>
<p>Note: the resulting merged predictor will NOT have calibrated percentile
ranks. Call <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.calibrate_percentile_ranks" title="mhcflurry.Class1AffinityPredictor.calibrate_percentile_ranks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calibrate_percentile_ranks</span></code></a> on it if these are needed.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>predictors</strong><span class="classifier">sequence of <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a></span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.merge_in_place">
<code class="sig-name descname">merge_in_place</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">others</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.merge_in_place"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.merge_in_place" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the models present in other predictors into the current predictor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>others</strong><span class="classifier">list of Class1AffinityPredictor</span></dt><dd><p>Other predictors to merge into the current predictor.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>list of string</strong><span class="classifier">names of newly added models</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.supported_alleles">
<em class="property">property </em><code class="sig-name descname">supported_alleles</code><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.supported_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Alleles for which predictions can be made.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.supported_peptide_lengths">
<em class="property">property </em><code class="sig-name descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(minimum, maximum) lengths of peptides supported by <em>all models</em>,
inclusive.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>(int, int) tuple</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.check_consistency">
<code class="sig-name descname">check_consistency</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.check_consistency"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.check_consistency" title="Permalink to this definition">¶</a></dt>
<dd><p>Verify that self.manifest_df is consistent with:
- self.class1_pan_allele_models
- self.allele_to_allele_specific_models</p>
<p>Currently only checks for agreement on the total number of models.</p>
<p>Throws AssertionError if inconsistent.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.save">
<code class="sig-name descname">save</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">models_dir</em>, <em class="sig-param">model_names_to_write=None</em>, <em class="sig-param">write_metadata=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the predictor to a directory on disk. If the directory does
not exist it will be created.</p>
<p>The serialization format consists of a file called “manifest.csv” with
the configurations of each Class1NeuralNetwork, along with per-network
files giving the model weights. If there are pan-allele predictors in
the ensemble, the allele sequences are also stored in the
directory. There is also a small file “index.txt” with basic metadata:
when the models were trained, by whom, on what host.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. It will be created if it doesn’t exist.</p>
</dd>
<dt><strong>model_names_to_write</strong><span class="classifier">list of string, optional</span></dt><dd><p>Only write the weights for the specified models. Useful for
incremental updates during training.</p>
</dd>
<dt><strong>write_metadata</strong><span class="classifier">boolean, optional</span></dt><dd><p>Whether to write optional metadata</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.load">
<em class="property">static </em><code class="sig-name descname">load</code><span class="sig-paren">(</span><em class="sig-param">models_dir=None</em>, <em class="sig-param">max_models=None</em>, <em class="sig-param">optimization_level=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserialize a predictor from a directory on disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. If unspecified the default downloaded models are
used.</p>
</dd>
<dt><strong>max_models</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum number of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a> instances to load</p>
</dd>
<dt><strong>optimization_level</strong><span class="classifier">int</span></dt><dd><p>If &gt;0, model optimization will be attempted. Defaults to value of
environment variable MHCFLURRY_OPTIMIZATION_LEVEL.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#mhcflurry.Class1AffinityPredictor" title="mhcflurry.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.optimize">
<code class="sig-name descname">optimize</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">warn=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.optimize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.optimize" title="Permalink to this definition">¶</a></dt>
<dd><p>EXPERIMENTAL: Optimize the predictor for faster predictions.</p>
<p>Currently the only optimization implemented is to merge multiple pan-
allele predictors at the tensorflow level.</p>
<p>The optimization is performed in-place, mutating the instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>bool</dt><dd><p>Whether optimization was performed</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.model_name">
<em class="property">static </em><code class="sig-name descname">model_name</code><span class="sig-paren">(</span><em class="sig-param">allele</em>, <em class="sig-param">num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.model_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.model_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a model name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>num</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.weights_path">
<em class="property">static </em><code class="sig-name descname">weights_path</code><span class="sig-paren">(</span><em class="sig-param">models_dir</em>, <em class="sig-param">model_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.weights_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.weights_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the path to the weights file for a model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>model_name</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.master_allele_encoding">
<em class="property">property </em><code class="sig-name descname">master_allele_encoding</code><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.master_allele_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>An AlleleEncoding containing the universe of alleles specified by
self.allele_to_sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>AlleleEncoding</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.fit_allele_specific_predictors">
<code class="sig-name descname">fit_allele_specific_predictors</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">n_models</em>, <em class="sig-param">architecture_hyperparameters_list</em>, <em class="sig-param">allele</em>, <em class="sig-param">peptides</em>, <em class="sig-param">affinities</em>, <em class="sig-param">inequalities=None</em>, <em class="sig-param">train_rounds=None</em>, <em class="sig-param">models_dir_for_save=None</em>, <em class="sig-param">verbose=0</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.fit_allele_specific_predictors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.fit_allele_specific_predictors" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit one or more allele specific predictors for a single allele using one
or more neural network architectures.</p>
<p>The new predictors are saved in the Class1AffinityPredictor instance
and will be used on subsequent calls to <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>n_models</strong><span class="classifier">int</span></dt><dd><p>Number of neural networks to fit</p>
</dd>
<dt><strong>architecture_hyperparameters_list</strong><span class="classifier">list of dict</span></dt><dd><p>List of hyperparameter sets.</p>
</dd>
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>peptides</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodableSequences</span></code> or list of string</span></dt><dd></dd>
<dt><strong>affinities</strong><span class="classifier">list of float</span></dt><dd><p>nM affinities</p>
</dd>
<dt><strong>inequalities</strong><span class="classifier">list of string, each element one of “&gt;”, “&lt;”, or “=”</span></dt><dd><p>See <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork.fit" title="mhcflurry.Class1NeuralNetwork.fit"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork.fit</span></code></a> for details.</p>
</dd>
<dt><strong>train_rounds</strong><span class="classifier">sequence of int</span></dt><dd><p>Each training point i will be used on training rounds r for which
train_rounds[i] &gt; r, r &gt;= 0.</p>
</dd>
<dt><strong>models_dir_for_save</strong><span class="classifier">string, optional</span></dt><dd><p>If specified, the Class1AffinityPredictor is (incrementally) written
to the given models dir after each neural network is fit.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Keras verbosity</p>
</dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd><p>Optional string of information to include in each progress update</p>
</dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd><p>How often (in seconds) to print progress. Set to None to disable.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.fit_class1_pan_allele_models">
<code class="sig-name descname">fit_class1_pan_allele_models</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">n_models</em>, <em class="sig-param">architecture_hyperparameters</em>, <em class="sig-param">alleles</em>, <em class="sig-param">peptides</em>, <em class="sig-param">affinities</em>, <em class="sig-param">inequalities</em>, <em class="sig-param">models_dir_for_save=None</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.fit_class1_pan_allele_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.fit_class1_pan_allele_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit one or more pan-allele predictors using a single neural network
architecture.</p>
<p>The new predictors are saved in the Class1AffinityPredictor instance
and will be used on subsequent calls to <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>n_models</strong><span class="classifier">int</span></dt><dd><p>Number of neural networks to fit</p>
</dd>
<dt><strong>architecture_hyperparameters</strong><span class="classifier">dict</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">list of string</span></dt><dd><p>Allele names (not sequences) corresponding to each peptide</p>
</dd>
<dt><strong>peptides</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodableSequences</span></code> or list of string</span></dt><dd></dd>
<dt><strong>affinities</strong><span class="classifier">list of float</span></dt><dd><p>nM affinities</p>
</dd>
<dt><strong>inequalities</strong><span class="classifier">list of string, each element one of “&gt;”, “&lt;”, or “=”</span></dt><dd><p>See Class1NeuralNetwork.fit for details.</p>
</dd>
<dt><strong>models_dir_for_save</strong><span class="classifier">string, optional</span></dt><dd><p>If specified, the Class1AffinityPredictor is (incrementally) written
to the given models dir after each neural network is fit.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Keras verbosity</p>
</dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd><p>Optional string of information to include in each progress update</p>
</dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd><p>How often (in seconds) to print progress. Set to None to disable.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of <a class="reference internal" href="#mhcflurry.Class1NeuralNetwork" title="mhcflurry.Class1NeuralNetwork"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></a></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.add_pan_allele_model">
<code class="sig-name descname">add_pan_allele_model</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">model</em>, <em class="sig-param">models_dir_for_save=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.add_pan_allele_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.add_pan_allele_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a pan-allele model to the ensemble and optionally do an incremental
save.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>model</strong><span class="classifier">Class1NeuralNetwork</span></dt><dd></dd>
<dt><strong>models_dir_for_save</strong><span class="classifier">string</span></dt><dd><p>Directory to save resulting ensemble to</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.percentile_ranks">
<code class="sig-name descname">percentile_ranks</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">affinities</em>, <em class="sig-param">allele=None</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Return percentile ranks for the given ic50 affinities and alleles.</p>
<p>The ‘allele’ and ‘alleles’ argument are as in the <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method.
Specify one of these.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>affinities</strong><span class="classifier">sequence of float</span></dt><dd><p>nM affinities</p>
</dd>
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">sequence of string</span></dt><dd></dd>
<dt><strong>throw</strong><span class="classifier">boolean</span></dt><dd><p>If True, a ValueError will be raised in the case of unsupported
alleles. If False, a warning will be logged and NaN will be returned
for those percentile ranks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of float</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">allele=None</em>, <em class="sig-param">throw=True</em>, <em class="sig-param">centrality_measure='mean'</em>, <em class="sig-param">model_kwargs={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict nM binding affinities.</p>
<p>If multiple predictors are available for an allele, the predictions are
the geometric means of the individual model (nM) predictions.</p>
<p>One of ‘allele’ or ‘alleles’ must be specified. If ‘allele’ is specified
all predictions will be for the given allele. If ‘alleles’ is specified
it must be the same length as ‘peptides’ and give the allele
corresponding to each peptide.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>peptides</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodableSequences</span></code> or list of string</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>throw</strong><span class="classifier">boolean</span></dt><dd><p>If True, a ValueError will be raised in the case of unsupported
alleles or peptide lengths. If False, a warning will be logged and
the predictions for the unsupported alleles or peptides will be NaN.</p>
</dd>
<dt><strong>centrality_measure</strong><span class="classifier">string or callable</span></dt><dd><p>Measure of central tendency to use to combine predictions in the
ensemble. Options include: mean, median, robust_mean.</p>
</dd>
<dt><strong>model_kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments to pass to Class1NeuralNetwork.predict</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of predictions</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.predict_to_dataframe">
<code class="sig-name descname">predict_to_dataframe</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">allele=None</em>, <em class="sig-param">throw=True</em>, <em class="sig-param">include_individual_model_predictions=False</em>, <em class="sig-param">include_percentile_ranks=True</em>, <em class="sig-param">include_confidence_intervals=True</em>, <em class="sig-param">centrality_measure='mean'</em>, <em class="sig-param">model_kwargs={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.predict_to_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.predict_to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict nM binding affinities. Gives more detailed output than <a class="reference internal" href="#mhcflurry.Class1AffinityPredictor.predict" title="mhcflurry.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>
method, including 5-95% prediction intervals.</p>
<p>If multiple predictors are available for an allele, the predictions are
the geometric means of the individual model predictions.</p>
<p>One of ‘allele’ or ‘alleles’ must be specified. If ‘allele’ is specified
all predictions will be for the given allele. If ‘alleles’ is specified
it must be the same length as ‘peptides’ and give the allele
corresponding to each peptide.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>peptides</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodableSequences</span></code> or list of string</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>throw</strong><span class="classifier">boolean</span></dt><dd><p>If True, a ValueError will be raised in the case of unsupported
alleles or peptide lengths. If False, a warning will be logged and
the predictions for the unsupported alleles or peptides will be NaN.</p>
</dd>
<dt><strong>include_individual_model_predictions</strong><span class="classifier">boolean</span></dt><dd><p>If True, the predictions of each individual model are included as
columns in the result DataFrame.</p>
</dd>
<dt><strong>include_percentile_ranks</strong><span class="classifier">boolean, default True</span></dt><dd><p>If True, a “prediction_percentile” column will be included giving
the percentile ranks. If no percentile rank info is available,
this will be ignored with a warning.</p>
</dd>
<dt><strong>centrality_measure</strong><span class="classifier">string or callable</span></dt><dd><p>Measure of central tendency to use to combine predictions in the
ensemble. Options include: mean, median, robust_mean.</p>
</dd>
<dt><strong>model_kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments to pass to Class1NeuralNetwork.predict</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> of predictions</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.calibrate_percentile_ranks">
<code class="sig-name descname">calibrate_percentile_ranks</code><span class="sig-paren">(</span><em class="sig-param">self, peptides=None, num_peptides_per_length=100000, alleles=None, bins=None, motif_summary=False, summary_top_peptide_fractions=[0.001], verbose=False, model_kwargs={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.calibrate_percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.calibrate_percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the cumulative distribution of ic50 values for a set of alleles
over a large universe of random peptides, to enable taking quantiles
of this distribution later.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">sequence of string or EncodableSequences, optional</span></dt><dd><p>Peptides to use</p>
</dd>
<dt><strong>num_peptides_per_length</strong><span class="classifier">int, optional</span></dt><dd><p>If peptides argument is not specified, then num_peptides_per_length
peptides are randomly sampled from a uniform distribution for each
supported length</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">sequence of string, optional</span></dt><dd><p>Alleles to perform calibration for. If not specified all supported
alleles will be calibrated.</p>
</dd>
<dt><strong>bins</strong><span class="classifier">object</span></dt><dd><p>Anything that can be passed to numpy.histogram’s “bins” argument
can be used here, i.e. either an integer or a sequence giving bin
edges. This is in ic50 space.</p>
</dd>
<dt><strong>motif_summary</strong><span class="classifier">bool</span></dt><dd><p>If True, the length distribution and per-position amino acid
frequencies are also calculated for the top x fraction of tightest-
binding peptides, where each value of x is given in the
summary_top_peptide_fractions list.</p>
</dd>
<dt><strong>summary_top_peptide_fractions</strong><span class="classifier">list of float</span></dt><dd><p>Only used if motif_summary is True</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">boolean</span></dt><dd><p>Whether to print status updates to stdout</p>
</dd>
<dt><strong>model_kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional low-level Class1NeuralNetwork.predict() kwargs.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>dict of string -&gt; pandas.DataFrame</dt><dd></dd>
<dt>If motif_summary is True, this will have keys  “frequency_matrices” and</dt><dd></dd>
<dt>“length_distributions”. Otherwise it will be empty.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1AffinityPredictor.model_select">
<code class="sig-name descname">model_select</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">score_function</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">min_models=1</em>, <em class="sig-param">max_models=10000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.model_select"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1AffinityPredictor.model_select" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform model selection using a user-specified scoring function.</p>
<p>This works only with allele-specific models, not pan-allele models.</p>
<p>Model selection is done using a “step up” variable selection procedure,
in which models are repeatedly added to an ensemble until the score
stops improving.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>score_function</strong><span class="classifier">Class1AffinityPredictor -&gt; float function</span></dt><dd><p>Scoring function</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">list of string, optional</span></dt><dd><p>If not specified, model selection is performed for all alleles.</p>
</dd>
<dt><strong>min_models</strong><span class="classifier">int, optional</span></dt><dd><p>Min models to select per allele</p>
</dd>
<dt><strong>max_models</strong><span class="classifier">int, optional</span></dt><dd><p>Max models to select per allele</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>Class1AffinityPredictor</strong><span class="classifier">predictor containing the selected models</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.Class1NeuralNetwork">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.</code><code class="sig-name descname">Class1NeuralNetwork</code><span class="sig-paren">(</span><em class="sig-param">**hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Low level class I predictor consisting of a single neural network.</p>
<p>Both single allele and pan-allele prediction are supported.</p>
<p>Users will generally use Class1AffinityPredictor, which gives a higher-level
interface and supports ensembles.</p>
<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.network_hyperparameter_defaults">
<code class="sig-name descname">network_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.network_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters (and their default values) that affect the neural network
architecture.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.compile_hyperparameter_defaults">
<code class="sig-name descname">compile_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.compile_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Loss and optimizer hyperparameters.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.fit_hyperparameter_defaults">
<code class="sig-name descname">fit_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.fit_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for neural network training.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.early_stopping_hyperparameter_defaults">
<code class="sig-name descname">early_stopping_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.early_stopping_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for early stopping.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.miscelaneous_hyperparameter_defaults">
<code class="sig-name descname">miscelaneous_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.miscelaneous_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Miscelaneous hyperaparameters. These parameters are not used by this class
but may be interpreted by other code.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.hyperparameter_defaults">
<code class="sig-name descname">hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Combined set of all supported hyperparameters and their default values.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.hyperparameter_renames">
<code class="sig-name descname">hyperparameter_renames</code><em class="property"> = {'embedding_init_method': None, 'embedding_input_dim': None, 'embedding_output_dim': None, 'kmer_size': None, 'left_edge': None, 'min_delta': None, 'mode': None, 'monitor': None, 'peptide_amino_acid_encoding': None, 'pseudosequence_use_embedding': None, 'right_edge': None, 'take_best_epoch': None, 'use_embedding': None, 'verbose': None}</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.hyperparameter_renames" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.apply_hyperparameter_renames">
<em class="property">classmethod </em><code class="sig-name descname">apply_hyperparameter_renames</code><span class="sig-paren">(</span><em class="sig-param">hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.apply_hyperparameter_renames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.apply_hyperparameter_renames" title="Permalink to this definition">¶</a></dt>
<dd><p>Handle hyperparameter renames.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>hyperparameters</strong><span class="classifier">dict</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>dict</strong><span class="classifier">updated hyperparameters</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1NeuralNetwork.KERAS_MODELS_CACHE">
<code class="sig-name descname">KERAS_MODELS_CACHE</code><em class="property"> = {}</em><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.KERAS_MODELS_CACHE" title="Permalink to this definition">¶</a></dt>
<dd><p>Process-wide keras model cache, a map from: architecture JSON string to
(Keras model, existing network weights)</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.clear_model_cache">
<em class="property">classmethod </em><code class="sig-name descname">clear_model_cache</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.clear_model_cache"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.clear_model_cache" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear the Keras model cache.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.borrow_cached_network">
<em class="property">classmethod </em><code class="sig-name descname">borrow_cached_network</code><span class="sig-paren">(</span><em class="sig-param">network_json</em>, <em class="sig-param">network_weights</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.borrow_cached_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.borrow_cached_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a keras Model with the specified architecture and weights.
As an optimization, when possible this will reuse architectures from a
process-wide cache.</p>
<p>The returned object is “borrowed” in the sense that its weights can
change later after subsequent calls to this method from other objects.</p>
<p>If you’re using this from a parallel implementation you’ll need to
hold a lock while using the returned object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>network_json</strong><span class="classifier">string of JSON</span></dt><dd></dd>
<dt><strong>network_weights</strong><span class="classifier">list of numpy.array</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>keras.models.Model</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.network">
<code class="sig-name descname">network</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">borrow=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the keras model associated with this predictor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>borrow</strong><span class="classifier">bool</span></dt><dd><p>Whether to return a cached model if possible. See
borrow_cached_network for details</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>keras.models.Model</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.update_network_description">
<code class="sig-name descname">update_network_description</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.update_network_description"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.update_network_description" title="Permalink to this definition">¶</a></dt>
<dd><p>Update self.network_json and self.network_weights properties based on
this instances’s neural network.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.keras_network_cache_key">
<em class="property">static </em><code class="sig-name descname">keras_network_cache_key</code><span class="sig-paren">(</span><em class="sig-param">network_json</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.keras_network_cache_key"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.keras_network_cache_key" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a Keras JSON description of a neural network, return a key that
uniquely defines this network. Networks that share the same key should
have compatible weights matrices and give the same prediction outputs
when their weights are the same.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>network_json</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.get_config">
<code class="sig-name descname">get_config</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.get_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.get_config" title="Permalink to this definition">¶</a></dt>
<dd><p>serialize to a dict all attributes except model weights</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>dict</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.from_config">
<em class="property">classmethod </em><code class="sig-name descname">from_config</code><span class="sig-paren">(</span><em class="sig-param">config</em>, <em class="sig-param">weights=None</em>, <em class="sig-param">weights_loader=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.from_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.from_config" title="Permalink to this definition">¶</a></dt>
<dd><p>deserialize from a dict returned by get_config().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>config</strong><span class="classifier">dict</span></dt><dd></dd>
<dt><strong>weights</strong><span class="classifier">list of array, optional</span></dt><dd><p>Network weights to restore</p>
</dd>
<dt><strong>weights_loader</strong><span class="classifier">callable, optional</span></dt><dd><p>Function to call (no arguments) to load weights when needed</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Class1NeuralNetwork</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.load_weights">
<code class="sig-name descname">load_weights</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.load_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.load_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Load weights by evaluating self.network_weights_loader, if needed.</p>
<p>After calling this, self.network_weights_loader will be None and
self.network_weights will be the weights list, if available.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.get_weights">
<code class="sig-name descname">get_weights</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.get_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.get_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the network weights</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of numpy.array giving weights for each layer or None if there is no</dt><dd></dd>
<dt>network</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.peptides_to_network_input">
<code class="sig-name descname">peptides_to_network_input</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.peptides_to_network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.peptides_to_network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode peptides to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">EncodableSequences or list of string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.supported_peptide_lengths">
<em class="property">property </em><code class="sig-name descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(minimum, maximum) lengths of peptides supported, inclusive.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>(int, int) tuple</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.allele_encoding_to_network_input">
<code class="sig-name descname">allele_encoding_to_network_input</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele_encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.allele_encoding_to_network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.allele_encoding_to_network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode alleles to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>allele_encoding</strong><span class="classifier">AlleleEncoding</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>(numpy.array, numpy.array)</dt><dd></dd>
<dt>Indices and allele representations.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.data_dependent_weights_initialization">
<em class="property">static </em><code class="sig-name descname">data_dependent_weights_initialization</code><span class="sig-paren">(</span><em class="sig-param">network</em>, <em class="sig-param">x_dict=None</em>, <em class="sig-param">method='lsuv'</em>, <em class="sig-param">verbose=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.data_dependent_weights_initialization"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.data_dependent_weights_initialization" title="Permalink to this definition">¶</a></dt>
<dd><p>Data dependent weights initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>network</strong><span class="classifier">keras.Model</span></dt><dd></dd>
<dt><strong>x_dict</strong><span class="classifier">dict of string -&gt; numpy.ndarray</span></dt><dd><p>Training data as would be passed keras.Model.fit().</p>
</dd>
<dt><strong>method</strong><span class="classifier">string</span></dt><dd><p>Initialization method. Currently only “lsuv” is supported.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Status updates printed to stdout if verbose &gt; 0</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.fit_generator">
<code class="sig-name descname">fit_generator</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">generator</em>, <em class="sig-param">validation_peptide_encoding</em>, <em class="sig-param">validation_affinities</em>, <em class="sig-param">validation_allele_encoding=None</em>, <em class="sig-param">validation_inequalities=None</em>, <em class="sig-param">validation_output_indices=None</em>, <em class="sig-param">steps_per_epoch=10</em>, <em class="sig-param">epochs=1000</em>, <em class="sig-param">min_epochs=0</em>, <em class="sig-param">patience=10</em>, <em class="sig-param">min_delta=0.0</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">progress_callback=None</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.fit_generator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.fit_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit using a generator. Does not support many of the features of fit(),
such as random negative peptides.</p>
<p>Fitting proceeds until early stopping is hit, using the peptides,
affinities, etc. given by the parameters starting with “<a href="#id6"><span class="problematic" id="id7">validation_</span></a>”.</p>
<p>This is used for pre-training pan-allele models using data synthesized
by the allele-specific models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>generator</strong><span class="classifier">generator yielding (alleles, peptides, affinities) tuples</span></dt><dd><p>where alleles and peptides are lists of strings, and affinities
is list of floats.</p>
</dd>
<dt><strong>validation_peptide_encoding</strong><span class="classifier">EncodableSequences</span></dt><dd></dd>
<dt><strong>validation_affinities</strong><span class="classifier">list of float</span></dt><dd></dd>
<dt><strong>validation_allele_encoding</strong><span class="classifier">AlleleEncoding</span></dt><dd></dd>
<dt><strong>validation_inequalities</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>validation_output_indices</strong><span class="classifier">list of int</span></dt><dd></dd>
<dt><strong>steps_per_epoch</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>epochs</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>min_epochs</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>patience</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>min_delta</strong><span class="classifier">float</span></dt><dd></dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>progress_callback</strong><span class="classifier">thunk</span></dt><dd></dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">affinities</em>, <em class="sig-param">allele_encoding=None</em>, <em class="sig-param">inequalities=None</em>, <em class="sig-param">output_indices=None</em>, <em class="sig-param">sample_weights=None</em>, <em class="sig-param">shuffle_permutation=None</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">progress_callback=None</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">EncodableSequences or list of string</span></dt><dd></dd>
<dt><strong>affinities</strong><span class="classifier">list of float</span></dt><dd><p>nM affinities. Must be same length of as peptides.</p>
</dd>
<dt><strong>allele_encoding</strong><span class="classifier">AlleleEncoding</span></dt><dd><p>If not specified, the model will be a single-allele predictor.</p>
</dd>
<dt><strong>inequalities</strong><span class="classifier">list of string, each element one of “&gt;”, “&lt;”, or “=”.</span></dt><dd><p>Inequalities to use for fitting. Same length as affinities.
Each element must be one of “&gt;”, “&lt;”, or “=”. For example, a “&gt;”
will train on y_pred &gt; y_true for that element in the training set.
Requires using a custom losses that support inequalities (e.g.
mse_with_ineqalities). If None all inequalities are taken to be “=”.</p>
</dd>
<dt><strong>output_indices</strong><span class="classifier">list of int</span></dt><dd><p>For multi-output models only. Same length as affinities. Indicates
the index of the output (starting from 0) for each training example.</p>
</dd>
<dt><strong>sample_weights</strong><span class="classifier">list of float</span></dt><dd><p>If not specified, all samples (including random negatives added
during training) will have equal weight. If specified, the random
negatives will be assigned weight=1.0.</p>
</dd>
<dt><strong>shuffle_permutation</strong><span class="classifier">list of int</span></dt><dd><p>Permutation (integer list) of same length as peptides and affinities
If None, then a random permutation will be generated.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Keras verbosity level</p>
</dd>
<dt><strong>progress_callback</strong><span class="classifier">function</span></dt><dd><p>No-argument function to call after each epoch.</p>
</dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd><p>Optional string of information to include in each progress update</p>
</dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd><p>How often (in seconds) to print progress update. Set to None to
disable.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">allele_encoding=None</em>, <em class="sig-param">batch_size=4096</em>, <em class="sig-param">output_index=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict affinities.</p>
<p>If peptides are specified as EncodableSequences, then the predictions
will be cached for this predictor as long as the EncodableSequences
object remains in memory. The cache is keyed in the object identity of
the EncodableSequences, not the sequences themselves. The cache is used
only for allele-specific models (i.e. when allele_encoding is None).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">EncodableSequences or list of string</span></dt><dd></dd>
<dt><strong>allele_encoding</strong><span class="classifier">AlleleEncoding, optional</span></dt><dd><p>Only required when this model is a pan-allele model</p>
</dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd><p>batch_size passed to Keras</p>
</dd>
<dt><strong>output_index</strong><span class="classifier">int or None</span></dt><dd><p>For multi-output models. Gives the output index to return. If set to
None, then all outputs are returned as a samples x outputs matrix.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of nM affinity predictions</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">models</em>, <em class="sig-param">merge_method='average'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge multiple models at the tensorflow (or other backend) level.</p>
<p>Only certain neural network architectures support merging. Others will
result in a NotImplementedError.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models</strong><span class="classifier">list of Class1NeuralNetwork</span></dt><dd><p>instances to merge</p>
</dd>
<dt><strong>merge_method</strong><span class="classifier">string, one of “average”, “sum”, or “concatenate”</span></dt><dd><p>How to merge the predictions of the different models</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Class1NeuralNetwork</dt><dd><p>The merged neural network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.make_network">
<code class="sig-name descname">make_network</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptide_encoding</em>, <em class="sig-param">allele_amino_acid_encoding</em>, <em class="sig-param">allele_dense_layer_sizes</em>, <em class="sig-param">peptide_dense_layer_sizes</em>, <em class="sig-param">peptide_allele_merge_method</em>, <em class="sig-param">peptide_allele_merge_activation</em>, <em class="sig-param">layer_sizes</em>, <em class="sig-param">dense_layer_l1_regularization</em>, <em class="sig-param">dense_layer_l2_regularization</em>, <em class="sig-param">activation</em>, <em class="sig-param">init</em>, <em class="sig-param">output_activation</em>, <em class="sig-param">dropout_probability</em>, <em class="sig-param">batch_normalization</em>, <em class="sig-param">locally_connected_layers</em>, <em class="sig-param">topology</em>, <em class="sig-param">num_outputs=1</em>, <em class="sig-param">allele_representations=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.make_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.make_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to make a keras network for class 1 affinity prediction.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.clear_allele_representations">
<code class="sig-name descname">clear_allele_representations</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.clear_allele_representations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.clear_allele_representations" title="Permalink to this definition">¶</a></dt>
<dd><p>Set allele representations to an empty array. Useful before saving to
save a smaller version of the model.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1NeuralNetwork.set_allele_representations">
<code class="sig-name descname">set_allele_representations</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele_representations</em>, <em class="sig-param">force_surgery=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.set_allele_representations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1NeuralNetwork.set_allele_representations" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the allele representations in use by this model. This means mutating
the weights for the allele input embedding layer.</p>
<p>Rationale: instead of passing in the allele sequence for each data point
during model training or prediction (which is expensive in terms of
memory usage), we pass in an allele index between 0 and n-1 where n is
the number of alleles in some universe of possible alleles. This index
is used in the model to lookup the corresponding allele sequence. This
function sets the lookup table.</p>
<p>See also: AlleleEncoding.allele_representations()</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>allele_representations</strong><span class="classifier">numpy.ndarray of shape (a, l, m)</span></dt><dd><dl class="simple">
<dt>where a is the total number of alleles,</dt><dd><p>l is the allele sequence length,
m is the length of the vectors used to represent amino acids</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.Class1ProcessingPredictor">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.</code><code class="sig-name descname">Class1ProcessingPredictor</code><span class="sig-paren">(</span><em class="sig-param">models</em>, <em class="sig-param">manifest_df=None</em>, <em class="sig-param">metadata_dataframes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>User-facing interface to antigen processing prediction.</p>
<p>Delegates to an ensemble of Class1ProcessingNeuralNetwork instances.</p>
<p>Instantiate a new Class1ProcessingPredictor</p>
<p>Users will generally call load() to restore a saved predictor rather
than using this constructor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models</strong><span class="classifier">list of Class1ProcessingNeuralNetwork</span></dt><dd><p>Neural networks in the ensemble.</p>
</dd>
<dt><strong>manifest_df</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>Manifest dataframe. If not specified a new one will be created when
needed.</p>
</dd>
<dt><strong>metadata_dataframes</strong><span class="classifier">dict of string -&gt; pandas.DataFrame</span></dt><dd><p>Arbitrary metadata associated with this predictor</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.sequence_lengths">
<em class="property">property </em><code class="sig-name descname">sequence_lengths</code><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.sequence_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Supported maximum sequence lengths.</p>
<p>Passing a peptide greater than the maximum supported length results
in an error.</p>
<p>Passing an N- or C-flank sequence greater than the maximum supported
length results in some part of it being ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>dict of string -&gt; int</dt><dd></dd>
<dt>Keys are “peptide”, “n_flank”, “c_flank”. Values give the maximum</dt><dd></dd>
<dt>supported sequence length.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.add_models">
<code class="sig-name descname">add_models</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">models</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.add_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.add_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Add models to the ensemble (in-place).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models</strong><span class="classifier">list of Class1ProcessingNeuralNetwork</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of string</dt><dd></dd>
<dt>Names of the new models.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.manifest_df">
<em class="property">property </em><code class="sig-name descname">manifest_df</code><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.manifest_df" title="Permalink to this definition">¶</a></dt>
<dd><p>A pandas.DataFrame describing the models included in this predictor.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.model_name">
<em class="property">static </em><code class="sig-name descname">model_name</code><span class="sig-paren">(</span><em class="sig-param">num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.model_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.model_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a model name</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.weights_path">
<em class="property">static </em><code class="sig-name descname">weights_path</code><span class="sig-paren">(</span><em class="sig-param">models_dir</em>, <em class="sig-param">model_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.weights_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.weights_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the path to the weights file for a model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>model_name</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string</span></dt><dd><p>Upstream sequence before each peptide</p>
</dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string</span></dt><dd><p>Downstream sequence after each peptide</p>
</dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd><p>Prediction keras batch size.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
<dt>Processing scores. Range is 0-1, higher indicates more favorable</dt><dd></dd>
<dt>processing.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.predict_to_dataframe">
<code class="sig-name descname">predict_to_dataframe</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.predict_to_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.predict_to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<p>See <a class="reference internal" href="#mhcflurry.Class1ProcessingPredictor.predict" title="mhcflurry.Class1ProcessingPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method for parameter descriptions.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
<dt>Processing predictions are in the “score” column. Also includes</dt><dd></dd>
<dt>peptides and flanking sequences.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.predict_to_dataframe_encoded">
<code class="sig-name descname">predict_to_dataframe_encoded</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.predict_to_dataframe_encoded"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.predict_to_dataframe_encoded" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<p>See <a class="reference internal" href="#mhcflurry.Class1ProcessingPredictor.predict" title="mhcflurry.Class1ProcessingPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method for more information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">FlankingEncoding</span></dt><dd></dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.check_consistency">
<code class="sig-name descname">check_consistency</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.check_consistency"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.check_consistency" title="Permalink to this definition">¶</a></dt>
<dd><p>Verify that self.manifest_df is consistent with instance variables.</p>
<p>Currently only checks for agreement on the total number of models.</p>
<p>Throws AssertionError if inconsistent.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.save">
<code class="sig-name descname">save</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">models_dir</em>, <em class="sig-param">model_names_to_write=None</em>, <em class="sig-param">write_metadata=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the predictor to a directory on disk. If the directory does
not exist it will be created.</p>
<p>The serialization format consists of a file called “manifest.csv” with
the configurations of each Class1ProcessingNeuralNetwork, along with
per-network files giving the model weights.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. It will be created if it doesn’t exist.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingPredictor.load">
<em class="property">classmethod </em><code class="sig-name descname">load</code><span class="sig-paren">(</span><em class="sig-param">models_dir=None</em>, <em class="sig-param">max_models=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingPredictor.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserialize a predictor from a directory on disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. If unspecified the default downloaded models are
used.</p>
</dd>
<dt><strong>max_models</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum number of models to load</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#mhcflurry.Class1ProcessingPredictor" title="mhcflurry.Class1ProcessingPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1ProcessingPredictor</span></code></a> instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.</code><code class="sig-name descname">Class1ProcessingNeuralNetwork</code><span class="sig-paren">(</span><em class="sig-param">**hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A neural network for antigen processing prediction</p>
<dl class="attribute">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.network_hyperparameter_defaults">
<code class="sig-name descname">network_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.network_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters (and their default values) that affect the neural network
architecture.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.fit_hyperparameter_defaults">
<code class="sig-name descname">fit_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.fit_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for neural network training.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.early_stopping_hyperparameter_defaults">
<code class="sig-name descname">early_stopping_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.early_stopping_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for early stopping.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.compile_hyperparameter_defaults">
<code class="sig-name descname">compile_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.compile_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Loss and optimizer hyperparameters. Any values supported by keras may be
used.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.auxiliary_input_hyperparameter_defaults">
<code class="sig-name descname">auxiliary_input_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.auxiliary_input_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Allele feature hyperparameters.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.hyperparameter_defaults">
<code class="sig-name descname">hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.sequence_lengths">
<em class="property">property </em><code class="sig-name descname">sequence_lengths</code><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.sequence_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Supported maximum sequence lengths</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>dict of string -&gt; int</dt><dd></dd>
<dt>Keys are “peptide”, “n_flank”, “c_flank”. Values give the maximum</dt><dd></dd>
<dt>supported sequence length.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.network">
<code class="sig-name descname">network</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the keras model associated with this network.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.update_network_description">
<code class="sig-name descname">update_network_description</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.update_network_description"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.update_network_description" title="Permalink to this definition">¶</a></dt>
<dd><p>Update self.network_json and self.network_weights properties based on
this instances’s neural network.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em>, <em class="sig-param">targets</em>, <em class="sig-param">sample_weights=None</em>, <em class="sig-param">shuffle_permutation=None</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">progress_callback=None</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">FlankingEncoding</span></dt><dd><p>Peptides and upstream/downstream flanking sequences</p>
</dd>
<dt><strong>targets</strong><span class="classifier">list of float</span></dt><dd><p>1 indicates hit, 0 indicates decoy</p>
</dd>
<dt><strong>sample_weights</strong><span class="classifier">list of float</span></dt><dd><p>If not specified all samples have equal weight.</p>
</dd>
<dt><strong>shuffle_permutation</strong><span class="classifier">list of int</span></dt><dd><p>Permutation (integer list) of same length as peptides and affinities
If None, then a random permutation will be generated.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Keras verbosity level</p>
</dd>
<dt><strong>progress_callback</strong><span class="classifier">function</span></dt><dd><p>No-argument function to call after each epoch.</p>
</dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd><p>Optional string of information to include in each progress update</p>
</dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd><p>How often (in seconds) to print progress update. Set to None to
disable.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string</span></dt><dd><p>Upstream sequence before each peptide</p>
</dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string</span></dt><dd><p>Downstream sequence after each peptide</p>
</dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd><p>Prediction keras batch size.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
<dt>Processing scores. Range is 0-1, higher indicates more favorable</dt><dd></dd>
<dt>processing.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.predict_encoded">
<code class="sig-name descname">predict_encoded</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.predict_encoded"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.predict_encoded" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">FlankingEncoding</span></dt><dd><p>Peptides and flanking sequences</p>
</dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd><p>Prediction keras batch size.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.network_input">
<code class="sig-name descname">network_input</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode peptides to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">FlankingEncoding</span></dt><dd><p>Peptides and flanking sequences</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.make_network">
<code class="sig-name descname">make_network</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">amino_acid_encoding</em>, <em class="sig-param">peptide_max_length</em>, <em class="sig-param">n_flank_length</em>, <em class="sig-param">c_flank_length</em>, <em class="sig-param">flanking_averages</em>, <em class="sig-param">convolutional_filters</em>, <em class="sig-param">convolutional_kernel_size</em>, <em class="sig-param">convolutional_activation</em>, <em class="sig-param">convolutional_kernel_l1_l2</em>, <em class="sig-param">dropout_rate</em>, <em class="sig-param">post_convolutional_dense_layer_sizes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.make_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.make_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to make a keras network given hyperparameters.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.get_weights">
<code class="sig-name descname">get_weights</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.get_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.get_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the network weights</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of numpy.array giving weights for each layer or None if there is no</dt><dd></dd>
<dt>network</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.get_config">
<code class="sig-name descname">get_config</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.get_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.get_config" title="Permalink to this definition">¶</a></dt>
<dd><p>serialize to a dict all attributes except model weights</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>dict</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1ProcessingNeuralNetwork.from_config">
<em class="property">classmethod </em><code class="sig-name descname">from_config</code><span class="sig-paren">(</span><em class="sig-param">config</em>, <em class="sig-param">weights=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.from_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1ProcessingNeuralNetwork.from_config" title="Permalink to this definition">¶</a></dt>
<dd><p>deserialize from a dict returned by get_config().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>config</strong><span class="classifier">dict</span></dt><dd></dd>
<dt><strong>weights</strong><span class="classifier">list of array, optional</span></dt><dd><p>Network weights to restore</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Class1ProcessingNeuralNetwork</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.Class1PresentationPredictor">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.</code><code class="sig-name descname">Class1PresentationPredictor</code><span class="sig-paren">(</span><em class="sig-param">affinity_predictor=None</em>, <em class="sig-param">processing_predictor_with_flanks=None</em>, <em class="sig-param">processing_predictor_without_flanks=None</em>, <em class="sig-param">weights_dataframe=None</em>, <em class="sig-param">metadata_dataframes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A logistic regression model over predicted binding affinity (BA) and antigen
processing (AP) score.</p>
<p>Instances of this class delegate to Class1AffinityPredictor and
Class1ProcessingPredictor instances to generate BA and AP predictions.
These predictions are combined using a logistic regression model to give
a “presentation score” prediction.</p>
<p>Most users will call the <a class="reference internal" href="#mhcflurry.Class1PresentationPredictor.load" title="mhcflurry.Class1PresentationPredictor.load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load</span></code></a> static method to get an instance of this
class, then call the <a class="reference internal" href="#mhcflurry.Class1PresentationPredictor.predict" title="mhcflurry.Class1PresentationPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method to generate predictions.</p>
<dl class="attribute">
<dt id="mhcflurry.Class1PresentationPredictor.model_inputs">
<code class="sig-name descname">model_inputs</code><em class="property"> = ['affinity_score', 'processing_score']</em><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.model_inputs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.supported_alleles">
<em class="property">property </em><code class="sig-name descname">supported_alleles</code><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.supported_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>List of alleles supported by the underlying Class1AffinityPredictor</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.supported_peptide_lengths">
<em class="property">property </em><code class="sig-name descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(min, max) of supported peptide lengths, inclusive.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.predict_affinity">
<code class="sig-name descname">predict_affinity</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">alleles</em>, <em class="sig-param">sample_names=None</em>, <em class="sig-param">include_affinity_percentile=True</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.predict_affinity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.predict_affinity" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict binding affinities across samples (each corresponding to up to
six MHC  I alleles).</p>
<p>Two modes are supported: each peptide can be evaluated for binding to
any of the alleles in any sample (this is what happens when sample_names
is None), or the i’th peptide can be evaluated for binding the alleles
of the sample given by the i’th entry in sample_names.</p>
<p>For example, if we don’t specify sample_names, then predictions
are taken for all combinations of samples and peptides, for a result
size of num peptides *  num samples:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1PresentationPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_affinity</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;PEPTIDE&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">    peptide  peptide_num sample_name      affinity best_allele</span>
<span class="go">0  SIINFEKL            0     sample1  12906.787792       A0201</span>
<span class="go">1   PEPTIDE            1     sample1  36827.681130       B0702</span>
<span class="go">2  SIINFEKL            0     sample2   3588.413748       C0202</span>
<span class="go">3   PEPTIDE            1     sample2  34362.109211       C0202</span>
</pre></div>
</div>
<p>In contrast, here we specify sample_names, so peptide is evaluated for
binding the alleles in the corresponding sample, for a result size equal
to the number of peptides:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_affinity</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;PEPTIDE&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">sample_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample2&quot;</span><span class="p">,</span> <span class="s2">&quot;sample1&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">    peptide  peptide_num sample_name      affinity best_allele</span>
<span class="go">0  SIINFEKL            0     sample2   3588.412141       C0202</span>
<span class="go">1   PEPTIDE            1     sample1  36827.682779       B0702</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">dict of string -&gt; list of string</span></dt><dd><p>Keys are sample names, values are the alleles (genotype) for
that sample</p>
</dd>
<dt><strong>sample_names</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>Sample names corresponding to each peptide. If None, then
predictions are generated for all sample genotypes  across all
peptides.</p>
</dd>
<dt><strong>include_affinity_percentile</strong><span class="classifier">bool</span></dt><dd><p>Whether to include affinity percentile ranks</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Set to 0 for quiet.</p>
</dd>
<dt><strong>throw</strong><span class="classifier">verbose</span></dt><dd><p>Whether to throw exception (vs. just log a warning) on invalid
peptides, etc.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>pandas.DataFrame</strong><span class="classifier">predictions</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.predict_processing">
<code class="sig-name descname">predict_processing</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">verbose=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.predict_processing"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.predict_processing" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing scores for individual peptides, optionally
including flanking sequences for better cleavage prediction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd></dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd></dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>numpy.array</strong><span class="classifier">Antigen processing scores for each peptide</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">targets</em>, <em class="sig-param">peptides</em>, <em class="sig-param">sample_names</em>, <em class="sig-param">alleles</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">verbose=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the presentation score logistic regression model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>targets</strong><span class="classifier">list of int/float</span></dt><dd><p>1 indicates hit, 0 indicates decoy</p>
</dd>
<dt><strong>peptides</strong><span class="classifier">list of string [same length as targets]</span></dt><dd></dd>
<dt><strong>sample_names</strong><span class="classifier">list of string [same length as targets]</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">dict of string -&gt; list of string</span></dt><dd><p>Keys are sample names, values are the alleles for that sample</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string [same length as targets]</span></dt><dd></dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string [same length as targets]</span></dt><dd></dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.get_model">
<code class="sig-name descname">get_model</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.get_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.get_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Load or instantiate a new logistic regression model. Private helper
method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name</strong><span class="classifier">string</span></dt><dd><p>If None (the default), an un-fit LR model is returned. Otherwise the
weights are loaded for the specified model.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>sklearn.linear_model.LogisticRegression</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">alleles</em>, <em class="sig-param">sample_names=None</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">include_affinity_percentile=False</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict presentation scores across a set of peptides.</p>
<p>Presentation scores combine predictions for MHC I binding affinity
and antigen processing.</p>
<p>This method returns a pandas.DataFrame giving presentation scores plus
the binding affinity and processing predictions and other intermediate
results.</p>
<p>Example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1PresentationPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;PEPTIDE&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">n_flanks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;NNN&quot;</span><span class="p">,</span> <span class="s2">&quot;SNS&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">c_flanks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CCC&quot;</span><span class="p">,</span> <span class="s2">&quot;CNC&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">    peptide n_flank c_flank  peptide_num sample_name      affinity best_allele  processing_score  presentation_score</span>
<span class="go">0  SIINFEKL     NNN     CCC            0     sample1  12906.787792       A0201          0.802466            0.140365</span>
<span class="go">1   PEPTIDE     SNS     CNC            1     sample1  36827.681130       B0702          0.105260            0.004059</span>
<span class="go">2  SIINFEKL     NNN     CCC            0     sample2   3588.413748       C0202          0.802466            0.338647</span>
<span class="go">3   PEPTIDE     SNS     CNC            1     sample2  34362.109211       C0202          0.105260            0.004317</span>
</pre></div>
</div>
<p>You can also specify sample_names, in which case peptide is evaluated
for binding the alleles in the corresponding sample only. See
<a class="reference internal" href="#mhcflurry.Class1PresentationPredictor.predict_affinity" title="mhcflurry.Class1PresentationPredictor.predict_affinity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict_affinity</span></code></a> for an examples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">list of string or dict of string -&gt; list of string</span></dt><dd><p>If you are predicting for a single sample, pass a list of strings
(up to 6) indicating the genotype. If you are predicting across
multiple samples, pass a dict where the keys are (arbitrary)
sample names and the values are the alleles to predict for that
sample.</p>
</dd>
<dt><strong>sample_names</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>If you are passing a dict for ‘alleles’, you can use this
argument to specify which peptides go with which samples. If it is
None, then predictions will be performed for each peptide across all
samples.</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>Upstream sequences before the peptide. Sequences of any length can
be given and a suffix of the size supported by the model will be
used.</p>
</dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>Downstream sequences after the peptide. Sequences of any length can
be given and a prefix of the size supported by the model will be
used.</p>
</dd>
<dt><strong>include_affinity_percentile</strong><span class="classifier">bool</span></dt><dd><p>Whether to include affinity percentile ranks</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Set to 0 for quiet.</p>
</dd>
<dt><strong>throw</strong><span class="classifier">verbose</span></dt><dd><p>Whether to throw exception (vs. just log a warning) on invalid
peptides, etc.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
<dt>Presentation scores and intermediate results.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.predict_sequences">
<code class="sig-name descname">predict_sequences</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em>, <em class="sig-param">alleles</em>, <em class="sig-param">result='best'</em>, <em class="sig-param">comparison_quantity='presentation_score'</em>, <em class="sig-param">filter_value=None</em>, <em class="sig-param">peptide_lengths=(8</em>, <em class="sig-param">9</em>, <em class="sig-param">10</em>, <em class="sig-param">11)</em>, <em class="sig-param">use_flanks=True</em>, <em class="sig-param">include_affinity_percentile=True</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.predict_sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.predict_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict presentation across protein sequences.</p>
<p>Example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1PresentationPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_sequences</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">sequences</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s1">&#39;protein1&#39;</span><span class="p">:</span> <span class="s2">&quot;MDSKGSSQKGSRLLLLLVVSNLL&quot;</span><span class="p">,</span>
<span class="gp">... </span>       <span class="s1">&#39;protein2&#39;</span><span class="p">:</span> <span class="s2">&quot;SSLPTPEDKEQAQQTHH&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">result</span><span class="o">=</span><span class="s2">&quot;filtered&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">comparison_quantity</span><span class="o">=</span><span class="s2">&quot;affinity&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">filter_value</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">  sequence_name  pos     peptide         n_flank     c_flank sample_name    affinity best_allele  affinity_percentile  processing_score  presentation_score</span>
<span class="go">0      protein1   13   LLLLVVSNL   MDSKGSSQKGSRL           L     sample1   38.206225       A0201             0.380125          0.017644            0.571060</span>
<span class="go">1      protein1   14   LLLVVSNLL  MDSKGSSQKGSRLL                 sample1   42.243472       A0201             0.420250          0.090984            0.619213</span>
<span class="go">2      protein1    5   SSQKGSRLL           MDSKG   LLLVVSNLL     sample2   66.749223       C0202             0.803375          0.383608            0.774468</span>
<span class="go">3      protein1    6   SQKGSRLLL          MDSKGS    LLVVSNLL     sample2  178.033474       C0202             1.820000          0.275019            0.482206</span>
<span class="go">4      protein1   13  LLLLVVSNLL   MDSKGSSQKGSRL                 sample1  202.208167       A0201             1.112500          0.058782            0.261320</span>
<span class="go">5      protein1   12  LLLLLVVSNL    MDSKGSSQKGSR           L     sample1  202.506582       A0201             1.112500          0.010025            0.225648</span>
<span class="go">6      protein2    0   SSLPTPEDK                    EQAQQTHH     sample1  335.529377       A0301             1.011750          0.010443            0.156798</span>
<span class="go">7      protein2    0   SSLPTPEDK                    EQAQQTHH     sample2  353.451759       C0202             2.674250          0.010443            0.150753</span>
<span class="go">8      protein1    8   KGSRLLLLL        MDSKGSSQ      VVSNLL     sample2  410.327286       C0202             2.887000          0.121374            0.194081</span>
<span class="go">9      protein1    5    SSQKGSRL           MDSKG  LLLLVVSNLL     sample2  477.285954       C0202             3.107375          0.111982            0.168572</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">str, list of string, or string -&gt; string dict</span></dt><dd><p>Protein sequences. If a dict is given, the keys are arbitrary (
e.g. protein names), and the values are the amino acid sequences.</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">list of string, list of list of string, or dict of string -&gt; list of string</span></dt><dd><p>MHC I alleles. Can be: (1) a string (a single allele), (2) a list of
strings (a single genotype), (3) a list of list of strings
(multiple genotypes, where the total number of genotypes must equal
the number of sequences), or (4) a dict giving multiple genotypes,
which will each be run over the sequences.</p>
</dd>
<dt><strong>result</strong><span class="classifier">string</span></dt><dd><p>Specify ‘best’ to return the strongest peptide for each sequence,
‘all’ to return predictions for all peptides, or ‘filtered’ to
return predictions where the comparison_quantity is stronger
(i.e (&lt;) for affinity, (&gt;) for scores) than filter_value.</p>
</dd>
<dt><strong>comparison_quantity</strong><span class="classifier">string</span></dt><dd><p>One of “presentation_score”, “processing_score”, “affinity”, or
“affinity_percentile”. Prediction to use to rank (if result is
“best”) or filter (if result is “filtered”) results.</p>
</dd>
<dt><strong>filter_value</strong><span class="classifier">float</span></dt><dd><p>Threshold value to use, only relevant when result is “filtered”.
If comparison_quantity is “affinity”, then all results less than
(i.e. tighter than) the specified nM affinity are retained. If it’s
“presentation_score” or “processing_score” then results greater than
the indicated filter_value are retained.</p>
</dd>
<dt><strong>peptide_lengths</strong><span class="classifier">list of int</span></dt><dd><p>Peptide lengths to predict for.</p>
</dd>
<dt><strong>use_flanks</strong><span class="classifier">bool</span></dt><dd><p>Whether to include flanking sequences when running the AP predictor
(for better cleavage prediction).</p>
</dd>
<dt><strong>include_affinity_percentile</strong><span class="classifier">bool</span></dt><dd><p>Whether to include affinity percentile ranks in output.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Set to 0 for quiet mode.</p>
</dd>
<dt><strong>throw</strong><span class="classifier">boolean</span></dt><dd><p>Whether to throw exceptions (vs. log warnings) on invalid inputs.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame with columns:</dt><dd><p>peptide, n_flank, c_flank, sequence_name, affinity, best_allele,
processing_score, presentation_score</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.save">
<code class="sig-name descname">save</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">models_dir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Save the predictor to a directory on disk. If the directory does
not exist it will be created.</p>
<p>The wrapped Class1AffinityPredictor and Class1ProcessingPredictor
instances are included in the saved data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. It will be created if it doesn’t exist.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.Class1PresentationPredictor.load">
<em class="property">classmethod </em><code class="sig-name descname">load</code><span class="sig-paren">(</span><em class="sig-param">models_dir=None</em>, <em class="sig-param">max_models=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.Class1PresentationPredictor.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserialize a predictor from a directory on disk.</p>
<p>This will also load the wrapped Class1AffinityPredictor and
Class1ProcessingPredictor instances.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. If unspecified the default downloaded models are
used.</p>
</dd>
<dt><strong>max_models</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum number of affinity and processing (counted separately)
models to load</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#mhcflurry.Class1PresentationPredictor" title="mhcflurry.Class1PresentationPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1PresentationPredictor</span></code></a> instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-mhcflurry.allele_encoding">
<span id="mhcflurry-allele-encoding-module"></span><h2>mhcflurry.allele_encoding module<a class="headerlink" href="#module-mhcflurry.allele_encoding" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.allele_encoding.AlleleEncoding">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.allele_encoding.</code><code class="sig-name descname">AlleleEncoding</code><span class="sig-paren">(</span><em class="sig-param">alleles=None</em>, <em class="sig-param">allele_to_sequence=None</em>, <em class="sig-param">borrow_from=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/allele_encoding.html#AlleleEncoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.allele_encoding.AlleleEncoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A place to cache encodings for a sequence of alleles.</p>
<p>We frequently work with alleles by integer indices, for example as
inputs to neural networks. This class is used to map allele names to
integer indices in a consistent way by keeping track of the universe
of alleles under use, i.e. a distinction is made between the universe
of supported alleles (what’s in <code class="xref py py-obj docutils literal notranslate"><span class="pre">allele_to_sequence</span></code>) and the actual
set of alleles used for some task (what’s in <code class="xref py py-obj docutils literal notranslate"><span class="pre">alleles</span></code>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>alleles</strong><span class="classifier">list of string</span></dt><dd><p>Allele names. If any allele is None instead of string, it will be
mapped to the special index value -1.</p>
</dd>
<dt><strong>allele_to_sequence</strong><span class="classifier">dict of str -&gt; str</span></dt><dd><p>Allele name to amino acid sequence</p>
</dd>
<dt><strong>borrow_from</strong><span class="classifier">AlleleEncoding, optional</span></dt><dd><p>If specified, do not specify allele_to_sequence. The sequences from
the provided instance are used. This guarantees that the mappings
from allele to index and from allele to sequence are the same
between the instances.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="mhcflurry.allele_encoding.AlleleEncoding.compact">
<code class="sig-name descname">compact</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/allele_encoding.html#AlleleEncoding.compact"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.allele_encoding.AlleleEncoding.compact" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new AlleleEncoding in which the universe of supported alleles
is only the alleles actually used.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>AlleleEncoding</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.allele_encoding.AlleleEncoding.allele_representations">
<code class="sig-name descname">allele_representations</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">encoding_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/allele_encoding.html#AlleleEncoding.allele_representations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.allele_encoding.AlleleEncoding.allele_representations" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode the universe of supported allele sequences to a matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>encoding_name</strong><span class="classifier">string</span></dt><dd><p>How to represent amino acids. Valid names are “BLOSUM62” or
“one-hot”. See <code class="xref py py-obj docutils literal notranslate"><span class="pre">amino_acid.ENCODING_DATA_FRAMES</span></code>.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of shape</dt><dd><p>(num alleles in universe, sequence length, vector size)</p>
</dd>
<dt>where vector size is usually 21 (20 amino acids + X character)</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.allele_encoding.AlleleEncoding.fixed_length_vector_encoded_sequences">
<code class="sig-name descname">fixed_length_vector_encoded_sequences</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">encoding_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/allele_encoding.html#AlleleEncoding.fixed_length_vector_encoded_sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.allele_encoding.AlleleEncoding.fixed_length_vector_encoded_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode allele sequences (not the universe of alleles) to a matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>encoding_name</strong><span class="classifier">string</span></dt><dd><p>How to represent amino acids. Valid names are “BLOSUM62” or
“one-hot”. See <code class="xref py py-obj docutils literal notranslate"><span class="pre">amino_acid.ENCODING_DATA_FRAMES</span></code>.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array with shape:</dt><dd><p>(num alleles, sequence length, vector size)</p>
</dd>
<dt>where vector size is usually 21 (20 amino acids + X character)</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.amino_acid">
<span id="mhcflurry-amino-acid-module"></span><h2>mhcflurry.amino_acid module<a class="headerlink" href="#module-mhcflurry.amino_acid" title="Permalink to this headline">¶</a></h2>
<p>Functions for encoding fixed length sequences of amino acids into various
vector representations, such as one-hot and BLOSUM62.</p>
<dl class="function">
<dt id="mhcflurry.amino_acid.available_vector_encodings">
<code class="sig-prename descclassname">mhcflurry.amino_acid.</code><code class="sig-name descname">available_vector_encodings</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/amino_acid.html#available_vector_encodings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.amino_acid.available_vector_encodings" title="Permalink to this definition">¶</a></dt>
<dd><p>Return list of supported amino acid vector encodings.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.amino_acid.vector_encoding_length">
<code class="sig-prename descclassname">mhcflurry.amino_acid.</code><code class="sig-name descname">vector_encoding_length</code><span class="sig-paren">(</span><em class="sig-param">name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/amino_acid.html#vector_encoding_length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.amino_acid.vector_encoding_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the given vector encoding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>int</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.amino_acid.index_encoding">
<code class="sig-prename descclassname">mhcflurry.amino_acid.</code><code class="sig-name descname">index_encoding</code><span class="sig-paren">(</span><em class="sig-param">sequences</em>, <em class="sig-param">letter_to_index_dict</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/amino_acid.html#index_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.amino_acid.index_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode a sequence of same-length strings to a matrix of integers of the
same shape. The map from characters to integers is given by
<code class="xref py py-obj docutils literal notranslate"><span class="pre">letter_to_index_dict</span></code>.</p>
<p>Given a sequence of <code class="xref py py-obj docutils literal notranslate"><span class="pre">n</span></code> strings all of length <code class="xref py py-obj docutils literal notranslate"><span class="pre">k</span></code>, return a <code class="xref py py-obj docutils literal notranslate"><span class="pre">k</span> <span class="pre">*</span> <span class="pre">n</span></code> array where
the (<code class="xref py py-obj docutils literal notranslate"><span class="pre">i</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">j</span></code>)th element is <code class="xref py py-obj docutils literal notranslate"><span class="pre">letter_to_index_dict[sequence[i][j]]</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">list of length n of strings of length k</span></dt><dd></dd>
<dt><strong>letter_to_index_dict</strong><span class="classifier">dict</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of integers with shape (<code class="xref py py-obj docutils literal notranslate"><span class="pre">k</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">n</span></code>)</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.amino_acid.fixed_vectors_encoding">
<code class="sig-prename descclassname">mhcflurry.amino_acid.</code><code class="sig-name descname">fixed_vectors_encoding</code><span class="sig-paren">(</span><em class="sig-param">index_encoded_sequences</em>, <em class="sig-param">letter_to_vector_df</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/amino_acid.html#fixed_vectors_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.amino_acid.fixed_vectors_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a <code class="xref py py-obj docutils literal notranslate"><span class="pre">n</span></code> x <code class="xref py py-obj docutils literal notranslate"><span class="pre">k</span></code> matrix of integers such as that returned by <a class="reference internal" href="#mhcflurry.amino_acid.index_encoding" title="mhcflurry.amino_acid.index_encoding"><code class="xref py py-obj docutils literal notranslate"><span class="pre">index_encoding()</span></code></a> and
a dataframe mapping each index to an arbitrary vector, return a <code class="xref py py-obj docutils literal notranslate"><span class="pre">n</span> <span class="pre">*</span> <span class="pre">k</span> <span class="pre">*</span> <span class="pre">m</span></code>
array where the (<code class="xref py py-obj docutils literal notranslate"><span class="pre">i</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">j</span></code>)’th element is <code class="xref py py-obj docutils literal notranslate"><span class="pre">letter_to_vector_df.iloc[sequence[i][j]]</span></code>.</p>
<p>The dataframe index and columns names are ignored here; the indexing is done
entirely by integer position in the dataframe.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>index_encoded_sequences</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">n</span></code> x <code class="xref py py-obj docutils literal notranslate"><span class="pre">k</span></code> array of integers</span></dt><dd></dd>
<dt><strong>letter_to_vector_df</strong><span class="classifier">pandas.DataFrame of shape (<code class="xref py py-obj docutils literal notranslate"><span class="pre">alphabet</span> <span class="pre">size</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">m</span></code>)</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of integers with shape (<code class="xref py py-obj docutils literal notranslate"><span class="pre">n</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">k</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">m</span></code>)</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.calibrate_percentile_ranks_command">
<span id="mhcflurry-calibrate-percentile-ranks-command-module"></span><h2>mhcflurry.calibrate_percentile_ranks_command module<a class="headerlink" href="#module-mhcflurry.calibrate_percentile_ranks_command" title="Permalink to this headline">¶</a></h2>
<p>Calibrate percentile ranks for models. Runs in-place.</p>
<dl class="function">
<dt id="mhcflurry.calibrate_percentile_ranks_command.run">
<code class="sig-prename descclassname">mhcflurry.calibrate_percentile_ranks_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/calibrate_percentile_ranks_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.calibrate_percentile_ranks_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.calibrate_percentile_ranks_command.do_calibrate_percentile_ranks">
<code class="sig-prename descclassname">mhcflurry.calibrate_percentile_ranks_command.</code><code class="sig-name descname">do_calibrate_percentile_ranks</code><span class="sig-paren">(</span><em class="sig-param">alleles</em>, <em class="sig-param">constant_data={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/calibrate_percentile_ranks_command.html#do_calibrate_percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.calibrate_percentile_ranks_command.do_calibrate_percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.calibrate_percentile_ranks_command.calibrate_percentile_ranks">
<code class="sig-prename descclassname">mhcflurry.calibrate_percentile_ranks_command.</code><code class="sig-name descname">calibrate_percentile_ranks</code><span class="sig-paren">(</span><em class="sig-param">allele, predictor, peptides=None, motif_summary=False, summary_top_peptide_fractions=[0.001], verbose=False, model_kwargs={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/calibrate_percentile_ranks_command.html#calibrate_percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.calibrate_percentile_ranks_command.calibrate_percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.class1_affinity_predictor">
<span id="mhcflurry-class1-affinity-predictor-module"></span><h2>mhcflurry.class1_affinity_predictor module<a class="headerlink" href="#module-mhcflurry.class1_affinity_predictor" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.class1_affinity_predictor.</code><code class="sig-name descname">Class1AffinityPredictor</code><span class="sig-paren">(</span><em class="sig-param">allele_to_allele_specific_models=None</em>, <em class="sig-param">class1_pan_allele_models=None</em>, <em class="sig-param">allele_to_sequence=None</em>, <em class="sig-param">manifest_df=None</em>, <em class="sig-param">allele_to_percent_rank_transform=None</em>, <em class="sig-param">metadata_dataframes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>High-level interface for peptide/MHC I binding affinity prediction.</p>
<p>This class manages low-level <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code> instances, each of which
wraps a single Keras network. The purpose of <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> is to
implement ensembles, handling of multiple alleles, and predictor loading and
saving. It also provides a place to keep track of metadata like prediction
histograms for percentile rank calibration.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>allele_to_allele_specific_models</strong><span class="classifier">dict of string -&gt; list of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></span></dt><dd><p>Ensemble of single-allele models to use for each allele.</p>
</dd>
<dt><strong>class1_pan_allele_models</strong><span class="classifier">list of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></span></dt><dd><p>Ensemble of pan-allele models.</p>
</dd>
<dt><strong>allele_to_sequence</strong><span class="classifier">dict of string -&gt; string</span></dt><dd><p>MHC allele name to fixed-length amino acid sequence (sometimes
referred to as the pseudosequence). Required only if
class1_pan_allele_models is specified.</p>
</dd>
<dt><strong>manifest_df</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>, optional</span></dt><dd><p>Must have columns: model_name, allele, config_json, model.
Only required if you want to update an existing serialization of a
Class1AffinityPredictor. Otherwise this dataframe will be generated
automatically based on the supplied models.</p>
</dd>
<dt><strong>allele_to_percent_rank_transform</strong><span class="classifier">dict of string -&gt; <code class="xref py py-obj docutils literal notranslate"><span class="pre">PercentRankTransform</span></code>, optional</span></dt><dd><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">PercentRankTransform</span></code> instances to use for each allele</p>
</dd>
<dt><strong>metadata_dataframes</strong><span class="classifier">dict of string -&gt; pandas.DataFrame, optional</span></dt><dd><p>Optional additional dataframes to write to the models dir when
save() is called. Useful for tracking provenance.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.manifest_df">
<em class="property">property </em><code class="sig-name descname">manifest_df</code><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.manifest_df" title="Permalink to this definition">¶</a></dt>
<dd><p>A pandas.DataFrame describing the models included in this predictor.</p>
<p>Based on:
- self.class1_pan_allele_models
- self.allele_to_allele_specific_models</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.clear_cache">
<code class="sig-name descname">clear_cache</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.clear_cache"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.clear_cache" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear values cached based on the neural networks in this predictor.</p>
<dl class="simple">
<dt>Users should call this after mutating any of the following:</dt><dd><ul class="simple">
<li><p>self.class1_pan_allele_models</p></li>
<li><p>self.allele_to_allele_specific_models</p></li>
<li><p>self.allele_to_sequence</p></li>
</ul>
</dd>
</dl>
<p>Methods that mutate these instance variables will call this method on
their own if needed.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.neural_networks">
<em class="property">property </em><code class="sig-name descname">neural_networks</code><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.neural_networks" title="Permalink to this definition">¶</a></dt>
<dd><p>List of the neural networks in the ensemble.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">predictors</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge the ensembles of two or more <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> instances.</p>
<p>Note: the resulting merged predictor will NOT have calibrated percentile
ranks. Call <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.calibrate_percentile_ranks" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.calibrate_percentile_ranks"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calibrate_percentile_ranks</span></code></a> on it if these are needed.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>predictors</strong><span class="classifier">sequence of <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a></span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.merge_in_place">
<code class="sig-name descname">merge_in_place</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">others</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.merge_in_place"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.merge_in_place" title="Permalink to this definition">¶</a></dt>
<dd><p>Add the models present in other predictors into the current predictor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>others</strong><span class="classifier">list of Class1AffinityPredictor</span></dt><dd><p>Other predictors to merge into the current predictor.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>list of string</strong><span class="classifier">names of newly added models</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.supported_alleles">
<em class="property">property </em><code class="sig-name descname">supported_alleles</code><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.supported_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Alleles for which predictions can be made.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.supported_peptide_lengths">
<em class="property">property </em><code class="sig-name descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(minimum, maximum) lengths of peptides supported by <em>all models</em>,
inclusive.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>(int, int) tuple</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.check_consistency">
<code class="sig-name descname">check_consistency</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.check_consistency"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.check_consistency" title="Permalink to this definition">¶</a></dt>
<dd><p>Verify that self.manifest_df is consistent with:
- self.class1_pan_allele_models
- self.allele_to_allele_specific_models</p>
<p>Currently only checks for agreement on the total number of models.</p>
<p>Throws AssertionError if inconsistent.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.save">
<code class="sig-name descname">save</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">models_dir</em>, <em class="sig-param">model_names_to_write=None</em>, <em class="sig-param">write_metadata=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the predictor to a directory on disk. If the directory does
not exist it will be created.</p>
<p>The serialization format consists of a file called “manifest.csv” with
the configurations of each Class1NeuralNetwork, along with per-network
files giving the model weights. If there are pan-allele predictors in
the ensemble, the allele sequences are also stored in the
directory. There is also a small file “index.txt” with basic metadata:
when the models were trained, by whom, on what host.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. It will be created if it doesn’t exist.</p>
</dd>
<dt><strong>model_names_to_write</strong><span class="classifier">list of string, optional</span></dt><dd><p>Only write the weights for the specified models. Useful for
incremental updates during training.</p>
</dd>
<dt><strong>write_metadata</strong><span class="classifier">boolean, optional</span></dt><dd><p>Whether to write optional metadata</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.load">
<em class="property">static </em><code class="sig-name descname">load</code><span class="sig-paren">(</span><em class="sig-param">models_dir=None</em>, <em class="sig-param">max_models=None</em>, <em class="sig-param">optimization_level=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserialize a predictor from a directory on disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. If unspecified the default downloaded models are
used.</p>
</dd>
<dt><strong>max_models</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum number of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code> instances to load</p>
</dd>
<dt><strong>optimization_level</strong><span class="classifier">int</span></dt><dd><p>If &gt;0, model optimization will be attempted. Defaults to value of
environment variable MHCFLURRY_OPTIMIZATION_LEVEL.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1AffinityPredictor</span></code></a> instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.optimize">
<code class="sig-name descname">optimize</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">warn=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.optimize"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.optimize" title="Permalink to this definition">¶</a></dt>
<dd><p>EXPERIMENTAL: Optimize the predictor for faster predictions.</p>
<p>Currently the only optimization implemented is to merge multiple pan-
allele predictors at the tensorflow level.</p>
<p>The optimization is performed in-place, mutating the instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>bool</dt><dd><p>Whether optimization was performed</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.model_name">
<em class="property">static </em><code class="sig-name descname">model_name</code><span class="sig-paren">(</span><em class="sig-param">allele</em>, <em class="sig-param">num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.model_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.model_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a model name</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>num</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.weights_path">
<em class="property">static </em><code class="sig-name descname">weights_path</code><span class="sig-paren">(</span><em class="sig-param">models_dir</em>, <em class="sig-param">model_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.weights_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.weights_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the path to the weights file for a model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>model_name</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.master_allele_encoding">
<em class="property">property </em><code class="sig-name descname">master_allele_encoding</code><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.master_allele_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>An AlleleEncoding containing the universe of alleles specified by
self.allele_to_sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>AlleleEncoding</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.fit_allele_specific_predictors">
<code class="sig-name descname">fit_allele_specific_predictors</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">n_models</em>, <em class="sig-param">architecture_hyperparameters_list</em>, <em class="sig-param">allele</em>, <em class="sig-param">peptides</em>, <em class="sig-param">affinities</em>, <em class="sig-param">inequalities=None</em>, <em class="sig-param">train_rounds=None</em>, <em class="sig-param">models_dir_for_save=None</em>, <em class="sig-param">verbose=0</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.fit_allele_specific_predictors"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.fit_allele_specific_predictors" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit one or more allele specific predictors for a single allele using one
or more neural network architectures.</p>
<p>The new predictors are saved in the Class1AffinityPredictor instance
and will be used on subsequent calls to <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>n_models</strong><span class="classifier">int</span></dt><dd><p>Number of neural networks to fit</p>
</dd>
<dt><strong>architecture_hyperparameters_list</strong><span class="classifier">list of dict</span></dt><dd><p>List of hyperparameter sets.</p>
</dd>
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>peptides</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodableSequences</span></code> or list of string</span></dt><dd></dd>
<dt><strong>affinities</strong><span class="classifier">list of float</span></dt><dd><p>nM affinities</p>
</dd>
<dt><strong>inequalities</strong><span class="classifier">list of string, each element one of “&gt;”, “&lt;”, or “=”</span></dt><dd><p>See <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork.fit</span></code> for details.</p>
</dd>
<dt><strong>train_rounds</strong><span class="classifier">sequence of int</span></dt><dd><p>Each training point i will be used on training rounds r for which
train_rounds[i] &gt; r, r &gt;= 0.</p>
</dd>
<dt><strong>models_dir_for_save</strong><span class="classifier">string, optional</span></dt><dd><p>If specified, the Class1AffinityPredictor is (incrementally) written
to the given models dir after each neural network is fit.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Keras verbosity</p>
</dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd><p>Optional string of information to include in each progress update</p>
</dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd><p>How often (in seconds) to print progress. Set to None to disable.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.fit_class1_pan_allele_models">
<code class="sig-name descname">fit_class1_pan_allele_models</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">n_models</em>, <em class="sig-param">architecture_hyperparameters</em>, <em class="sig-param">alleles</em>, <em class="sig-param">peptides</em>, <em class="sig-param">affinities</em>, <em class="sig-param">inequalities</em>, <em class="sig-param">models_dir_for_save=None</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.fit_class1_pan_allele_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.fit_class1_pan_allele_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit one or more pan-allele predictors using a single neural network
architecture.</p>
<p>The new predictors are saved in the Class1AffinityPredictor instance
and will be used on subsequent calls to <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>n_models</strong><span class="classifier">int</span></dt><dd><p>Number of neural networks to fit</p>
</dd>
<dt><strong>architecture_hyperparameters</strong><span class="classifier">dict</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">list of string</span></dt><dd><p>Allele names (not sequences) corresponding to each peptide</p>
</dd>
<dt><strong>peptides</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodableSequences</span></code> or list of string</span></dt><dd></dd>
<dt><strong>affinities</strong><span class="classifier">list of float</span></dt><dd><p>nM affinities</p>
</dd>
<dt><strong>inequalities</strong><span class="classifier">list of string, each element one of “&gt;”, “&lt;”, or “=”</span></dt><dd><p>See Class1NeuralNetwork.fit for details.</p>
</dd>
<dt><strong>models_dir_for_save</strong><span class="classifier">string, optional</span></dt><dd><p>If specified, the Class1AffinityPredictor is (incrementally) written
to the given models dir after each neural network is fit.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Keras verbosity</p>
</dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd><p>Optional string of information to include in each progress update</p>
</dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd><p>How often (in seconds) to print progress. Set to None to disable.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of <code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1NeuralNetwork</span></code></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.add_pan_allele_model">
<code class="sig-name descname">add_pan_allele_model</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">model</em>, <em class="sig-param">models_dir_for_save=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.add_pan_allele_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.add_pan_allele_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a pan-allele model to the ensemble and optionally do an incremental
save.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>model</strong><span class="classifier">Class1NeuralNetwork</span></dt><dd></dd>
<dt><strong>models_dir_for_save</strong><span class="classifier">string</span></dt><dd><p>Directory to save resulting ensemble to</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.percentile_ranks">
<code class="sig-name descname">percentile_ranks</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">affinities</em>, <em class="sig-param">allele=None</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Return percentile ranks for the given ic50 affinities and alleles.</p>
<p>The ‘allele’ and ‘alleles’ argument are as in the <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method.
Specify one of these.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>affinities</strong><span class="classifier">sequence of float</span></dt><dd><p>nM affinities</p>
</dd>
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">sequence of string</span></dt><dd></dd>
<dt><strong>throw</strong><span class="classifier">boolean</span></dt><dd><p>If True, a ValueError will be raised in the case of unsupported
alleles. If False, a warning will be logged and NaN will be returned
for those percentile ranks.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of float</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">allele=None</em>, <em class="sig-param">throw=True</em>, <em class="sig-param">centrality_measure='mean'</em>, <em class="sig-param">model_kwargs={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict nM binding affinities.</p>
<p>If multiple predictors are available for an allele, the predictions are
the geometric means of the individual model (nM) predictions.</p>
<p>One of ‘allele’ or ‘alleles’ must be specified. If ‘allele’ is specified
all predictions will be for the given allele. If ‘alleles’ is specified
it must be the same length as ‘peptides’ and give the allele
corresponding to each peptide.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>peptides</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodableSequences</span></code> or list of string</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>throw</strong><span class="classifier">boolean</span></dt><dd><p>If True, a ValueError will be raised in the case of unsupported
alleles or peptide lengths. If False, a warning will be logged and
the predictions for the unsupported alleles or peptides will be NaN.</p>
</dd>
<dt><strong>centrality_measure</strong><span class="classifier">string or callable</span></dt><dd><p>Measure of central tendency to use to combine predictions in the
ensemble. Options include: mean, median, robust_mean.</p>
</dd>
<dt><strong>model_kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments to pass to Class1NeuralNetwork.predict</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of predictions</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict_to_dataframe">
<code class="sig-name descname">predict_to_dataframe</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">allele=None</em>, <em class="sig-param">throw=True</em>, <em class="sig-param">include_individual_model_predictions=False</em>, <em class="sig-param">include_percentile_ranks=True</em>, <em class="sig-param">include_confidence_intervals=True</em>, <em class="sig-param">centrality_measure='mean'</em>, <em class="sig-param">model_kwargs={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.predict_to_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict_to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict nM binding affinities. Gives more detailed output than <a class="reference internal" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict" title="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a>
method, including 5-95% prediction intervals.</p>
<p>If multiple predictors are available for an allele, the predictions are
the geometric means of the individual model predictions.</p>
<p>One of ‘allele’ or ‘alleles’ must be specified. If ‘allele’ is specified
all predictions will be for the given allele. If ‘alleles’ is specified
it must be the same length as ‘peptides’ and give the allele
corresponding to each peptide.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>peptides</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">EncodableSequences</span></code> or list of string</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>allele</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>throw</strong><span class="classifier">boolean</span></dt><dd><p>If True, a ValueError will be raised in the case of unsupported
alleles or peptide lengths. If False, a warning will be logged and
the predictions for the unsupported alleles or peptides will be NaN.</p>
</dd>
<dt><strong>include_individual_model_predictions</strong><span class="classifier">boolean</span></dt><dd><p>If True, the predictions of each individual model are included as
columns in the result DataFrame.</p>
</dd>
<dt><strong>include_percentile_ranks</strong><span class="classifier">boolean, default True</span></dt><dd><p>If True, a “prediction_percentile” column will be included giving
the percentile ranks. If no percentile rank info is available,
this will be ignored with a warning.</p>
</dd>
<dt><strong>centrality_measure</strong><span class="classifier">string or callable</span></dt><dd><p>Measure of central tendency to use to combine predictions in the
ensemble. Options include: mean, median, robust_mean.</p>
</dd>
<dt><strong>model_kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments to pass to Class1NeuralNetwork.predict</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> of predictions</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.calibrate_percentile_ranks">
<code class="sig-name descname">calibrate_percentile_ranks</code><span class="sig-paren">(</span><em class="sig-param">self, peptides=None, num_peptides_per_length=100000, alleles=None, bins=None, motif_summary=False, summary_top_peptide_fractions=[0.001], verbose=False, model_kwargs={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.calibrate_percentile_ranks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.calibrate_percentile_ranks" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the cumulative distribution of ic50 values for a set of alleles
over a large universe of random peptides, to enable taking quantiles
of this distribution later.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">sequence of string or EncodableSequences, optional</span></dt><dd><p>Peptides to use</p>
</dd>
<dt><strong>num_peptides_per_length</strong><span class="classifier">int, optional</span></dt><dd><p>If peptides argument is not specified, then num_peptides_per_length
peptides are randomly sampled from a uniform distribution for each
supported length</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">sequence of string, optional</span></dt><dd><p>Alleles to perform calibration for. If not specified all supported
alleles will be calibrated.</p>
</dd>
<dt><strong>bins</strong><span class="classifier">object</span></dt><dd><p>Anything that can be passed to numpy.histogram’s “bins” argument
can be used here, i.e. either an integer or a sequence giving bin
edges. This is in ic50 space.</p>
</dd>
<dt><strong>motif_summary</strong><span class="classifier">bool</span></dt><dd><p>If True, the length distribution and per-position amino acid
frequencies are also calculated for the top x fraction of tightest-
binding peptides, where each value of x is given in the
summary_top_peptide_fractions list.</p>
</dd>
<dt><strong>summary_top_peptide_fractions</strong><span class="classifier">list of float</span></dt><dd><p>Only used if motif_summary is True</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">boolean</span></dt><dd><p>Whether to print status updates to stdout</p>
</dd>
<dt><strong>model_kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional low-level Class1NeuralNetwork.predict() kwargs.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>dict of string -&gt; pandas.DataFrame</dt><dd></dd>
<dt>If motif_summary is True, this will have keys  “frequency_matrices” and</dt><dd></dd>
<dt>“length_distributions”. Otherwise it will be empty.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.model_select">
<code class="sig-name descname">model_select</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">score_function</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">min_models=1</em>, <em class="sig-param">max_models=10000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_affinity_predictor.html#Class1AffinityPredictor.model_select"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_affinity_predictor.Class1AffinityPredictor.model_select" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform model selection using a user-specified scoring function.</p>
<p>This works only with allele-specific models, not pan-allele models.</p>
<p>Model selection is done using a “step up” variable selection procedure,
in which models are repeatedly added to an ensemble until the score
stops improving.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>score_function</strong><span class="classifier">Class1AffinityPredictor -&gt; float function</span></dt><dd><p>Scoring function</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">list of string, optional</span></dt><dd><p>If not specified, model selection is performed for all alleles.</p>
</dd>
<dt><strong>min_models</strong><span class="classifier">int, optional</span></dt><dd><p>Min models to select per allele</p>
</dd>
<dt><strong>max_models</strong><span class="classifier">int, optional</span></dt><dd><p>Max models to select per allele</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>Class1AffinityPredictor</strong><span class="classifier">predictor containing the selected models</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.class1_neural_network">
<span id="mhcflurry-class1-neural-network-module"></span><h2>mhcflurry.class1_neural_network module<a class="headerlink" href="#module-mhcflurry.class1_neural_network" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.class1_neural_network.</code><code class="sig-name descname">Class1NeuralNetwork</code><span class="sig-paren">(</span><em class="sig-param">**hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Low level class I predictor consisting of a single neural network.</p>
<p>Both single allele and pan-allele prediction are supported.</p>
<p>Users will generally use Class1AffinityPredictor, which gives a higher-level
interface and supports ensembles.</p>
<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.network_hyperparameter_defaults">
<code class="sig-name descname">network_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.network_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters (and their default values) that affect the neural network
architecture.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.compile_hyperparameter_defaults">
<code class="sig-name descname">compile_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.compile_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Loss and optimizer hyperparameters.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.fit_hyperparameter_defaults">
<code class="sig-name descname">fit_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.fit_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for neural network training.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.early_stopping_hyperparameter_defaults">
<code class="sig-name descname">early_stopping_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.early_stopping_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for early stopping.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.miscelaneous_hyperparameter_defaults">
<code class="sig-name descname">miscelaneous_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.miscelaneous_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Miscelaneous hyperaparameters. These parameters are not used by this class
but may be interpreted by other code.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.hyperparameter_defaults">
<code class="sig-name descname">hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Combined set of all supported hyperparameters and their default values.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.hyperparameter_renames">
<code class="sig-name descname">hyperparameter_renames</code><em class="property"> = {'embedding_init_method': None, 'embedding_input_dim': None, 'embedding_output_dim': None, 'kmer_size': None, 'left_edge': None, 'min_delta': None, 'mode': None, 'monitor': None, 'peptide_amino_acid_encoding': None, 'pseudosequence_use_embedding': None, 'right_edge': None, 'take_best_epoch': None, 'use_embedding': None, 'verbose': None}</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.hyperparameter_renames" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.apply_hyperparameter_renames">
<em class="property">classmethod </em><code class="sig-name descname">apply_hyperparameter_renames</code><span class="sig-paren">(</span><em class="sig-param">hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.apply_hyperparameter_renames"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.apply_hyperparameter_renames" title="Permalink to this definition">¶</a></dt>
<dd><p>Handle hyperparameter renames.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>hyperparameters</strong><span class="classifier">dict</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>dict</strong><span class="classifier">updated hyperparameters</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.KERAS_MODELS_CACHE">
<code class="sig-name descname">KERAS_MODELS_CACHE</code><em class="property"> = {}</em><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.KERAS_MODELS_CACHE" title="Permalink to this definition">¶</a></dt>
<dd><p>Process-wide keras model cache, a map from: architecture JSON string to
(Keras model, existing network weights)</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.clear_model_cache">
<em class="property">classmethod </em><code class="sig-name descname">clear_model_cache</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.clear_model_cache"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.clear_model_cache" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear the Keras model cache.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.borrow_cached_network">
<em class="property">classmethod </em><code class="sig-name descname">borrow_cached_network</code><span class="sig-paren">(</span><em class="sig-param">network_json</em>, <em class="sig-param">network_weights</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.borrow_cached_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.borrow_cached_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a keras Model with the specified architecture and weights.
As an optimization, when possible this will reuse architectures from a
process-wide cache.</p>
<p>The returned object is “borrowed” in the sense that its weights can
change later after subsequent calls to this method from other objects.</p>
<p>If you’re using this from a parallel implementation you’ll need to
hold a lock while using the returned object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>network_json</strong><span class="classifier">string of JSON</span></dt><dd></dd>
<dt><strong>network_weights</strong><span class="classifier">list of numpy.array</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>keras.models.Model</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.network">
<code class="sig-name descname">network</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">borrow=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the keras model associated with this predictor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>borrow</strong><span class="classifier">bool</span></dt><dd><p>Whether to return a cached model if possible. See
borrow_cached_network for details</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>keras.models.Model</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.update_network_description">
<code class="sig-name descname">update_network_description</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.update_network_description"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.update_network_description" title="Permalink to this definition">¶</a></dt>
<dd><p>Update self.network_json and self.network_weights properties based on
this instances’s neural network.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.keras_network_cache_key">
<em class="property">static </em><code class="sig-name descname">keras_network_cache_key</code><span class="sig-paren">(</span><em class="sig-param">network_json</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.keras_network_cache_key"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.keras_network_cache_key" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a Keras JSON description of a neural network, return a key that
uniquely defines this network. Networks that share the same key should
have compatible weights matrices and give the same prediction outputs
when their weights are the same.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>network_json</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.get_config">
<code class="sig-name descname">get_config</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.get_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.get_config" title="Permalink to this definition">¶</a></dt>
<dd><p>serialize to a dict all attributes except model weights</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>dict</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.from_config">
<em class="property">classmethod </em><code class="sig-name descname">from_config</code><span class="sig-paren">(</span><em class="sig-param">config</em>, <em class="sig-param">weights=None</em>, <em class="sig-param">weights_loader=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.from_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.from_config" title="Permalink to this definition">¶</a></dt>
<dd><p>deserialize from a dict returned by get_config().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>config</strong><span class="classifier">dict</span></dt><dd></dd>
<dt><strong>weights</strong><span class="classifier">list of array, optional</span></dt><dd><p>Network weights to restore</p>
</dd>
<dt><strong>weights_loader</strong><span class="classifier">callable, optional</span></dt><dd><p>Function to call (no arguments) to load weights when needed</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Class1NeuralNetwork</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.load_weights">
<code class="sig-name descname">load_weights</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.load_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.load_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Load weights by evaluating self.network_weights_loader, if needed.</p>
<p>After calling this, self.network_weights_loader will be None and
self.network_weights will be the weights list, if available.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.get_weights">
<code class="sig-name descname">get_weights</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.get_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.get_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the network weights</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of numpy.array giving weights for each layer or None if there is no</dt><dd></dd>
<dt>network</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.peptides_to_network_input">
<code class="sig-name descname">peptides_to_network_input</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.peptides_to_network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.peptides_to_network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode peptides to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">EncodableSequences or list of string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.supported_peptide_lengths">
<em class="property">property </em><code class="sig-name descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(minimum, maximum) lengths of peptides supported, inclusive.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>(int, int) tuple</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.allele_encoding_to_network_input">
<code class="sig-name descname">allele_encoding_to_network_input</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele_encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.allele_encoding_to_network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.allele_encoding_to_network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode alleles to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>allele_encoding</strong><span class="classifier">AlleleEncoding</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>(numpy.array, numpy.array)</dt><dd></dd>
<dt>Indices and allele representations.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.data_dependent_weights_initialization">
<em class="property">static </em><code class="sig-name descname">data_dependent_weights_initialization</code><span class="sig-paren">(</span><em class="sig-param">network</em>, <em class="sig-param">x_dict=None</em>, <em class="sig-param">method='lsuv'</em>, <em class="sig-param">verbose=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.data_dependent_weights_initialization"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.data_dependent_weights_initialization" title="Permalink to this definition">¶</a></dt>
<dd><p>Data dependent weights initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>network</strong><span class="classifier">keras.Model</span></dt><dd></dd>
<dt><strong>x_dict</strong><span class="classifier">dict of string -&gt; numpy.ndarray</span></dt><dd><p>Training data as would be passed keras.Model.fit().</p>
</dd>
<dt><strong>method</strong><span class="classifier">string</span></dt><dd><p>Initialization method. Currently only “lsuv” is supported.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Status updates printed to stdout if verbose &gt; 0</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.fit_generator">
<code class="sig-name descname">fit_generator</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">generator</em>, <em class="sig-param">validation_peptide_encoding</em>, <em class="sig-param">validation_affinities</em>, <em class="sig-param">validation_allele_encoding=None</em>, <em class="sig-param">validation_inequalities=None</em>, <em class="sig-param">validation_output_indices=None</em>, <em class="sig-param">steps_per_epoch=10</em>, <em class="sig-param">epochs=1000</em>, <em class="sig-param">min_epochs=0</em>, <em class="sig-param">patience=10</em>, <em class="sig-param">min_delta=0.0</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">progress_callback=None</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.fit_generator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.fit_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit using a generator. Does not support many of the features of fit(),
such as random negative peptides.</p>
<p>Fitting proceeds until early stopping is hit, using the peptides,
affinities, etc. given by the parameters starting with “<a href="#id8"><span class="problematic" id="id9">validation_</span></a>”.</p>
<p>This is used for pre-training pan-allele models using data synthesized
by the allele-specific models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>generator</strong><span class="classifier">generator yielding (alleles, peptides, affinities) tuples</span></dt><dd><p>where alleles and peptides are lists of strings, and affinities
is list of floats.</p>
</dd>
<dt><strong>validation_peptide_encoding</strong><span class="classifier">EncodableSequences</span></dt><dd></dd>
<dt><strong>validation_affinities</strong><span class="classifier">list of float</span></dt><dd></dd>
<dt><strong>validation_allele_encoding</strong><span class="classifier">AlleleEncoding</span></dt><dd></dd>
<dt><strong>validation_inequalities</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>validation_output_indices</strong><span class="classifier">list of int</span></dt><dd></dd>
<dt><strong>steps_per_epoch</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>epochs</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>min_epochs</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>patience</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>min_delta</strong><span class="classifier">float</span></dt><dd></dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>progress_callback</strong><span class="classifier">thunk</span></dt><dd></dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">affinities</em>, <em class="sig-param">allele_encoding=None</em>, <em class="sig-param">inequalities=None</em>, <em class="sig-param">output_indices=None</em>, <em class="sig-param">sample_weights=None</em>, <em class="sig-param">shuffle_permutation=None</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">progress_callback=None</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">EncodableSequences or list of string</span></dt><dd></dd>
<dt><strong>affinities</strong><span class="classifier">list of float</span></dt><dd><p>nM affinities. Must be same length of as peptides.</p>
</dd>
<dt><strong>allele_encoding</strong><span class="classifier">AlleleEncoding</span></dt><dd><p>If not specified, the model will be a single-allele predictor.</p>
</dd>
<dt><strong>inequalities</strong><span class="classifier">list of string, each element one of “&gt;”, “&lt;”, or “=”.</span></dt><dd><p>Inequalities to use for fitting. Same length as affinities.
Each element must be one of “&gt;”, “&lt;”, or “=”. For example, a “&gt;”
will train on y_pred &gt; y_true for that element in the training set.
Requires using a custom losses that support inequalities (e.g.
mse_with_ineqalities). If None all inequalities are taken to be “=”.</p>
</dd>
<dt><strong>output_indices</strong><span class="classifier">list of int</span></dt><dd><p>For multi-output models only. Same length as affinities. Indicates
the index of the output (starting from 0) for each training example.</p>
</dd>
<dt><strong>sample_weights</strong><span class="classifier">list of float</span></dt><dd><p>If not specified, all samples (including random negatives added
during training) will have equal weight. If specified, the random
negatives will be assigned weight=1.0.</p>
</dd>
<dt><strong>shuffle_permutation</strong><span class="classifier">list of int</span></dt><dd><p>Permutation (integer list) of same length as peptides and affinities
If None, then a random permutation will be generated.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Keras verbosity level</p>
</dd>
<dt><strong>progress_callback</strong><span class="classifier">function</span></dt><dd><p>No-argument function to call after each epoch.</p>
</dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd><p>Optional string of information to include in each progress update</p>
</dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd><p>How often (in seconds) to print progress update. Set to None to
disable.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">allele_encoding=None</em>, <em class="sig-param">batch_size=4096</em>, <em class="sig-param">output_index=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict affinities.</p>
<p>If peptides are specified as EncodableSequences, then the predictions
will be cached for this predictor as long as the EncodableSequences
object remains in memory. The cache is keyed in the object identity of
the EncodableSequences, not the sequences themselves. The cache is used
only for allele-specific models (i.e. when allele_encoding is None).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">EncodableSequences or list of string</span></dt><dd></dd>
<dt><strong>allele_encoding</strong><span class="classifier">AlleleEncoding, optional</span></dt><dd><p>Only required when this model is a pan-allele model</p>
</dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd><p>batch_size passed to Keras</p>
</dd>
<dt><strong>output_index</strong><span class="classifier">int or None</span></dt><dd><p>For multi-output models. Gives the output index to return. If set to
None, then all outputs are returned as a samples x outputs matrix.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of nM affinity predictions</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">models</em>, <em class="sig-param">merge_method='average'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.merge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge multiple models at the tensorflow (or other backend) level.</p>
<p>Only certain neural network architectures support merging. Others will
result in a NotImplementedError.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models</strong><span class="classifier">list of Class1NeuralNetwork</span></dt><dd><p>instances to merge</p>
</dd>
<dt><strong>merge_method</strong><span class="classifier">string, one of “average”, “sum”, or “concatenate”</span></dt><dd><p>How to merge the predictions of the different models</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Class1NeuralNetwork</dt><dd><p>The merged neural network</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.make_network">
<code class="sig-name descname">make_network</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptide_encoding</em>, <em class="sig-param">allele_amino_acid_encoding</em>, <em class="sig-param">allele_dense_layer_sizes</em>, <em class="sig-param">peptide_dense_layer_sizes</em>, <em class="sig-param">peptide_allele_merge_method</em>, <em class="sig-param">peptide_allele_merge_activation</em>, <em class="sig-param">layer_sizes</em>, <em class="sig-param">dense_layer_l1_regularization</em>, <em class="sig-param">dense_layer_l2_regularization</em>, <em class="sig-param">activation</em>, <em class="sig-param">init</em>, <em class="sig-param">output_activation</em>, <em class="sig-param">dropout_probability</em>, <em class="sig-param">batch_normalization</em>, <em class="sig-param">locally_connected_layers</em>, <em class="sig-param">topology</em>, <em class="sig-param">num_outputs=1</em>, <em class="sig-param">allele_representations=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.make_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.make_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to make a keras network for class 1 affinity prediction.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.clear_allele_representations">
<code class="sig-name descname">clear_allele_representations</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.clear_allele_representations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.clear_allele_representations" title="Permalink to this definition">¶</a></dt>
<dd><p>Set allele representations to an empty array. Useful before saving to
save a smaller version of the model.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_neural_network.Class1NeuralNetwork.set_allele_representations">
<code class="sig-name descname">set_allele_representations</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele_representations</em>, <em class="sig-param">force_surgery=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_neural_network.html#Class1NeuralNetwork.set_allele_representations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_neural_network.Class1NeuralNetwork.set_allele_representations" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the allele representations in use by this model. This means mutating
the weights for the allele input embedding layer.</p>
<p>Rationale: instead of passing in the allele sequence for each data point
during model training or prediction (which is expensive in terms of
memory usage), we pass in an allele index between 0 and n-1 where n is
the number of alleles in some universe of possible alleles. This index
is used in the model to lookup the corresponding allele sequence. This
function sets the lookup table.</p>
<p>See also: AlleleEncoding.allele_representations()</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>allele_representations</strong><span class="classifier">numpy.ndarray of shape (a, l, m)</span></dt><dd><dl class="simple">
<dt>where a is the total number of alleles,</dt><dd><p>l is the allele sequence length,
m is the length of the vectors used to represent amino acids</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.class1_presentation_predictor">
<span id="mhcflurry-class1-presentation-predictor-module"></span><h2>mhcflurry.class1_presentation_predictor module<a class="headerlink" href="#module-mhcflurry.class1_presentation_predictor" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.class1_presentation_predictor.</code><code class="sig-name descname">Class1PresentationPredictor</code><span class="sig-paren">(</span><em class="sig-param">affinity_predictor=None</em>, <em class="sig-param">processing_predictor_with_flanks=None</em>, <em class="sig-param">processing_predictor_without_flanks=None</em>, <em class="sig-param">weights_dataframe=None</em>, <em class="sig-param">metadata_dataframes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A logistic regression model over predicted binding affinity (BA) and antigen
processing (AP) score.</p>
<p>Instances of this class delegate to Class1AffinityPredictor and
Class1ProcessingPredictor instances to generate BA and AP predictions.
These predictions are combined using a logistic regression model to give
a “presentation score” prediction.</p>
<p>Most users will call the <a class="reference internal" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.load" title="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load</span></code></a> static method to get an instance of this
class, then call the <a class="reference internal" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict" title="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method to generate predictions.</p>
<dl class="attribute">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.model_inputs">
<code class="sig-name descname">model_inputs</code><em class="property"> = ['affinity_score', 'processing_score']</em><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.model_inputs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.supported_alleles">
<em class="property">property </em><code class="sig-name descname">supported_alleles</code><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.supported_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>List of alleles supported by the underlying Class1AffinityPredictor</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.supported_peptide_lengths">
<em class="property">property </em><code class="sig-name descname">supported_peptide_lengths</code><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.supported_peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>(min, max) of supported peptide lengths, inclusive.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict_affinity">
<code class="sig-name descname">predict_affinity</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">alleles</em>, <em class="sig-param">sample_names=None</em>, <em class="sig-param">include_affinity_percentile=True</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.predict_affinity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict_affinity" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict binding affinities across samples (each corresponding to up to
six MHC  I alleles).</p>
<p>Two modes are supported: each peptide can be evaluated for binding to
any of the alleles in any sample (this is what happens when sample_names
is None), or the i’th peptide can be evaluated for binding the alleles
of the sample given by the i’th entry in sample_names.</p>
<p>For example, if we don’t specify sample_names, then predictions
are taken for all combinations of samples and peptides, for a result
size of num peptides *  num samples:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1PresentationPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_affinity</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;PEPTIDE&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">    peptide  peptide_num sample_name      affinity best_allele</span>
<span class="go">0  SIINFEKL            0     sample1  12906.787792       A0201</span>
<span class="go">1   PEPTIDE            1     sample1  36827.681130       B0702</span>
<span class="go">2  SIINFEKL            0     sample2   3588.413748       C0202</span>
<span class="go">3   PEPTIDE            1     sample2  34362.109211       C0202</span>
</pre></div>
</div>
<p>In contrast, here we specify sample_names, so peptide is evaluated for
binding the alleles in the corresponding sample, for a result size equal
to the number of peptides:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_affinity</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;PEPTIDE&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">sample_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sample2&quot;</span><span class="p">,</span> <span class="s2">&quot;sample1&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">    peptide  peptide_num sample_name      affinity best_allele</span>
<span class="go">0  SIINFEKL            0     sample2   3588.412141       C0202</span>
<span class="go">1   PEPTIDE            1     sample1  36827.682779       B0702</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">dict of string -&gt; list of string</span></dt><dd><p>Keys are sample names, values are the alleles (genotype) for
that sample</p>
</dd>
<dt><strong>sample_names</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>Sample names corresponding to each peptide. If None, then
predictions are generated for all sample genotypes  across all
peptides.</p>
</dd>
<dt><strong>include_affinity_percentile</strong><span class="classifier">bool</span></dt><dd><p>Whether to include affinity percentile ranks</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Set to 0 for quiet.</p>
</dd>
<dt><strong>throw</strong><span class="classifier">verbose</span></dt><dd><p>Whether to throw exception (vs. just log a warning) on invalid
peptides, etc.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>pandas.DataFrame</strong><span class="classifier">predictions</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict_processing">
<code class="sig-name descname">predict_processing</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">verbose=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.predict_processing"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict_processing" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing scores for individual peptides, optionally
including flanking sequences for better cleavage prediction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd></dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd></dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>numpy.array</strong><span class="classifier">Antigen processing scores for each peptide</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">targets</em>, <em class="sig-param">peptides</em>, <em class="sig-param">sample_names</em>, <em class="sig-param">alleles</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">verbose=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the presentation score logistic regression model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>targets</strong><span class="classifier">list of int/float</span></dt><dd><p>1 indicates hit, 0 indicates decoy</p>
</dd>
<dt><strong>peptides</strong><span class="classifier">list of string [same length as targets]</span></dt><dd></dd>
<dt><strong>sample_names</strong><span class="classifier">list of string [same length as targets]</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">dict of string -&gt; list of string</span></dt><dd><p>Keys are sample names, values are the alleles for that sample</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string [same length as targets]</span></dt><dd></dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string [same length as targets]</span></dt><dd></dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.get_model">
<code class="sig-name descname">get_model</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.get_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.get_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Load or instantiate a new logistic regression model. Private helper
method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name</strong><span class="classifier">string</span></dt><dd><p>If None (the default), an un-fit LR model is returned. Otherwise the
weights are loaded for the specified model.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>sklearn.linear_model.LogisticRegression</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">alleles</em>, <em class="sig-param">sample_names=None</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">include_affinity_percentile=False</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict presentation scores across a set of peptides.</p>
<p>Presentation scores combine predictions for MHC I binding affinity
and antigen processing.</p>
<p>This method returns a pandas.DataFrame giving presentation scores plus
the binding affinity and processing predictions and other intermediate
results.</p>
<p>Example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1PresentationPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">peptides</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SIINFEKL&quot;</span><span class="p">,</span> <span class="s2">&quot;PEPTIDE&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">n_flanks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;NNN&quot;</span><span class="p">,</span> <span class="s2">&quot;SNS&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">c_flanks</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CCC&quot;</span><span class="p">,</span> <span class="s2">&quot;CNC&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">    peptide n_flank c_flank  peptide_num sample_name      affinity best_allele  processing_score  presentation_score</span>
<span class="go">0  SIINFEKL     NNN     CCC            0     sample1  12906.787792       A0201          0.802466            0.140365</span>
<span class="go">1   PEPTIDE     SNS     CNC            1     sample1  36827.681130       B0702          0.105260            0.004059</span>
<span class="go">2  SIINFEKL     NNN     CCC            0     sample2   3588.413748       C0202          0.802466            0.338647</span>
<span class="go">3   PEPTIDE     SNS     CNC            1     sample2  34362.109211       C0202          0.105260            0.004317</span>
</pre></div>
</div>
<p>You can also specify sample_names, in which case peptide is evaluated
for binding the alleles in the corresponding sample only. See
<a class="reference internal" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict_affinity" title="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict_affinity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict_affinity</span></code></a> for an examples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">list of string or dict of string -&gt; list of string</span></dt><dd><p>If you are predicting for a single sample, pass a list of strings
(up to 6) indicating the genotype. If you are predicting across
multiple samples, pass a dict where the keys are (arbitrary)
sample names and the values are the alleles to predict for that
sample.</p>
</dd>
<dt><strong>sample_names</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>If you are passing a dict for ‘alleles’, you can use this
argument to specify which peptides go with which samples. If it is
None, then predictions will be performed for each peptide across all
samples.</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>Upstream sequences before the peptide. Sequences of any length can
be given and a suffix of the size supported by the model will be
used.</p>
</dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>Downstream sequences after the peptide. Sequences of any length can
be given and a prefix of the size supported by the model will be
used.</p>
</dd>
<dt><strong>include_affinity_percentile</strong><span class="classifier">bool</span></dt><dd><p>Whether to include affinity percentile ranks</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Set to 0 for quiet.</p>
</dd>
<dt><strong>throw</strong><span class="classifier">verbose</span></dt><dd><p>Whether to throw exception (vs. just log a warning) on invalid
peptides, etc.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
<dt>Presentation scores and intermediate results.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict_sequences">
<code class="sig-name descname">predict_sequences</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em>, <em class="sig-param">alleles</em>, <em class="sig-param">result='best'</em>, <em class="sig-param">comparison_quantity='presentation_score'</em>, <em class="sig-param">filter_value=None</em>, <em class="sig-param">peptide_lengths=(8</em>, <em class="sig-param">9</em>, <em class="sig-param">10</em>, <em class="sig-param">11)</em>, <em class="sig-param">use_flanks=True</em>, <em class="sig-param">include_affinity_percentile=True</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">throw=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.predict_sequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.predict_sequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict presentation across protein sequences.</p>
<p>Example:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span> <span class="o">=</span> <span class="n">Class1PresentationPredictor</span><span class="o">.</span><span class="n">load</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">predictor</span><span class="o">.</span><span class="n">predict_sequences</span><span class="p">(</span>
<span class="gp">... </span>   <span class="n">sequences</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s1">&#39;protein1&#39;</span><span class="p">:</span> <span class="s2">&quot;MDSKGSSQKGSRLLLLLVVSNLL&quot;</span><span class="p">,</span>
<span class="gp">... </span>       <span class="s1">&#39;protein2&#39;</span><span class="p">:</span> <span class="s2">&quot;SSLPTPEDKEQAQQTHH&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">alleles</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>       <span class="s2">&quot;sample1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0201&quot;</span><span class="p">,</span> <span class="s2">&quot;A0301&quot;</span><span class="p">,</span> <span class="s2">&quot;B0702&quot;</span><span class="p">],</span>
<span class="gp">... </span>       <span class="s2">&quot;sample2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A0101&quot;</span><span class="p">,</span> <span class="s2">&quot;C0202&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="p">},</span>
<span class="gp">... </span>   <span class="n">result</span><span class="o">=</span><span class="s2">&quot;filtered&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">comparison_quantity</span><span class="o">=</span><span class="s2">&quot;affinity&quot;</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">filter_value</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span>
<span class="gp">... </span>   <span class="n">verbose</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">  sequence_name  pos     peptide         n_flank     c_flank sample_name    affinity best_allele  affinity_percentile  processing_score  presentation_score</span>
<span class="go">0      protein1   13   LLLLVVSNL   MDSKGSSQKGSRL           L     sample1   38.206225       A0201             0.380125          0.017644            0.571060</span>
<span class="go">1      protein1   14   LLLVVSNLL  MDSKGSSQKGSRLL                 sample1   42.243472       A0201             0.420250          0.090984            0.619213</span>
<span class="go">2      protein1    5   SSQKGSRLL           MDSKG   LLLVVSNLL     sample2   66.749223       C0202             0.803375          0.383608            0.774468</span>
<span class="go">3      protein1    6   SQKGSRLLL          MDSKGS    LLVVSNLL     sample2  178.033474       C0202             1.820000          0.275019            0.482206</span>
<span class="go">4      protein1   13  LLLLVVSNLL   MDSKGSSQKGSRL                 sample1  202.208167       A0201             1.112500          0.058782            0.261320</span>
<span class="go">5      protein1   12  LLLLLVVSNL    MDSKGSSQKGSR           L     sample1  202.506582       A0201             1.112500          0.010025            0.225648</span>
<span class="go">6      protein2    0   SSLPTPEDK                    EQAQQTHH     sample1  335.529377       A0301             1.011750          0.010443            0.156798</span>
<span class="go">7      protein2    0   SSLPTPEDK                    EQAQQTHH     sample2  353.451759       C0202             2.674250          0.010443            0.150753</span>
<span class="go">8      protein1    8   KGSRLLLLL        MDSKGSSQ      VVSNLL     sample2  410.327286       C0202             2.887000          0.121374            0.194081</span>
<span class="go">9      protein1    5    SSQKGSRL           MDSKG  LLLLVVSNLL     sample2  477.285954       C0202             3.107375          0.111982            0.168572</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">str, list of string, or string -&gt; string dict</span></dt><dd><p>Protein sequences. If a dict is given, the keys are arbitrary (
e.g. protein names), and the values are the amino acid sequences.</p>
</dd>
<dt><strong>alleles</strong><span class="classifier">list of string, list of list of string, or dict of string -&gt; list of string</span></dt><dd><p>MHC I alleles. Can be: (1) a string (a single allele), (2) a list of
strings (a single genotype), (3) a list of list of strings
(multiple genotypes, where the total number of genotypes must equal
the number of sequences), or (4) a dict giving multiple genotypes,
which will each be run over the sequences.</p>
</dd>
<dt><strong>result</strong><span class="classifier">string</span></dt><dd><p>Specify ‘best’ to return the strongest peptide for each sequence,
‘all’ to return predictions for all peptides, or ‘filtered’ to
return predictions where the comparison_quantity is stronger
(i.e (&lt;) for affinity, (&gt;) for scores) than filter_value.</p>
</dd>
<dt><strong>comparison_quantity</strong><span class="classifier">string</span></dt><dd><p>One of “presentation_score”, “processing_score”, “affinity”, or
“affinity_percentile”. Prediction to use to rank (if result is
“best”) or filter (if result is “filtered”) results.</p>
</dd>
<dt><strong>filter_value</strong><span class="classifier">float</span></dt><dd><p>Threshold value to use, only relevant when result is “filtered”.
If comparison_quantity is “affinity”, then all results less than
(i.e. tighter than) the specified nM affinity are retained. If it’s
“presentation_score” or “processing_score” then results greater than
the indicated filter_value are retained.</p>
</dd>
<dt><strong>peptide_lengths</strong><span class="classifier">list of int</span></dt><dd><p>Peptide lengths to predict for.</p>
</dd>
<dt><strong>use_flanks</strong><span class="classifier">bool</span></dt><dd><p>Whether to include flanking sequences when running the AP predictor
(for better cleavage prediction).</p>
</dd>
<dt><strong>include_affinity_percentile</strong><span class="classifier">bool</span></dt><dd><p>Whether to include affinity percentile ranks in output.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Set to 0 for quiet mode.</p>
</dd>
<dt><strong>throw</strong><span class="classifier">boolean</span></dt><dd><p>Whether to throw exceptions (vs. log warnings) on invalid inputs.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame with columns:</dt><dd><p>peptide, n_flank, c_flank, sequence_name, affinity, best_allele,
processing_score, presentation_score</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.save">
<code class="sig-name descname">save</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">models_dir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Save the predictor to a directory on disk. If the directory does
not exist it will be created.</p>
<p>The wrapped Class1AffinityPredictor and Class1ProcessingPredictor
instances are included in the saved data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. It will be created if it doesn’t exist.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.load">
<em class="property">classmethod </em><code class="sig-name descname">load</code><span class="sig-paren">(</span><em class="sig-param">models_dir=None</em>, <em class="sig-param">max_models=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_presentation_predictor.html#Class1PresentationPredictor.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserialize a predictor from a directory on disk.</p>
<p>This will also load the wrapped Class1AffinityPredictor and
Class1ProcessingPredictor instances.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. If unspecified the default downloaded models are
used.</p>
</dd>
<dt><strong>max_models</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum number of affinity and processing (counted separately)
models to load</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#mhcflurry.class1_presentation_predictor.Class1PresentationPredictor" title="mhcflurry.class1_presentation_predictor.Class1PresentationPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1PresentationPredictor</span></code></a> instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.class1_processing_neural_network">
<span id="mhcflurry-class1-processing-neural-network-module"></span><h2>mhcflurry.class1_processing_neural_network module<a class="headerlink" href="#module-mhcflurry.class1_processing_neural_network" title="Permalink to this headline">¶</a></h2>
<p>Antigen processing neural network implementation</p>
<dl class="class">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.class1_processing_neural_network.</code><code class="sig-name descname">Class1ProcessingNeuralNetwork</code><span class="sig-paren">(</span><em class="sig-param">**hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A neural network for antigen processing prediction</p>
<dl class="attribute">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.network_hyperparameter_defaults">
<code class="sig-name descname">network_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.network_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters (and their default values) that affect the neural network
architecture.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.fit_hyperparameter_defaults">
<code class="sig-name descname">fit_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.fit_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for neural network training.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.early_stopping_hyperparameter_defaults">
<code class="sig-name descname">early_stopping_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.early_stopping_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperparameters for early stopping.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.compile_hyperparameter_defaults">
<code class="sig-name descname">compile_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.compile_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Loss and optimizer hyperparameters. Any values supported by keras may be
used.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.auxiliary_input_hyperparameter_defaults">
<code class="sig-name descname">auxiliary_input_hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.auxiliary_input_hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Allele feature hyperparameters.</p>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.hyperparameter_defaults">
<code class="sig-name descname">hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.sequence_lengths">
<em class="property">property </em><code class="sig-name descname">sequence_lengths</code><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.sequence_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Supported maximum sequence lengths</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>dict of string -&gt; int</dt><dd></dd>
<dt>Keys are “peptide”, “n_flank”, “c_flank”. Values give the maximum</dt><dd></dd>
<dt>supported sequence length.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.network">
<code class="sig-name descname">network</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.network" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the keras model associated with this network.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.update_network_description">
<code class="sig-name descname">update_network_description</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.update_network_description"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.update_network_description" title="Permalink to this definition">¶</a></dt>
<dd><p>Update self.network_json and self.network_weights properties based on
this instances’s neural network.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em>, <em class="sig-param">targets</em>, <em class="sig-param">sample_weights=None</em>, <em class="sig-param">shuffle_permutation=None</em>, <em class="sig-param">verbose=1</em>, <em class="sig-param">progress_callback=None</em>, <em class="sig-param">progress_preamble=''</em>, <em class="sig-param">progress_print_interval=5.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the neural network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">FlankingEncoding</span></dt><dd><p>Peptides and upstream/downstream flanking sequences</p>
</dd>
<dt><strong>targets</strong><span class="classifier">list of float</span></dt><dd><p>1 indicates hit, 0 indicates decoy</p>
</dd>
<dt><strong>sample_weights</strong><span class="classifier">list of float</span></dt><dd><p>If not specified all samples have equal weight.</p>
</dd>
<dt><strong>shuffle_permutation</strong><span class="classifier">list of int</span></dt><dd><p>Permutation (integer list) of same length as peptides and affinities
If None, then a random permutation will be generated.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">int</span></dt><dd><p>Keras verbosity level</p>
</dd>
<dt><strong>progress_callback</strong><span class="classifier">function</span></dt><dd><p>No-argument function to call after each epoch.</p>
</dd>
<dt><strong>progress_preamble</strong><span class="classifier">string</span></dt><dd><p>Optional string of information to include in each progress update</p>
</dd>
<dt><strong>progress_print_interval</strong><span class="classifier">float</span></dt><dd><p>How often (in seconds) to print progress update. Set to None to
disable.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string</span></dt><dd><p>Upstream sequence before each peptide</p>
</dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string</span></dt><dd><p>Downstream sequence after each peptide</p>
</dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd><p>Prediction keras batch size.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
<dt>Processing scores. Range is 0-1, higher indicates more favorable</dt><dd></dd>
<dt>processing.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.predict_encoded">
<code class="sig-name descname">predict_encoded</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.predict_encoded"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.predict_encoded" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">FlankingEncoding</span></dt><dd><p>Peptides and flanking sequences</p>
</dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd><p>Prediction keras batch size.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.network_input">
<code class="sig-name descname">network_input</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.network_input"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.network_input" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode peptides to the fixed-length encoding expected by the neural
network (which depends on the architecture).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">FlankingEncoding</span></dt><dd><p>Peptides and flanking sequences</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.make_network">
<code class="sig-name descname">make_network</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">amino_acid_encoding</em>, <em class="sig-param">peptide_max_length</em>, <em class="sig-param">n_flank_length</em>, <em class="sig-param">c_flank_length</em>, <em class="sig-param">flanking_averages</em>, <em class="sig-param">convolutional_filters</em>, <em class="sig-param">convolutional_kernel_size</em>, <em class="sig-param">convolutional_activation</em>, <em class="sig-param">convolutional_kernel_l1_l2</em>, <em class="sig-param">dropout_rate</em>, <em class="sig-param">post_convolutional_dense_layer_sizes</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.make_network"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.make_network" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to make a keras network given hyperparameters.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.get_weights">
<code class="sig-name descname">get_weights</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.get_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.get_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the network weights</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of numpy.array giving weights for each layer or None if there is no</dt><dd></dd>
<dt>network</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.get_config">
<code class="sig-name descname">get_config</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.get_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.get_config" title="Permalink to this definition">¶</a></dt>
<dd><p>serialize to a dict all attributes except model weights</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>dict</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.from_config">
<em class="property">classmethod </em><code class="sig-name descname">from_config</code><span class="sig-paren">(</span><em class="sig-param">config</em>, <em class="sig-param">weights=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_neural_network.html#Class1ProcessingNeuralNetwork.from_config"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_neural_network.Class1ProcessingNeuralNetwork.from_config" title="Permalink to this definition">¶</a></dt>
<dd><p>deserialize from a dict returned by get_config().</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>config</strong><span class="classifier">dict</span></dt><dd></dd>
<dt><strong>weights</strong><span class="classifier">list of array, optional</span></dt><dd><p>Network weights to restore</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Class1ProcessingNeuralNetwork</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.class1_processing_predictor">
<span id="mhcflurry-class1-processing-predictor-module"></span><h2>mhcflurry.class1_processing_predictor module<a class="headerlink" href="#module-mhcflurry.class1_processing_predictor" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.class1_processing_predictor.</code><code class="sig-name descname">Class1ProcessingPredictor</code><span class="sig-paren">(</span><em class="sig-param">models</em>, <em class="sig-param">manifest_df=None</em>, <em class="sig-param">metadata_dataframes=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>User-facing interface to antigen processing prediction.</p>
<p>Delegates to an ensemble of Class1ProcessingNeuralNetwork instances.</p>
<p>Instantiate a new Class1ProcessingPredictor</p>
<p>Users will generally call load() to restore a saved predictor rather
than using this constructor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models</strong><span class="classifier">list of Class1ProcessingNeuralNetwork</span></dt><dd><p>Neural networks in the ensemble.</p>
</dd>
<dt><strong>manifest_df</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>Manifest dataframe. If not specified a new one will be created when
needed.</p>
</dd>
<dt><strong>metadata_dataframes</strong><span class="classifier">dict of string -&gt; pandas.DataFrame</span></dt><dd><p>Arbitrary metadata associated with this predictor</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.sequence_lengths">
<em class="property">property </em><code class="sig-name descname">sequence_lengths</code><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.sequence_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Supported maximum sequence lengths.</p>
<p>Passing a peptide greater than the maximum supported length results
in an error.</p>
<p>Passing an N- or C-flank sequence greater than the maximum supported
length results in some part of it being ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>dict of string -&gt; int</dt><dd></dd>
<dt>Keys are “peptide”, “n_flank”, “c_flank”. Values give the maximum</dt><dd></dd>
<dt>supported sequence length.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.add_models">
<code class="sig-name descname">add_models</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">models</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.add_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.add_models" title="Permalink to this definition">¶</a></dt>
<dd><p>Add models to the ensemble (in-place).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models</strong><span class="classifier">list of Class1ProcessingNeuralNetwork</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of string</dt><dd></dd>
<dt>Names of the new models.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.manifest_df">
<em class="property">property </em><code class="sig-name descname">manifest_df</code><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.manifest_df" title="Permalink to this definition">¶</a></dt>
<dd><p>A pandas.DataFrame describing the models included in this predictor.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.model_name">
<em class="property">static </em><code class="sig-name descname">model_name</code><span class="sig-paren">(</span><em class="sig-param">num</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.model_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.model_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a model name</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.weights_path">
<em class="property">static </em><code class="sig-name descname">weights_path</code><span class="sig-paren">(</span><em class="sig-param">models_dir</em>, <em class="sig-param">model_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.weights_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.weights_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate the path to the weights file for a model</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>model_name</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string</span></dt><dd><p>Upstream sequence before each peptide</p>
</dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string</span></dt><dd><p>Downstream sequence after each peptide</p>
</dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd><p>Prediction keras batch size.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
<dt>Processing scores. Range is 0-1, higher indicates more favorable</dt><dd></dd>
<dt>processing.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict_to_dataframe">
<code class="sig-name descname">predict_to_dataframe</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">n_flanks=None</em>, <em class="sig-param">c_flanks=None</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.predict_to_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict_to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<p>See <a class="reference internal" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict" title="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method for parameter descriptions.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
<dt>Processing predictions are in the “score” column. Also includes</dt><dd></dd>
<dt>peptides and flanking sequences.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict_to_dataframe_encoded">
<code class="sig-name descname">predict_to_dataframe_encoded</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">sequences</em>, <em class="sig-param">batch_size=4096</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.predict_to_dataframe_encoded"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict_to_dataframe_encoded" title="Permalink to this definition">¶</a></dt>
<dd><p>Predict antigen processing.</p>
<p>See <a class="reference internal" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict" title="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.predict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">predict</span></code></a> method for more information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">FlankingEncoding</span></dt><dd></dd>
<dt><strong>batch_size</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.check_consistency">
<code class="sig-name descname">check_consistency</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.check_consistency"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.check_consistency" title="Permalink to this definition">¶</a></dt>
<dd><p>Verify that self.manifest_df is consistent with instance variables.</p>
<p>Currently only checks for agreement on the total number of models.</p>
<p>Throws AssertionError if inconsistent.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.save">
<code class="sig-name descname">save</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">models_dir</em>, <em class="sig-param">model_names_to_write=None</em>, <em class="sig-param">write_metadata=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the predictor to a directory on disk. If the directory does
not exist it will be created.</p>
<p>The serialization format consists of a file called “manifest.csv” with
the configurations of each Class1ProcessingNeuralNetwork, along with
per-network files giving the model weights.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. It will be created if it doesn’t exist.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.load">
<em class="property">classmethod </em><code class="sig-name descname">load</code><span class="sig-paren">(</span><em class="sig-param">models_dir=None</em>, <em class="sig-param">max_models=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/class1_processing_predictor.html#Class1ProcessingPredictor.load"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Deserialize a predictor from a directory on disk.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>models_dir</strong><span class="classifier">string</span></dt><dd><p>Path to directory. If unspecified the default downloaded models are
used.</p>
</dd>
<dt><strong>max_models</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum number of models to load</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#mhcflurry.class1_processing_predictor.Class1ProcessingPredictor" title="mhcflurry.class1_processing_predictor.Class1ProcessingPredictor"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Class1ProcessingPredictor</span></code></a> instance</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.cluster_parallelism">
<span id="mhcflurry-cluster-parallelism-module"></span><h2>mhcflurry.cluster_parallelism module<a class="headerlink" href="#module-mhcflurry.cluster_parallelism" title="Permalink to this headline">¶</a></h2>
<p>Simple, relatively naive parallel map implementation for HPC clusters.</p>
<p>Used for training MHCflurry models.</p>
<dl class="function">
<dt id="mhcflurry.cluster_parallelism.add_cluster_parallelism_args">
<code class="sig-prename descclassname">mhcflurry.cluster_parallelism.</code><code class="sig-name descname">add_cluster_parallelism_args</code><span class="sig-paren">(</span><em class="sig-param">parser</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/cluster_parallelism.html#add_cluster_parallelism_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.cluster_parallelism.add_cluster_parallelism_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Add commandline arguments controlling cluster parallelism to an argparse
ArgumentParser.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>parser</strong><span class="classifier">argparse.ArgumentParser</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.cluster_parallelism.cluster_results_from_args">
<code class="sig-prename descclassname">mhcflurry.cluster_parallelism.</code><code class="sig-name descname">cluster_results_from_args</code><span class="sig-paren">(</span><em class="sig-param">args</em>, <em class="sig-param">work_function</em>, <em class="sig-param">work_items</em>, <em class="sig-param">constant_data=None</em>, <em class="sig-param">input_serialization_method='pickle'</em>, <em class="sig-param">result_serialization_method='pickle'</em>, <em class="sig-param">clear_constant_data=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/cluster_parallelism.html#cluster_results_from_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.cluster_parallelism.cluster_results_from_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Parallel map configurable using commandline arguments. See the
cluster_results() function for docs.</p>
<p>The <code class="xref py py-obj docutils literal notranslate"><span class="pre">args</span></code> parameter should be an argparse.Namespace from an argparse parser
generated using the add_cluster_parallelism_args() function.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>args</strong></dt><dd></dd>
<dt><strong>work_function</strong></dt><dd></dd>
<dt><strong>work_items</strong></dt><dd></dd>
<dt><strong>constant_data</strong></dt><dd></dd>
<dt><strong>result_serialization_method</strong></dt><dd></dd>
<dt><strong>clear_constant_data</strong></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>generator</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.cluster_parallelism.cluster_results">
<code class="sig-prename descclassname">mhcflurry.cluster_parallelism.</code><code class="sig-name descname">cluster_results</code><span class="sig-paren">(</span><em class="sig-param">work_function</em>, <em class="sig-param">work_items</em>, <em class="sig-param">constant_data=None</em>, <em class="sig-param">submit_command='sh'</em>, <em class="sig-param">results_workdir='./cluster-workdir'</em>, <em class="sig-param">additional_complete_file=None</em>, <em class="sig-param">script_prefix_path=None</em>, <em class="sig-param">input_serialization_method='pickle'</em>, <em class="sig-param">result_serialization_method='pickle'</em>, <em class="sig-param">max_retries=3</em>, <em class="sig-param">clear_constant_data=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/cluster_parallelism.html#cluster_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.cluster_parallelism.cluster_results" title="Permalink to this definition">¶</a></dt>
<dd><p>Parallel map on an HPC cluster.</p>
<p>Returns [work_function(item) for item in work_items] where each invocation
of work_function is performed as a separate HPC cluster job. Order is
preserved.</p>
<p>Optionally, “constant data” can be specified, which will be passed to
each work_function() invocation as a keyword argument called constant_data.
This data is serialized once and all workers read it from the same source,
which is more efficient than serializing it separately for each worker.</p>
<p>Each worker’s input is serialized to a shared NFS directory and the
submit_command is used to launch a job to process that input. The shared
filesystem is polled occasionally to watch for results, which are fed back
to the user.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>work_function</strong><span class="classifier">A -&gt; B</span></dt><dd></dd>
<dt><strong>work_items</strong><span class="classifier">list of A</span></dt><dd></dd>
<dt><strong>constant_data</strong><span class="classifier">object</span></dt><dd></dd>
<dt><strong>submit_command</strong><span class="classifier">string</span></dt><dd><p>For running on LSF, we use “bsub” here.</p>
</dd>
<dt><strong>results_workdir</strong><span class="classifier">string</span></dt><dd><p>Path to NFS shared directory where inputs and results can be written</p>
</dd>
<dt><strong>script_prefix_path</strong><span class="classifier">string</span></dt><dd><p>Path to script that will be invoked to run each worker. A line calling
the _mhcflurry-cluster-worker-entry-point command will be appended to
the contents of this file.</p>
</dd>
<dt><strong>result_serialization_method</strong><span class="classifier">string, one of “pickle” or “save_predictor”</span></dt><dd><p>The “save_predictor” works only when the return type of work_function
is Class1AffinityPredictor</p>
</dd>
<dt><strong>max_retries</strong><span class="classifier">int</span></dt><dd><p>How many times to attempt to re-launch a failed worker</p>
</dd>
<dt><strong>clear_constant_data</strong><span class="classifier">bool</span></dt><dd><p>If True, the constant data dict is cleared on the launching host after
it is serialized to disk.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>generator of B</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.cluster_parallelism.worker_entry_point">
<code class="sig-prename descclassname">mhcflurry.cluster_parallelism.</code><code class="sig-name descname">worker_entry_point</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/cluster_parallelism.html#worker_entry_point"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.cluster_parallelism.worker_entry_point" title="Permalink to this definition">¶</a></dt>
<dd><p>Entry point for the worker command.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>argv</strong><span class="classifier">list of string</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.common">
<span id="mhcflurry-common-module"></span><h2>mhcflurry.common module<a class="headerlink" href="#module-mhcflurry.common" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="mhcflurry.common.set_keras_backend">
<code class="sig-prename descclassname">mhcflurry.common.</code><code class="sig-name descname">set_keras_backend</code><span class="sig-paren">(</span><em class="sig-param">backend=None</em>, <em class="sig-param">gpu_device_nums=None</em>, <em class="sig-param">num_threads=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#set_keras_backend"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.set_keras_backend" title="Permalink to this definition">¶</a></dt>
<dd><p>Configure Keras backend to use GPU or CPU. Only tensorflow is supported.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>backend</strong><span class="classifier">string, optional</span></dt><dd><p>one of ‘tensorflow-default’, ‘tensorflow-cpu’, ‘tensorflow-gpu’</p>
</dd>
<dt><strong>gpu_device_nums</strong><span class="classifier">list of int, optional</span></dt><dd><p>GPU devices to potentially use</p>
</dd>
<dt><strong>num_threads</strong><span class="classifier">int, optional</span></dt><dd><p>Tensorflow threads to use</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.configure_logging">
<code class="sig-prename descclassname">mhcflurry.common.</code><code class="sig-name descname">configure_logging</code><span class="sig-paren">(</span><em class="sig-param">verbose=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#configure_logging"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.configure_logging" title="Permalink to this definition">¶</a></dt>
<dd><p>Configure logging module using defaults.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>verbose</strong><span class="classifier">boolean</span></dt><dd><p>If true, output will be at level DEBUG, otherwise, INFO.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.amino_acid_distribution">
<code class="sig-prename descclassname">mhcflurry.common.</code><code class="sig-name descname">amino_acid_distribution</code><span class="sig-paren">(</span><em class="sig-param">peptides</em>, <em class="sig-param">smoothing=0.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#amino_acid_distribution"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.amino_acid_distribution" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the fraction of each amino acid across a collection of peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>smoothing</strong><span class="classifier">float, optional</span></dt><dd><p>Small number (e.g. 0.01) to add to all amino acid fractions. The higher
the number the more uniform the distribution.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.Series indexed by amino acids</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.random_peptides">
<code class="sig-prename descclassname">mhcflurry.common.</code><code class="sig-name descname">random_peptides</code><span class="sig-paren">(</span><em class="sig-param">num</em>, <em class="sig-param">length=9</em>, <em class="sig-param">distribution=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#random_peptides"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.random_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate random peptides (kmers).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>num</strong><span class="classifier">int</span></dt><dd><p>Number of peptides to return</p>
</dd>
<dt><strong>length</strong><span class="classifier">int</span></dt><dd><p>Length of each peptide</p>
</dd>
<dt><strong>distribution</strong><span class="classifier">pandas.Series</span></dt><dd><p>Maps 1-letter amino acid abbreviations to
probabilities. If not specified a uniform
distribution is used.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.positional_frequency_matrix">
<code class="sig-prename descclassname">mhcflurry.common.</code><code class="sig-name descname">positional_frequency_matrix</code><span class="sig-paren">(</span><em class="sig-param">peptides</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#positional_frequency_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.positional_frequency_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a set of peptides, calculate a length x amino acids frequency matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>All of same length</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd><p>Index is position, columns are amino acids</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.save_weights">
<code class="sig-prename descclassname">mhcflurry.common.</code><code class="sig-name descname">save_weights</code><span class="sig-paren">(</span><em class="sig-param">weights_list</em>, <em class="sig-param">filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#save_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.save_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Save model weights to the given filename using numpy’s “.npz” format.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>weights_list</strong><span class="classifier">list of numpy array</span></dt><dd></dd>
<dt><strong>filename</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.common.load_weights">
<code class="sig-prename descclassname">mhcflurry.common.</code><code class="sig-name descname">load_weights</code><span class="sig-paren">(</span><em class="sig-param">filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#load_weights"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.load_weights" title="Permalink to this definition">¶</a></dt>
<dd><p>Restore model weights from the given filename, which should have been
created with <a class="reference internal" href="#mhcflurry.common.save_weights" title="mhcflurry.common.save_weights"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_weights</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>filename</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of array</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="mhcflurry.common.NumpyJSONEncoder">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.common.</code><code class="sig-name descname">NumpyJSONEncoder</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">skipkeys=False</em>, <em class="sig-param">ensure_ascii=True</em>, <em class="sig-param">check_circular=True</em>, <em class="sig-param">allow_nan=True</em>, <em class="sig-param">sort_keys=False</em>, <em class="sig-param">indent=None</em>, <em class="sig-param">separators=None</em>, <em class="sig-param">default=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#NumpyJSONEncoder"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.NumpyJSONEncoder" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">json.encoder.JSONEncoder</span></code></p>
<p>JSON encoder (used with json module) that can handle numpy arrays.</p>
<p>Constructor for JSONEncoder, with sensible defaults.</p>
<p>If skipkeys is false, then it is a TypeError to attempt
encoding of keys that are not str, int, float or None.  If
skipkeys is True, such items are simply skipped.</p>
<p>If ensure_ascii is true, the output is guaranteed to be str
objects with all incoming non-ASCII characters escaped.  If
ensure_ascii is false, the output can contain non-ASCII characters.</p>
<p>If check_circular is true, then lists, dicts, and custom encoded
objects will be checked for circular references during encoding to
prevent an infinite recursion (which would cause an OverflowError).
Otherwise, no such check takes place.</p>
<p>If allow_nan is true, then NaN, Infinity, and -Infinity will be
encoded as such.  This behavior is not JSON specification compliant,
but is consistent with most JavaScript based encoders and decoders.
Otherwise, it will be a ValueError to encode such floats.</p>
<p>If sort_keys is true, then the output of dictionaries will be
sorted by key; this is useful for regression tests to ensure
that JSON serializations can be compared on a day-to-day basis.</p>
<p>If indent is a non-negative integer, then JSON array
elements and object members will be pretty-printed with that
indent level.  An indent level of 0 will only insert newlines.
None is the most compact representation.</p>
<p>If specified, separators should be an (item_separator, key_separator)
tuple.  The default is (‘, ‘, ‘: ‘) if <em>indent</em> is <code class="docutils literal notranslate"><span class="pre">None</span></code> and
(‘,’, ‘: ‘) otherwise.  To get the most compact JSON representation,
you should specify (‘,’, ‘:’) to eliminate whitespace.</p>
<p>If specified, default is a function that gets called for objects
that can’t otherwise be serialized.  It should return a JSON encodable
version of the object or raise a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code>.</p>
<dl class="method">
<dt id="mhcflurry.common.NumpyJSONEncoder.default">
<code class="sig-name descname">default</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/common.html#NumpyJSONEncoder.default"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.common.NumpyJSONEncoder.default" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement this method in a subclass such that it returns
a serializable object for <code class="docutils literal notranslate"><span class="pre">o</span></code>, or calls the base implementation
(to raise a <code class="docutils literal notranslate"><span class="pre">TypeError</span></code>).</p>
<p>For example, to support arbitrary iterators, you could
implement default like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">default</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">o</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">iterable</span> <span class="o">=</span> <span class="nb">iter</span><span class="p">(</span><span class="n">o</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">list</span><span class="p">(</span><span class="n">iterable</span><span class="p">)</span>
    <span class="c1"># Let the base class default method raise the TypeError</span>
    <span class="k">return</span> <span class="n">JSONEncoder</span><span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">o</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.custom_loss">
<span id="mhcflurry-custom-loss-module"></span><h2>mhcflurry.custom_loss module<a class="headerlink" href="#module-mhcflurry.custom_loss" title="Permalink to this headline">¶</a></h2>
<p>Custom loss functions.</p>
<p>For losses supporting inequalities, each training data point is associated with
one of (=), (&lt;), or (&gt;). For e.g. (&gt;) inequalities, penalization is applied only
if the prediction is less than the given value.</p>
<dl class="function">
<dt id="mhcflurry.custom_loss.get_loss">
<code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">get_loss</code><span class="sig-paren">(</span><em class="sig-param">name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#get_loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.get_loss" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a custom_loss.Loss instance by name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>custom_loss.Loss</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="class">
<dt id="mhcflurry.custom_loss.Loss">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">Loss</code><span class="sig-paren">(</span><em class="sig-param">name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#Loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.Loss" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Thin wrapper to keep track of neural network loss functions, which could
be custom or baked into Keras.</p>
<p>Each subclass or instance should define these properties/methods:
- name : string
- loss : string or function</p>
<blockquote>
<div><p>This is what gets passed to keras.fit()</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>encode_y<span class="classifier">numpy.ndarray -&gt; numpy.ndarray</span></dt><dd><p>Transformation to apply to regression target before fitting</p>
</dd>
</dl>
</li>
</ul>
<dl class="method">
<dt id="mhcflurry.custom_loss.Loss.loss">
<code class="sig-name descname">loss</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">y_true</em>, <em class="sig-param">y_pred</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#Loss.loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.Loss.loss" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.Loss.get_keras_loss">
<code class="sig-name descname">get_keras_loss</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">reduction='sum_over_batch_size'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#Loss.get_keras_loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.Loss.get_keras_loss" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.custom_loss.StandardKerasLoss">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">StandardKerasLoss</code><span class="sig-paren">(</span><em class="sig-param">loss_name='mse'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#StandardKerasLoss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.StandardKerasLoss" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#mhcflurry.custom_loss.Loss" title="mhcflurry.custom_loss.Loss"><code class="xref py py-class docutils literal notranslate"><span class="pre">mhcflurry.custom_loss.Loss</span></code></a></p>
<p>A loss function supported by Keras, such as MSE.</p>
<dl class="attribute">
<dt id="mhcflurry.custom_loss.StandardKerasLoss.supports_inequalities">
<code class="sig-name descname">supports_inequalities</code><em class="property"> = False</em><a class="headerlink" href="#mhcflurry.custom_loss.StandardKerasLoss.supports_inequalities" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.StandardKerasLoss.supports_multiple_outputs">
<code class="sig-name descname">supports_multiple_outputs</code><em class="property"> = False</em><a class="headerlink" href="#mhcflurry.custom_loss.StandardKerasLoss.supports_multiple_outputs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.StandardKerasLoss.encode_y">
<em class="property">static </em><code class="sig-name descname">encode_y</code><span class="sig-paren">(</span><em class="sig-param">y</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#StandardKerasLoss.encode_y"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.StandardKerasLoss.encode_y" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.custom_loss.TransformPredictionsLossWrapper">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">TransformPredictionsLossWrapper</code><span class="sig-paren">(</span><em class="sig-param">loss</em>, <em class="sig-param">y_pred_transform=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#TransformPredictionsLossWrapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.TransformPredictionsLossWrapper" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#mhcflurry.custom_loss.Loss" title="mhcflurry.custom_loss.Loss"><code class="xref py py-class docutils literal notranslate"><span class="pre">mhcflurry.custom_loss.Loss</span></code></a></p>
<p>Wrapper that applies an arbitrary transform to y_pred before calling an
underlying loss function.</p>
<p>The y_pred_transform function should be a tensor -&gt; tensor function.</p>
<dl class="method">
<dt id="mhcflurry.custom_loss.TransformPredictionsLossWrapper.encode_y">
<code class="sig-name descname">encode_y</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#TransformPredictionsLossWrapper.encode_y"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.TransformPredictionsLossWrapper.encode_y" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.TransformPredictionsLossWrapper.loss">
<code class="sig-name descname">loss</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">y_true</em>, <em class="sig-param">y_pred</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#TransformPredictionsLossWrapper.loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.TransformPredictionsLossWrapper.loss" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.custom_loss.MSEWithInequalities">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">MSEWithInequalities</code><span class="sig-paren">(</span><em class="sig-param">name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#mhcflurry.custom_loss.Loss" title="mhcflurry.custom_loss.Loss"><code class="xref py py-class docutils literal notranslate"><span class="pre">mhcflurry.custom_loss.Loss</span></code></a></p>
<p>Supports training a regression model on data that includes inequalities
(e.g. x &lt; 100). Mean square error is used as the loss for elements with
an (=) inequality. For elements with e.g. a (&gt; 0.5) inequality, then the loss
for that element is (y - 0.5)^2 (standard MSE) if y &lt; 500 and 0 otherwise.</p>
<p>This loss assumes that the normal range for y_true and y_pred is 0 - 1. As a
hack, the implementation uses other intervals for y_pred to encode the
inequality information.</p>
<p>y_true is interpreted as follows:</p>
<dl class="simple">
<dt>between 0 - 1</dt><dd><p>Regular MSE loss is used. Penalty (y_pred - y_true)**2 is applied if
y_pred is greater or less than y_true.</p>
</dd>
<dt>between 2 - 3:</dt><dd><p>Treated as a “&gt;” inequality. Penalty (y_pred - (y_true - 2))**2 is
applied only if y_pred is less than y_true - 2.</p>
</dd>
<dt>between 4 - 5:</dt><dd><p>Treated as a “&lt;” inequality. Penalty (y_pred - (y_true - 4))**2 is
applied only if y_pred is greater than y_true - 4.</p>
</dd>
</dl>
<dl class="attribute">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.name">
<code class="sig-name descname">name</code><em class="property"> = 'mse_with_inequalities'</em><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.supports_inequalities">
<code class="sig-name descname">supports_inequalities</code><em class="property"> = True</em><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.supports_inequalities" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.supports_multiple_outputs">
<code class="sig-name descname">supports_multiple_outputs</code><em class="property"> = False</em><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.supports_multiple_outputs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.encode_y">
<em class="property">static </em><code class="sig-name descname">encode_y</code><span class="sig-paren">(</span><em class="sig-param">y</em>, <em class="sig-param">inequalities=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalities.encode_y"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.encode_y" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.MSEWithInequalities.loss">
<code class="sig-name descname">loss</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">y_true</em>, <em class="sig-param">y_pred</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalities.loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalities.loss" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">MSEWithInequalitiesAndMultipleOutputs</code><span class="sig-paren">(</span><em class="sig-param">name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalitiesAndMultipleOutputs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#mhcflurry.custom_loss.Loss" title="mhcflurry.custom_loss.Loss"><code class="xref py py-class docutils literal notranslate"><span class="pre">mhcflurry.custom_loss.Loss</span></code></a></p>
<p>Loss supporting inequalities and multiple outputs.</p>
<p>This loss assumes that the normal range for y_true and y_pred is 0 - 1. As a
hack, the implementation uses other intervals for y_pred to encode the
inequality and output-index information.</p>
<p>Inequalities are encoded into the regression target as in
the MSEWithInequalities loss.</p>
<p>Multiple outputs are encoded by mapping each regression target x (after
transforming for inequalities) using the rule x -&gt; x + i * 10 where i is
the output index.</p>
<p>The reason for explicitly encoding multiple outputs this way (rather than
just making the regression target a matrix instead of a vector) is that
in our use cases we frequently have missing data in the regression target.
This encoding gives a simple way to penalize only on (data point, output
index) pairs that have labels.</p>
<dl class="attribute">
<dt id="mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.name">
<code class="sig-name descname">name</code><em class="property"> = 'mse_with_inequalities_and_multiple_outputs'</em><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.supports_inequalities">
<code class="sig-name descname">supports_inequalities</code><em class="property"> = True</em><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.supports_inequalities" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.supports_multiple_outputs">
<code class="sig-name descname">supports_multiple_outputs</code><em class="property"> = True</em><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.supports_multiple_outputs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.encode_y">
<em class="property">static </em><code class="sig-name descname">encode_y</code><span class="sig-paren">(</span><em class="sig-param">y</em>, <em class="sig-param">inequalities=None</em>, <em class="sig-param">output_indices=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalitiesAndMultipleOutputs.encode_y"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.encode_y" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.loss">
<code class="sig-name descname">loss</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">y_true</em>, <em class="sig-param">y_pred</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MSEWithInequalitiesAndMultipleOutputs.loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MSEWithInequalitiesAndMultipleOutputs.loss" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.custom_loss.MultiallelicMassSpecLoss">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">MultiallelicMassSpecLoss</code><span class="sig-paren">(</span><em class="sig-param">delta=0.2</em>, <em class="sig-param">multiplier=1.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MultiallelicMassSpecLoss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MultiallelicMassSpecLoss" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#mhcflurry.custom_loss.Loss" title="mhcflurry.custom_loss.Loss"><code class="xref py py-class docutils literal notranslate"><span class="pre">mhcflurry.custom_loss.Loss</span></code></a></p>
<dl class="attribute">
<dt id="mhcflurry.custom_loss.MultiallelicMassSpecLoss.name">
<code class="sig-name descname">name</code><em class="property"> = 'multiallelic_mass_spec_loss'</em><a class="headerlink" href="#mhcflurry.custom_loss.MultiallelicMassSpecLoss.name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.MultiallelicMassSpecLoss.supports_inequalities">
<code class="sig-name descname">supports_inequalities</code><em class="property"> = True</em><a class="headerlink" href="#mhcflurry.custom_loss.MultiallelicMassSpecLoss.supports_inequalities" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.MultiallelicMassSpecLoss.supports_multiple_outputs">
<code class="sig-name descname">supports_multiple_outputs</code><em class="property"> = False</em><a class="headerlink" href="#mhcflurry.custom_loss.MultiallelicMassSpecLoss.supports_multiple_outputs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.MultiallelicMassSpecLoss.encode_y">
<em class="property">static </em><code class="sig-name descname">encode_y</code><span class="sig-paren">(</span><em class="sig-param">y</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MultiallelicMassSpecLoss.encode_y"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MultiallelicMassSpecLoss.encode_y" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.custom_loss.MultiallelicMassSpecLoss.loss">
<code class="sig-name descname">loss</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">y_true</em>, <em class="sig-param">y_pred</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#MultiallelicMassSpecLoss.loss"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.MultiallelicMassSpecLoss.loss" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="mhcflurry.custom_loss.check_shape">
<code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">check_shape</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">arr</em>, <em class="sig-param">expected_shape</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/custom_loss.html#check_shape"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.custom_loss.check_shape" title="Permalink to this definition">¶</a></dt>
<dd><p>Raise ValueError if arr.shape != expected_shape.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>name</strong><span class="classifier">string</span></dt><dd><p>Included in error message to aid debugging</p>
</dd>
<dt><strong>arr</strong><span class="classifier">numpy.ndarray</span></dt><dd></dd>
<dt><strong>expected_shape</strong><span class="classifier">tuple of int</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="mhcflurry.custom_loss.cls">
<code class="sig-prename descclassname">mhcflurry.custom_loss.</code><code class="sig-name descname">cls</code><a class="headerlink" href="#mhcflurry.custom_loss.cls" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="#mhcflurry.custom_loss.MultiallelicMassSpecLoss" title="mhcflurry.custom_loss.MultiallelicMassSpecLoss"><code class="xref py py-class docutils literal notranslate"><span class="pre">mhcflurry.custom_loss.MultiallelicMassSpecLoss</span></code></a></p>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.data_dependent_weights_initialization">
<span id="mhcflurry-data-dependent-weights-initialization-module"></span><h2>mhcflurry.data_dependent_weights_initialization module<a class="headerlink" href="#module-mhcflurry.data_dependent_weights_initialization" title="Permalink to this headline">¶</a></h2>
<p>Layer-sequential unit-variance initialization for neural networks.</p>
<dl class="simple">
<dt>See:</dt><dd><p>Mishkin and Matas, “All you need is a good init”. 2016.
<a class="reference external" href="https://arxiv.org/abs/1511.06422">https://arxiv.org/abs/1511.06422</a></p>
</dd>
</dl>
<dl class="function">
<dt id="mhcflurry.data_dependent_weights_initialization.svd_orthonormal">
<code class="sig-prename descclassname">mhcflurry.data_dependent_weights_initialization.</code><code class="sig-name descname">svd_orthonormal</code><span class="sig-paren">(</span><em class="sig-param">shape</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/data_dependent_weights_initialization.html#svd_orthonormal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.data_dependent_weights_initialization.svd_orthonormal" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.data_dependent_weights_initialization.get_activations">
<code class="sig-prename descclassname">mhcflurry.data_dependent_weights_initialization.</code><code class="sig-name descname">get_activations</code><span class="sig-paren">(</span><em class="sig-param">model</em>, <em class="sig-param">layer</em>, <em class="sig-param">X_batch</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/data_dependent_weights_initialization.html#get_activations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.data_dependent_weights_initialization.get_activations" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.data_dependent_weights_initialization.lsuv_init">
<code class="sig-prename descclassname">mhcflurry.data_dependent_weights_initialization.</code><code class="sig-name descname">lsuv_init</code><span class="sig-paren">(</span><em class="sig-param">model</em>, <em class="sig-param">batch</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">margin=0.1</em>, <em class="sig-param">max_iter=100</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/data_dependent_weights_initialization.html#lsuv_init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.data_dependent_weights_initialization.lsuv_init" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize neural network weights using layer-sequential unit-variance
initialization.</p>
<dl class="simple">
<dt>See:</dt><dd><p>Mishkin and Matas, “All you need is a good init”. 2016.
<a class="reference external" href="https://arxiv.org/abs/1511.06422">https://arxiv.org/abs/1511.06422</a></p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>model</strong><span class="classifier">keras.Model</span></dt><dd></dd>
<dt><strong>batch</strong><span class="classifier">dict</span></dt><dd><p>Training data, as would be passed keras.Model.fit()</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">boolean</span></dt><dd><p>Whether to print progress to stdout</p>
</dd>
<dt><strong>margin</strong><span class="classifier">float</span></dt><dd></dd>
<dt><strong>max_iter</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>keras.Model</dt><dd><p>Same as what was passed in.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.downloads">
<span id="mhcflurry-downloads-module"></span><h2>mhcflurry.downloads module<a class="headerlink" href="#module-mhcflurry.downloads" title="Permalink to this headline">¶</a></h2>
<p>Manage local downloaded data.</p>
<dl class="function">
<dt id="mhcflurry.downloads.get_downloads_dir">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">get_downloads_dir</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_downloads_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_downloads_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the path to local downloaded data</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_current_release">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">get_current_release</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_current_release"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_current_release" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the current downloaded data release</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_downloads_metadata">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">get_downloads_metadata</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_downloads_metadata"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_downloads_metadata" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the contents of downloads.yml as a dict</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_default_class1_models_dir">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">get_default_class1_models_dir</code><span class="sig-paren">(</span><em class="sig-param">test_exists=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_default_class1_models_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_default_class1_models_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the absolute path to the default class1 models dir.</p>
<p>If environment variable MHCFLURRY_DEFAULT_CLASS1_MODELS is set to an
absolute path, return that path. If it’s set to a relative path (i.e. does
not start with /) then return that path taken to be relative to the mhcflurry
downloads dir.</p>
<p>If environment variable MHCFLURRY_DEFAULT_CLASS1_MODELS is NOT set,
then return the path to downloaded models in the “models_class1” download.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>test_exists</strong><span class="classifier">boolean, optional</span></dt><dd><p>Whether to raise an exception of the path does not exist</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>string</strong><span class="classifier">absolute path</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_default_class1_presentation_models_dir">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">get_default_class1_presentation_models_dir</code><span class="sig-paren">(</span><em class="sig-param">test_exists=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_default_class1_presentation_models_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_default_class1_presentation_models_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the absolute path to the default class1 presentation models dir.</p>
<p>See <a class="reference internal" href="#mhcflurry.downloads.get_default_class1_models_dir" title="mhcflurry.downloads.get_default_class1_models_dir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_default_class1_models_dir</span></code></a>.</p>
<p>If environment variable MHCFLURRY_DEFAULT_CLASS1_PRESENTATION_MODELS is set
to an absolute path, return that path. If it’s set to a relative path (does
not start with /) then return that path taken to be relative to the mhcflurry
downloads dir.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>test_exists</strong><span class="classifier">boolean, optional</span></dt><dd><p>Whether to raise an exception of the path does not exist</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>string</strong><span class="classifier">absolute path</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_default_class1_processing_models_dir">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">get_default_class1_processing_models_dir</code><span class="sig-paren">(</span><em class="sig-param">test_exists=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_default_class1_processing_models_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_default_class1_processing_models_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the absolute path to the default class1 processing models dir.</p>
<p>See <a class="reference internal" href="#mhcflurry.downloads.get_default_class1_models_dir" title="mhcflurry.downloads.get_default_class1_models_dir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_default_class1_models_dir</span></code></a>.</p>
<p>If environment variable MHCFLURRY_DEFAULT_CLASS1_PROCESSING_MODELS is set
to an absolute path, return that path. If it’s set to a relative path (does
not start with /) then return that path taken to be relative to the mhcflurry
downloads dir.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>test_exists</strong><span class="classifier">boolean, optional</span></dt><dd><p>Whether to raise an exception of the path does not exist</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>string</strong><span class="classifier">absolute path</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_current_release_downloads">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">get_current_release_downloads</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_current_release_downloads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_current_release_downloads" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dict of all available downloads in the current release.</p>
<p>The dict keys are the names of the downloads. The values are a dict
with two entries:</p>
<dl class="simple">
<dt>downloaded<span class="classifier">bool</span></dt><dd><p>Whether the download is currently available locally</p>
</dd>
<dt>metadata<span class="classifier">dict</span></dt><dd><p>Info about the download from downloads.yml such as URL</p>
</dd>
<dt>up_to_date<span class="classifier">bool or None</span></dt><dd><p>Whether the download URL(s) match what was used to download the current
data. This is None if it cannot be determined.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.get_path">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">get_path</code><span class="sig-paren">(</span><em class="sig-param">download_name</em>, <em class="sig-param">filename=''</em>, <em class="sig-param">test_exists=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#get_path"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.get_path" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the local path to a file in a MHCflurry download</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>download_name</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>filename</strong><span class="classifier">string</span></dt><dd><p>Relative path within the download to the file of interest</p>
</dd>
<dt><strong>test_exists</strong><span class="classifier">boolean</span></dt><dd><p>If True (default) throw an error telling the user how to download the
data if the file does not exist</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>string giving local absolute path</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads.configure">
<code class="sig-prename descclassname">mhcflurry.downloads.</code><code class="sig-name descname">configure</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads.html#configure"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads.configure" title="Permalink to this definition">¶</a></dt>
<dd><p>Setup various global variables based on environment variables.</p>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.downloads_command">
<span id="mhcflurry-downloads-command-module"></span><h2>mhcflurry.downloads_command module<a class="headerlink" href="#module-mhcflurry.downloads_command" title="Permalink to this headline">¶</a></h2>
<p>Download MHCflurry released datasets and trained models.</p>
<p>Examples</p>
<dl class="simple">
<dt>Fetch the default downloads:</dt><dd><p>$ mhcflurry-downloads fetch</p>
</dd>
<dt>Fetch a specific download:</dt><dd><p>$ mhcflurry-downloads fetch models_class1_pan</p>
</dd>
<dt>Get the path to a download:</dt><dd><p>$ mhcflurry-downloads path models_class1_pan</p>
</dd>
<dt>Get the URL of a download:</dt><dd><p>$ mhcflurry-downloads url models_class1_pan</p>
</dd>
<dt>Summarize available and fetched downloads:</dt><dd><p>$ mhcflurry-downloads info</p>
</dd>
</dl>
<dl class="function">
<dt id="mhcflurry.downloads_command.run">
<code class="sig-prename descclassname">mhcflurry.downloads_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.mkdir_p">
<code class="sig-prename descclassname">mhcflurry.downloads_command.</code><code class="sig-name descname">mkdir_p</code><span class="sig-paren">(</span><em class="sig-param">path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#mkdir_p"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.mkdir_p" title="Permalink to this definition">¶</a></dt>
<dd><p>Make directories as needed, similar to mkdir -p in a shell.</p>
<p>From:
<a class="reference external" href="http://stackoverflow.com/questions/600268/mkdir-p-functionality-in-python">http://stackoverflow.com/questions/600268/mkdir-p-functionality-in-python</a></p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.yes_no">
<code class="sig-prename descclassname">mhcflurry.downloads_command.</code><code class="sig-name descname">yes_no</code><span class="sig-paren">(</span><em class="sig-param">boolean</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#yes_no"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.yes_no" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="mhcflurry.downloads_command.TqdmUpTo">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.downloads_command.</code><code class="sig-name descname">TqdmUpTo</code><span class="sig-paren">(</span><em class="sig-param">iterable=None</em>, <em class="sig-param">desc=None</em>, <em class="sig-param">total=None</em>, <em class="sig-param">leave=True</em>, <em class="sig-param">file=None</em>, <em class="sig-param">ncols=None</em>, <em class="sig-param">mininterval=0.1</em>, <em class="sig-param">maxinterval=10.0</em>, <em class="sig-param">miniters=None</em>, <em class="sig-param">ascii=None</em>, <em class="sig-param">disable=False</em>, <em class="sig-param">unit='it'</em>, <em class="sig-param">unit_scale=False</em>, <em class="sig-param">dynamic_ncols=False</em>, <em class="sig-param">smoothing=0.3</em>, <em class="sig-param">bar_format=None</em>, <em class="sig-param">initial=0</em>, <em class="sig-param">position=None</em>, <em class="sig-param">postfix=None</em>, <em class="sig-param">unit_divisor=1000</em>, <em class="sig-param">gui=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#TqdmUpTo"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.TqdmUpTo" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tqdm._tqdm.tqdm</span></code></p>
<p>Provides <code class="xref py py-obj docutils literal notranslate"><span class="pre">update_to(n)</span></code> which uses <code class="xref py py-obj docutils literal notranslate"><span class="pre">tqdm.update(delta_n)</span></code>.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl>
<dt><strong>iterable</strong><span class="classifier">iterable, optional</span></dt><dd><p>Iterable to decorate with a progressbar.
Leave blank to manually manage the updates.</p>
</dd>
<dt><strong>desc</strong><span class="classifier">str, optional</span></dt><dd><p>Prefix for the progressbar.</p>
</dd>
<dt><strong>total</strong><span class="classifier">int, optional</span></dt><dd><p>The number of expected iterations. If unspecified,
len(iterable) is used if possible. As a last resort, only basic
progress statistics are displayed (no ETA, no progressbar).
If <code class="xref py py-obj docutils literal notranslate"><span class="pre">gui</span></code> is True and this parameter needs subsequent updating,
specify an initial arbitrary large positive integer,
e.g. int(9e9).</p>
</dd>
<dt><strong>leave</strong><span class="classifier">bool, optional</span></dt><dd><p>If [default: True], keeps all traces of the progressbar
upon termination of iteration.</p>
</dd>
<dt><strong>file</strong><span class="classifier"><code class="xref py py-obj docutils literal notranslate"><span class="pre">io.TextIOWrapper</span></code> or <code class="xref py py-obj docutils literal notranslate"><span class="pre">io.StringIO</span></code>, optional</span></dt><dd><p>Specifies where to output the progress messages
(default: sys.stderr). Uses <code class="xref py py-obj docutils literal notranslate"><span class="pre">file.write(str)</span></code> and <code class="xref py py-obj docutils literal notranslate"><span class="pre">file.flush()</span></code>
methods.</p>
</dd>
<dt><strong>ncols</strong><span class="classifier">int, optional</span></dt><dd><p>The width of the entire output message. If specified,
dynamically resizes the progressbar to stay within this bound.
If unspecified, attempts to use environment width. The
fallback is a meter width of 10 and no limit for the counter and
statistics. If 0, will not print any meter (only stats).</p>
</dd>
<dt><strong>mininterval</strong><span class="classifier">float, optional</span></dt><dd><p>Minimum progress display update interval, in seconds [default: 0.1].</p>
</dd>
<dt><strong>maxinterval</strong><span class="classifier">float, optional</span></dt><dd><p>Maximum progress display update interval, in seconds [default: 10].
Automatically adjusts <code class="xref py py-obj docutils literal notranslate"><span class="pre">miniters</span></code> to correspond to <code class="xref py py-obj docutils literal notranslate"><span class="pre">mininterval</span></code>
after long display update lag. Only works if <code class="xref py py-obj docutils literal notranslate"><span class="pre">dynamic_miniters</span></code>
or monitor thread is enabled.</p>
</dd>
<dt><strong>miniters</strong><span class="classifier">int, optional</span></dt><dd><p>Minimum progress display update interval, in iterations.
If 0 and <code class="xref py py-obj docutils literal notranslate"><span class="pre">dynamic_miniters</span></code>, will automatically adjust to equal
<code class="xref py py-obj docutils literal notranslate"><span class="pre">mininterval</span></code> (more CPU efficient, good for tight loops).
If &gt; 0, will skip display of specified number of iterations.
Tweak this and <code class="xref py py-obj docutils literal notranslate"><span class="pre">mininterval</span></code> to get very efficient loops.
If your progress is erratic with both fast and slow iterations
(network, skipping items, etc) you should set miniters=1.</p>
</dd>
<dt><strong>ascii</strong><span class="classifier">bool, optional</span></dt><dd><p>If unspecified or False, use unicode (smooth blocks) to fill
the meter. The fallback is to use ASCII characters <code class="xref py py-obj docutils literal notranslate"><span class="pre">1-9</span> <span class="pre">#</span></code>.</p>
</dd>
<dt><strong>disable</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to disable the entire progressbar wrapper
[default: False]. If set to None, disable on non-TTY.</p>
</dd>
<dt><strong>unit</strong><span class="classifier">str, optional</span></dt><dd><p>String that will be used to define the unit of each iteration
[default: it].</p>
</dd>
<dt><strong>unit_scale</strong><span class="classifier">bool or int or float, optional</span></dt><dd><p>If 1 or True, the number of iterations will be reduced/scaled
automatically and a metric prefix following the
International System of Units standard will be added
(kilo, mega, etc.) [default: False]. If any other non-zero
number, will scale <code class="xref py py-obj docutils literal notranslate"><span class="pre">total</span></code> and <code class="xref py py-obj docutils literal notranslate"><span class="pre">n</span></code>.</p>
</dd>
<dt><strong>dynamic_ncols</strong><span class="classifier">bool, optional</span></dt><dd><p>If set, constantly alters <code class="xref py py-obj docutils literal notranslate"><span class="pre">ncols</span></code> to the environment (allowing
for window resizes) [default: False].</p>
</dd>
<dt><strong>smoothing</strong><span class="classifier">float, optional</span></dt><dd><p>Exponential moving average smoothing factor for speed estimates
(ignored in GUI mode). Ranges from 0 (average speed) to 1
(current/instantaneous speed) [default: 0.3].</p>
</dd>
<dt><strong>bar_format</strong><span class="classifier">str, optional</span></dt><dd><p>Specify a custom bar string formatting. May impact performance.
[default: ‘{l_bar}{bar}{r_bar}’], where
l_bar=’{desc}: {percentage:3.0f}%|’ and
r_bar=’| {n_fmt}/{total_fmt} [{elapsed}&lt;{remaining}, ‘</p>
<blockquote>
<div><p>‘{rate_fmt}{postfix}]’</p>
</div></blockquote>
<dl class="simple">
<dt>Possible vars: l_bar, bar, r_bar, n, n_fmt, total, total_fmt,</dt><dd><p>percentage, rate, rate_fmt, rate_noinv, rate_noinv_fmt,
rate_inv, rate_inv_fmt, elapsed, remaining, desc, postfix.</p>
</dd>
</dl>
<p>Note that a trailing “: ” is automatically removed after {desc}
if the latter is empty.</p>
</dd>
<dt><strong>initial</strong><span class="classifier">int, optional</span></dt><dd><p>The initial counter value. Useful when restarting a progress
bar [default: 0].</p>
</dd>
<dt><strong>position</strong><span class="classifier">int, optional</span></dt><dd><p>Specify the line offset to print this bar (starting from 0)
Automatic if unspecified.
Useful to manage multiple bars at once (eg, from threads).</p>
</dd>
<dt><strong>postfix</strong><span class="classifier">dict or <a href="#id2"><span class="problematic" id="id3">*</span></a>, optional</span></dt><dd><p>Specify additional stats to display at the end of the bar.
Calls <code class="xref py py-obj docutils literal notranslate"><span class="pre">set_postfix(**postfix)</span></code> if possible (dict).</p>
</dd>
<dt><strong>unit_divisor</strong><span class="classifier">float, optional</span></dt><dd><p>[default: 1000], ignored unless <code class="xref py py-obj docutils literal notranslate"><span class="pre">unit_scale</span></code> is True.</p>
</dd>
<dt><strong>gui</strong><span class="classifier">bool, optional</span></dt><dd><p>WARNING: internal parameter - do not use.
Use tqdm_gui(…) instead. If set, will attempt to use
matplotlib animations for a graphical output [default: False].</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>out</strong><span class="classifier">decorated iterator.</span></dt><dd></dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="mhcflurry.downloads_command.TqdmUpTo.update_to">
<code class="sig-name descname">update_to</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">b=1</em>, <em class="sig-param">bsize=1</em>, <em class="sig-param">tsize=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#TqdmUpTo.update_to"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.TqdmUpTo.update_to" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>b<span class="classifier">int, optional</span></dt><dd><p>Number of blocks transferred so far [default: 1].</p>
</dd>
<dt>bsize<span class="classifier">int, optional</span></dt><dd><p>Size of each block (in tqdm units) [default: 1].</p>
</dd>
<dt>tsize<span class="classifier">int, optional</span></dt><dd><p>Total size (in tqdm units). If [default: None] remains unchanged.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.fetch_subcommand">
<code class="sig-prename descclassname">mhcflurry.downloads_command.</code><code class="sig-name descname">fetch_subcommand</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#fetch_subcommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.fetch_subcommand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.info_subcommand">
<code class="sig-prename descclassname">mhcflurry.downloads_command.</code><code class="sig-name descname">info_subcommand</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#info_subcommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.info_subcommand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.path_subcommand">
<code class="sig-prename descclassname">mhcflurry.downloads_command.</code><code class="sig-name descname">path_subcommand</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#path_subcommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.path_subcommand" title="Permalink to this definition">¶</a></dt>
<dd><p>Print the local path to a download</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.downloads_command.url_subcommand">
<code class="sig-prename descclassname">mhcflurry.downloads_command.</code><code class="sig-name descname">url_subcommand</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/downloads_command.html#url_subcommand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.downloads_command.url_subcommand" title="Permalink to this definition">¶</a></dt>
<dd><p>Print the URL(s) for a download</p>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.encodable_sequences">
<span id="mhcflurry-encodable-sequences-module"></span><h2>mhcflurry.encodable_sequences module<a class="headerlink" href="#module-mhcflurry.encodable_sequences" title="Permalink to this headline">¶</a></h2>
<p>Class for encoding variable-length peptides to fixed-size numerical matrices</p>
<dl class="exception">
<dt id="mhcflurry.encodable_sequences.EncodingError">
<em class="property">exception </em><code class="sig-prename descclassname">mhcflurry.encodable_sequences.</code><code class="sig-name descname">EncodingError</code><span class="sig-paren">(</span><em class="sig-param">message</em>, <em class="sig-param">supported_peptide_lengths</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodingError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodingError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code></p>
<p>Exception raised when peptides cannot be encoded</p>
</dd></dl>

<dl class="class">
<dt id="mhcflurry.encodable_sequences.EncodableSequences">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.encodable_sequences.</code><code class="sig-name descname">EncodableSequences</code><span class="sig-paren">(</span><em class="sig-param">sequences</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class for encoding variable-length peptides to fixed-size numerical matrices</p>
<p>This class caches various encodings of a list of sequences.</p>
<p>In practice this is used only for peptides. To encode MHC allele sequences,
see AlleleEncoding.</p>
<dl class="attribute">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.unknown_character">
<code class="sig-name descname">unknown_character</code><em class="property"> = 'X'</em><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.unknown_character" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.create">
<em class="property">classmethod </em><code class="sig-name descname">create</code><span class="sig-paren">(</span><em class="sig-param">sequences</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences.create"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.create" title="Permalink to this definition">¶</a></dt>
<dd><p>Factory that returns an EncodableSequences given a list of
strings. As a convenience, you can also pass it an EncodableSequences
instance, in which case the object is returned unchanged.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.variable_length_to_fixed_length_categorical">
<code class="sig-name descname">variable_length_to_fixed_length_categorical</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">alignment_method='pad_middle'</em>, <em class="sig-param">left_edge=4</em>, <em class="sig-param">right_edge=4</em>, <em class="sig-param">max_length=15</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences.variable_length_to_fixed_length_categorical"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.variable_length_to_fixed_length_categorical" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode variable-length sequences to a fixed-size index-encoded (integer)
matrix.</p>
<p>See <a class="reference internal" href="#mhcflurry.encodable_sequences.EncodableSequences.sequences_to_fixed_length_index_encoded_array" title="mhcflurry.encodable_sequences.EncodableSequences.sequences_to_fixed_length_index_encoded_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequences_to_fixed_length_index_encoded_array</span></code></a> for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>alignment_method</strong><span class="classifier">string</span></dt><dd><p>One of “pad_middle” or “left_pad_right_pad”</p>
</dd>
<dt><strong>left_edge</strong><span class="classifier">int, size of fixed-position left side</span></dt><dd><p>Only relevant for pad_middle alignment method</p>
</dd>
<dt><strong>right_edge</strong><span class="classifier">int, size of the fixed-position right side</span></dt><dd><p>Only relevant for pad_middle alignment method</p>
</dd>
<dt><strong>max_length</strong><span class="classifier">maximum supported peptide length</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of integers with shape (num sequences, encoded length)</dt><dd></dd>
<dt>For pad_middle, the encoded length is max_length. For left_pad_right_pad,</dt><dd></dd>
<dt>it’s 3 * max_length.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.variable_length_to_fixed_length_vector_encoding">
<code class="sig-name descname">variable_length_to_fixed_length_vector_encoding</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">vector_encoding_name</em>, <em class="sig-param">alignment_method='pad_middle'</em>, <em class="sig-param">left_edge=4</em>, <em class="sig-param">right_edge=4</em>, <em class="sig-param">max_length=15</em>, <em class="sig-param">trim=False</em>, <em class="sig-param">allow_unsupported_amino_acids=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences.variable_length_to_fixed_length_vector_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.variable_length_to_fixed_length_vector_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode variable-length sequences to a fixed-size matrix. Amino acids
are encoded as specified by the vector_encoding_name argument.</p>
<p>See <a class="reference internal" href="#mhcflurry.encodable_sequences.EncodableSequences.sequences_to_fixed_length_index_encoded_array" title="mhcflurry.encodable_sequences.EncodableSequences.sequences_to_fixed_length_index_encoded_array"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequences_to_fixed_length_index_encoded_array</span></code></a> for details.</p>
<p>See also: variable_length_to_fixed_length_categorical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>vector_encoding_name</strong><span class="classifier">string</span></dt><dd><p>How to represent amino acids.
One of “BLOSUM62”, “one-hot”, etc. Full list of supported vector
encodings is given by available_vector_encodings().</p>
</dd>
<dt><strong>alignment_method</strong><span class="classifier">string</span></dt><dd><p>One of “pad_middle” or “left_pad_right_pad”</p>
</dd>
<dt><strong>left_edge</strong><span class="classifier">int</span></dt><dd><p>Size of fixed-position left side.
Only relevant for pad_middle alignment method</p>
</dd>
<dt><strong>right_edge</strong><span class="classifier">int</span></dt><dd><p>Size of the fixed-position right side.
Only relevant for pad_middle alignment method</p>
</dd>
<dt><strong>max_length</strong><span class="classifier">int</span></dt><dd><p>Maximum supported peptide length</p>
</dd>
<dt><strong>trim</strong><span class="classifier">bool</span></dt><dd><p>If True, longer sequences will be trimmed to fit the maximum
supported length. Not supported for all alignment methods.</p>
</dd>
<dt><strong>allow_unsupported_amino_acids</strong><span class="classifier">bool</span></dt><dd><p>If True, non-canonical amino acids will be replaced with the X
character before encoding.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array with shape (num sequences, encoded length, m)</dt><dd></dd>
<dt>where</dt><dd><ul class="simple">
<li><p>m is the vector encoding length (usually 21).</p></li>
<li><p>encoded length is max_length if alignment_method is pad_middle;
3 * max_length if it’s left_pad_right_pad.</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.encodable_sequences.EncodableSequences.sequences_to_fixed_length_index_encoded_array">
<em class="property">classmethod </em><code class="sig-name descname">sequences_to_fixed_length_index_encoded_array</code><span class="sig-paren">(</span><em class="sig-param">sequences</em>, <em class="sig-param">alignment_method='pad_middle'</em>, <em class="sig-param">left_edge=4</em>, <em class="sig-param">right_edge=4</em>, <em class="sig-param">max_length=15</em>, <em class="sig-param">trim=False</em>, <em class="sig-param">allow_unsupported_amino_acids=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/encodable_sequences.html#EncodableSequences.sequences_to_fixed_length_index_encoded_array"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.encodable_sequences.EncodableSequences.sequences_to_fixed_length_index_encoded_array" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode variable-length sequences to a fixed-size index-encoded (integer)
matrix.</p>
<p>How variable length sequences get mapped to fixed length is set by the
“alignment_method” argument. Supported alignment methods are:</p>
<blockquote>
<div><dl>
<dt>pad_middle</dt><dd><p>Encoding designed for preserving the anchor positions of class
I peptides. This is what is used in allele-specific models.</p>
<p>Each string must be of length at least left_edge + right_edge
and at most max_length. The first left_edge characters in the
input always map to the first left_edge characters in the
output. Similarly for the last right_edge characters. The
middle characters are filled in based on the length, with the
X character filling in the blanks.</p>
<p>Example:</p>
<p>AAAACDDDD -&gt; AAAAXXXCXXXDDDD</p>
</dd>
<dt>left_pad_centered_right_pad</dt><dd><p>Encoding that makes no assumptions on anchor positions but is
3x larger than pad_middle, since it duplicates the peptide
(left aligned + centered + right aligned). This is what is used
for the pan-allele models.</p>
<p>Example:</p>
<p>AAAACDDDD -&gt; AAAACDDDDXXXXXXXXXAAAACDDDDXXXXXXXXXAAAACDDDD</p>
</dd>
<dt>left_pad_right_pad</dt><dd><p>Same as left_pad_centered_right_pad but only includes left-
and right-padded peptide.</p>
<p>Example:</p>
<p>AAAACDDDD -&gt; AAAACDDDDXXXXXXXXXXXXAAAACDDDD</p>
</dd>
</dl>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequences</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>alignment_method</strong><span class="classifier">string</span></dt><dd><p>One of “pad_middle” or “left_pad_right_pad”</p>
</dd>
<dt><strong>left_edge</strong><span class="classifier">int</span></dt><dd><p>Size of fixed-position left side.
Only relevant for pad_middle alignment method</p>
</dd>
<dt><strong>right_edge</strong><span class="classifier">int</span></dt><dd><p>Size of the fixed-position right side.
Only relevant for pad_middle alignment method</p>
</dd>
<dt><strong>max_length</strong><span class="classifier">int</span></dt><dd><p>maximum supported peptide length</p>
</dd>
<dt><strong>trim</strong><span class="classifier">bool</span></dt><dd><p>If True, longer sequences will be trimmed to fit the maximum
supported length. Not supported for all alignment methods.</p>
</dd>
<dt><strong>allow_unsupported_amino_acids</strong><span class="classifier">bool</span></dt><dd><p>If True, non-canonical amino acids will be replaced with the X
character before encoding.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of integers with shape (num sequences, encoded length)</dt><dd></dd>
<dt>For pad_middle, the encoded length is max_length. For left_pad_right_pad,</dt><dd></dd>
<dt>it’s 2 * max_length. For left_pad_centered_right_pad, it’s</dt><dd></dd>
<dt>3 * max_length.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.ensemble_centrality">
<span id="mhcflurry-ensemble-centrality-module"></span><h2>mhcflurry.ensemble_centrality module<a class="headerlink" href="#module-mhcflurry.ensemble_centrality" title="Permalink to this headline">¶</a></h2>
<p>Measures of centrality (e.g. mean) used to combine predictions across an
ensemble. The input to these functions are log affinities, and they are expected
to return a centrality measure also in log-space.</p>
<dl class="function">
<dt id="mhcflurry.ensemble_centrality.robust_mean">
<code class="sig-prename descclassname">mhcflurry.ensemble_centrality.</code><code class="sig-name descname">robust_mean</code><span class="sig-paren">(</span><em class="sig-param">log_values</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/ensemble_centrality.html#robust_mean"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.ensemble_centrality.robust_mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Mean of values falling within the 25-75 percentiles.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>log_values</strong><span class="classifier">2-d numpy.array</span></dt><dd><p>Center is computed along the second axis (i.e. per row).</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt><strong>center</strong><span class="classifier">numpy.array of length log_values.shape[1]</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.fasta">
<span id="mhcflurry-fasta-module"></span><h2>mhcflurry.fasta module<a class="headerlink" href="#module-mhcflurry.fasta" title="Permalink to this headline">¶</a></h2>
<p>Adapted from pyensembl, github.com/openvax/pyensembl
Original implementation by Alex Rubinsteyn.</p>
<p>The worse sin in bioinformatics is to write your own FASTA parser.
We’re doing this to avoid adding another dependency to MHCflurry, however.</p>
<dl class="function">
<dt id="mhcflurry.fasta.read_fasta_to_dataframe">
<code class="sig-prename descclassname">mhcflurry.fasta.</code><code class="sig-name descname">read_fasta_to_dataframe</code><span class="sig-paren">(</span><em class="sig-param">filename</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/fasta.html#read_fasta_to_dataframe"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.fasta.read_fasta_to_dataframe" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="mhcflurry.fasta.FastaParser">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.fasta.</code><code class="sig-name descname">FastaParser</code><a class="reference internal" href="_modules/mhcflurry/fasta.html#FastaParser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.fasta.FastaParser" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>FastaParser object consumes lines of a FASTA file incrementally.</p>
<dl class="method">
<dt id="mhcflurry.fasta.FastaParser.iterate_over_file">
<code class="sig-name descname">iterate_over_file</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">fasta_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/fasta.html#FastaParser.iterate_over_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.fasta.FastaParser.iterate_over_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Generator that yields identifiers paired with sequences.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.fasta.FastaParser.open_file">
<em class="property">static </em><code class="sig-name descname">open_file</code><span class="sig-paren">(</span><em class="sig-param">fasta_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/fasta.html#FastaParser.open_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.fasta.FastaParser.open_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Open either a text file or compressed gzip file as a stream of bytes.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.flanking_encoding">
<span id="mhcflurry-flanking-encoding-module"></span><h2>mhcflurry.flanking_encoding module<a class="headerlink" href="#module-mhcflurry.flanking_encoding" title="Permalink to this headline">¶</a></h2>
<p>Class for encoding variable-length flanking and peptides to
fixed-size numerical matrices</p>
<dl class="class">
<dt id="mhcflurry.flanking_encoding.EncodingResult">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.flanking_encoding.</code><code class="sig-name descname">EncodingResult</code><span class="sig-paren">(</span><em class="sig-param">array</em>, <em class="sig-param">peptide_lengths</em><span class="sig-paren">)</span><a class="headerlink" href="#mhcflurry.flanking_encoding.EncodingResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<p>Create new instance of EncodingResult(array, peptide_lengths)</p>
<dl class="method">
<dt id="mhcflurry.flanking_encoding.EncodingResult.array">
<em class="property">property </em><code class="sig-name descname">array</code><a class="headerlink" href="#mhcflurry.flanking_encoding.EncodingResult.array" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.flanking_encoding.EncodingResult.peptide_lengths">
<em class="property">property </em><code class="sig-name descname">peptide_lengths</code><a class="headerlink" href="#mhcflurry.flanking_encoding.EncodingResult.peptide_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.flanking_encoding.FlankingEncoding">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.flanking_encoding.</code><code class="sig-name descname">FlankingEncoding</code><span class="sig-paren">(</span><em class="sig-param">peptides</em>, <em class="sig-param">n_flanks</em>, <em class="sig-param">c_flanks</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/flanking_encoding.html#FlankingEncoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.flanking_encoding.FlankingEncoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Encode peptides and optionally their N- and C-flanking sequences into fixed
size numerical matrices. Similar to EncodableSequences but with support
for flanking sequences and the encoding scheme used by the processing
predictor.</p>
<p>Instances of this class have an immutable list of peptides with
flanking sequences. Encodings are cached in the instances for faster
performance when the same set of peptides needs to encoded more than once.</p>
<p>Constructor. Sequences of any lengths can be passed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd><p>Peptide sequences</p>
</dd>
<dt><strong>n_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>Upstream sequences</p>
</dd>
<dt><strong>c_flanks</strong><span class="classifier">list of string [same length as peptides]</span></dt><dd><p>Downstream sequences</p>
</dd>
</dl>
</dd>
</dl>
<dl class="attribute">
<dt id="mhcflurry.flanking_encoding.FlankingEncoding.unknown_character">
<code class="sig-name descname">unknown_character</code><em class="property"> = 'X'</em><a class="headerlink" href="#mhcflurry.flanking_encoding.FlankingEncoding.unknown_character" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.flanking_encoding.FlankingEncoding.vector_encode">
<code class="sig-name descname">vector_encode</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">vector_encoding_name</em>, <em class="sig-param">peptide_max_length</em>, <em class="sig-param">n_flank_length</em>, <em class="sig-param">c_flank_length</em>, <em class="sig-param">allow_unsupported_amino_acids=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/flanking_encoding.html#FlankingEncoding.vector_encode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.flanking_encoding.FlankingEncoding.vector_encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode variable-length sequences to a fixed-size matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>vector_encoding_name</strong><span class="classifier">string</span></dt><dd><p>How to represent amino acids. One of “BLOSUM62”, “one-hot”, etc.
See <code class="xref py py-obj docutils literal notranslate"><span class="pre">amino_acid.available_vector_encodings()</span></code>.</p>
</dd>
<dt><strong>peptide_max_length</strong><span class="classifier">int</span></dt><dd><p>Maximum supported peptide length.</p>
</dd>
<dt><strong>n_flank_length</strong><span class="classifier">int</span></dt><dd><p>Maximum supported N-flank length</p>
</dd>
<dt><strong>c_flank_length</strong><span class="classifier">int</span></dt><dd><p>Maximum supported C-flank length</p>
</dd>
<dt><strong>allow_unsupported_amino_acids</strong><span class="classifier">bool</span></dt><dd><p>If True, non-canonical amino acids will be replaced with the X
character before encoding.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array with shape (num sequences, length, m)</dt><dd></dd>
<dt>where</dt><dd><ul class="simple">
<li><p>num sequences is number of peptides, i.e. len(self)</p></li>
<li><p>length is peptide_max_length + n_flank_length + c_flank_length</p></li>
<li><p>m is the vector encoding length (usually 21).</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.flanking_encoding.FlankingEncoding.encode">
<em class="property">static </em><code class="sig-name descname">encode</code><span class="sig-paren">(</span><em class="sig-param">vector_encoding_name</em>, <em class="sig-param">df</em>, <em class="sig-param">peptide_max_length</em>, <em class="sig-param">n_flank_length</em>, <em class="sig-param">c_flank_length</em>, <em class="sig-param">allow_unsupported_amino_acids=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/flanking_encoding.html#FlankingEncoding.encode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.flanking_encoding.FlankingEncoding.encode" title="Permalink to this definition">¶</a></dt>
<dd><p>Encode variable-length sequences to a fixed-size matrix.</p>
<p>Helper function. Users should use <a class="reference internal" href="#mhcflurry.flanking_encoding.FlankingEncoding.vector_encode" title="mhcflurry.flanking_encoding.FlankingEncoding.vector_encode"><code class="xref py py-obj docutils literal notranslate"><span class="pre">vector_encode</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>vector_encoding_name</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>df</strong><span class="classifier">pandas.DataFrame</span></dt><dd></dd>
<dt><strong>peptide_max_length</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>n_flank_length</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>c_flank_length</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>allow_unsupported_amino_acids</strong><span class="classifier">bool</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.hyperparameters">
<span id="mhcflurry-hyperparameters-module"></span><h2>mhcflurry.hyperparameters module<a class="headerlink" href="#module-mhcflurry.hyperparameters" title="Permalink to this headline">¶</a></h2>
<p>Hyperparameter (neural network options) management</p>
<dl class="class">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.hyperparameters.</code><code class="sig-name descname">HyperparameterDefaults</code><span class="sig-paren">(</span><em class="sig-param">**defaults</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class for managing hyperparameters. Thin wrapper around a dict.</p>
<p>Instances of this class are a specification of the hyperparameters
<em>supported</em> by a model and their defaults. The particular
hyperparameter settings to be used, for example, to train a model
are kept in plain dicts.</p>
<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.extend">
<code class="sig-name descname">extend</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">other</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.extend"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.extend" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a new HyperparameterDefaults instance containing the
hyperparameters from the current instance combined with
those from other.</p>
<p>It is an error if self and other have any hyperparameters in
common.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.with_defaults">
<code class="sig-name descname">with_defaults</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.with_defaults"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.with_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dict of hyperparameter settings, return a dict containing
those settings augmented by the defaults for any keys missing from
the dict.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.subselect">
<code class="sig-name descname">subselect</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.subselect"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.subselect" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter a dict of hyperparameter settings to only those keys defined
in this HyperparameterDefaults  .</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.check_valid_keys">
<code class="sig-name descname">check_valid_keys</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">obj</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.check_valid_keys"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.check_valid_keys" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a dict of hyperparameter settings, throw an exception if any
keys are not defined in this HyperparameterDefaults instance.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.hyperparameters.HyperparameterDefaults.models_grid">
<code class="sig-name descname">models_grid</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/hyperparameters.html#HyperparameterDefaults.models_grid"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.hyperparameters.HyperparameterDefaults.models_grid" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a grid of models by taking the cartesian product of all specified
model parameter lists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>The valid kwarg parameters are the entries of this</strong></dt><dd></dd>
<dt><strong>HyperparameterDefaults instance. Each parameter must be a list</strong></dt><dd></dd>
<dt><strong>giving the values to search across.</strong></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>list of dict giving the parameters for each model. The length of the</dt><dd></dd>
<dt>list is the product of the lengths of the input lists.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.local_parallelism">
<span id="mhcflurry-local-parallelism-module"></span><h2>mhcflurry.local_parallelism module<a class="headerlink" href="#module-mhcflurry.local_parallelism" title="Permalink to this headline">¶</a></h2>
<p>Infrastructure for “local” parallelism, i.e. multiprocess parallelism on one
compute node.</p>
<dl class="function">
<dt id="mhcflurry.local_parallelism.add_local_parallelism_args">
<code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">add_local_parallelism_args</code><span class="sig-paren">(</span><em class="sig-param">parser</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#add_local_parallelism_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.add_local_parallelism_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Add local parallelism arguments to the given argparse.ArgumentParser.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>parser</strong><span class="classifier">argparse.ArgumentParser</span></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.local_parallelism.worker_pool_with_gpu_assignments_from_args">
<code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">worker_pool_with_gpu_assignments_from_args</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#worker_pool_with_gpu_assignments_from_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.worker_pool_with_gpu_assignments_from_args" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a multiprocessing.Pool where each worker uses its own GPU.</p>
<p>Uses commandline arguments. See <a class="reference internal" href="#mhcflurry.local_parallelism.worker_pool_with_gpu_assignments" title="mhcflurry.local_parallelism.worker_pool_with_gpu_assignments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">worker_pool_with_gpu_assignments</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>args</strong><span class="classifier">argparse.ArgumentParser</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>multiprocessing.Pool</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.local_parallelism.worker_pool_with_gpu_assignments">
<code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">worker_pool_with_gpu_assignments</code><span class="sig-paren">(</span><em class="sig-param">num_jobs</em>, <em class="sig-param">num_gpus=0</em>, <em class="sig-param">backend=None</em>, <em class="sig-param">max_workers_per_gpu=1</em>, <em class="sig-param">max_tasks_per_worker=None</em>, <em class="sig-param">worker_log_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#worker_pool_with_gpu_assignments"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.worker_pool_with_gpu_assignments" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a multiprocessing.Pool where each worker uses its own GPU.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>num_jobs</strong><span class="classifier">int</span></dt><dd><p>Number of worker processes.</p>
</dd>
<dt><strong>num_gpus</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>backend</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>max_workers_per_gpu</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>max_tasks_per_worker</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>worker_log_dir</strong><span class="classifier">string</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>multiprocessing.Pool</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.local_parallelism.make_worker_pool">
<code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">make_worker_pool</code><span class="sig-paren">(</span><em class="sig-param">processes=None</em>, <em class="sig-param">initializer=None</em>, <em class="sig-param">initializer_kwargs_per_process=None</em>, <em class="sig-param">max_tasks_per_worker=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#make_worker_pool"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.make_worker_pool" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience wrapper to create a multiprocessing.Pool.</p>
<p>This function adds support for per-worker initializer arguments, which are
not natively supported by the multiprocessing module. The motivation for
this feature is to support allocating each worker to a (different) GPU.</p>
<dl>
<dt>IMPLEMENTATION NOTE:</dt><dd><p>The per-worker initializer arguments are implemented using a Queue. Each
worker reads its arguments from this queue when it starts. When it
terminates, it adds its initializer arguments back to the queue, so a
future process can initialize itself using these arguments.</p>
<p>There is one issue with this approach, however. If a worker crashes, it
never repopulates the queue of initializer arguments. This will prevent
any future worker from re-using those arguments. To deal with this
issue we add a second ‘backup queue’. This queue always contains the
full set of initializer arguments: whenever a worker reads from it, it
always pushes the pop’d args back to the end of the queue immediately.
If the primary arg queue is ever empty, then workers will read
from this backup queue.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>processes</strong><span class="classifier">int</span></dt><dd><p>Number of workers. Default: num CPUs.</p>
</dd>
<dt><strong>initializer</strong><span class="classifier">function, optional</span></dt><dd><p>Init function to call in each worker</p>
</dd>
<dt><strong>initializer_kwargs_per_process</strong><span class="classifier">list of dict, optional</span></dt><dd><p>Arguments to pass to initializer function for each worker. Length of
list must equal the number of workers.</p>
</dd>
<dt><strong>max_tasks_per_worker</strong><span class="classifier">int, optional</span></dt><dd><p>Restart workers after this many tasks. Requires Python &gt;=3.2.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>multiprocessing.Pool</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.local_parallelism.worker_init_entry_point">
<code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">worker_init_entry_point</code><span class="sig-paren">(</span><em class="sig-param">init_function</em>, <em class="sig-param">arg_queue=None</em>, <em class="sig-param">backup_arg_queue=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#worker_init_entry_point"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.worker_init_entry_point" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.local_parallelism.worker_init">
<code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">worker_init</code><span class="sig-paren">(</span><em class="sig-param">keras_backend=None</em>, <em class="sig-param">gpu_device_nums=None</em>, <em class="sig-param">worker_log_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#worker_init"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.worker_init" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="exception">
<dt id="mhcflurry.local_parallelism.WrapException">
<em class="property">exception </em><code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">WrapException</code><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#WrapException"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.WrapException" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Add traceback info to exception so exceptions raised in worker processes
can still show traceback info when re-raised in the parent.</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.local_parallelism.call_wrapped">
<code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">call_wrapped</code><span class="sig-paren">(</span><em class="sig-param">function</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#call_wrapped"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.call_wrapped" title="Permalink to this definition">¶</a></dt>
<dd><p>Run function on args and kwargs and return result, wrapping any exception
raised in a WrapException.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>function</strong><span class="classifier">arbitrary function</span></dt><dd></dd>
<dt><strong>Any other arguments provided are passed to the function.</strong></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>object</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.local_parallelism.call_wrapped_kwargs">
<code class="sig-prename descclassname">mhcflurry.local_parallelism.</code><code class="sig-name descname">call_wrapped_kwargs</code><span class="sig-paren">(</span><em class="sig-param">function</em>, <em class="sig-param">kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/local_parallelism.html#call_wrapped_kwargs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.local_parallelism.call_wrapped_kwargs" title="Permalink to this definition">¶</a></dt>
<dd><p>Invoke function on given kwargs and return result, wrapping any exception
raised in a WrapException.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>function</strong><span class="classifier">arbitrary function</span></dt><dd></dd>
<dt><strong>kwargs</strong><span class="classifier">dict</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>object</dt><dd></dd>
<dt>result of calling function(<a href="#id4"><span class="problematic" id="id5">**</span></a>kwargs)</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.percent_rank_transform">
<span id="mhcflurry-percent-rank-transform-module"></span><h2>mhcflurry.percent_rank_transform module<a class="headerlink" href="#module-mhcflurry.percent_rank_transform" title="Permalink to this headline">¶</a></h2>
<p>Class for transforming arbitrary values into percent ranks given a distribution.</p>
<dl class="class">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.percent_rank_transform.</code><code class="sig-name descname">PercentRankTransform</code><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Transform arbitrary values into percent ranks.</p>
<dl class="method">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">values</em>, <em class="sig-param">bins</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform.fit"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the transform using the given values (in our case ic50s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>values</strong><span class="classifier">ic50 values</span></dt><dd></dd>
<dt><strong>bins</strong><span class="classifier">bins for the cumulative distribution function</span></dt><dd><p>Anything that can be passed to numpy.histogram’s “bins” argument
can be used here.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform.transform">
<code class="sig-name descname">transform</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">values</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform.transform"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform.transform" title="Permalink to this definition">¶</a></dt>
<dd><p>Return percent ranks (range [0, 100]) for the given values.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform.to_series">
<code class="sig-name descname">to_series</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform.to_series"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform.to_series" title="Permalink to this definition">¶</a></dt>
<dd><p>Serialize the fit to a pandas.Series.</p>
<p>The index on the series gives the bin edges and the valeus give the CDF.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>pandas.Series</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.percent_rank_transform.PercentRankTransform.from_series">
<em class="property">static </em><code class="sig-name descname">from_series</code><span class="sig-paren">(</span><em class="sig-param">series</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/percent_rank_transform.html#PercentRankTransform.from_series"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.percent_rank_transform.PercentRankTransform.from_series" title="Permalink to this definition">¶</a></dt>
<dd><p>Deseralize a PercentRankTransform the given pandas.Series, as returned
by <a class="reference internal" href="#mhcflurry.percent_rank_transform.PercentRankTransform.to_series" title="mhcflurry.percent_rank_transform.PercentRankTransform.to_series"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_series()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>series</strong><span class="classifier">pandas.Series</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>PercentRankTransform</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.predict_command">
<span id="mhcflurry-predict-command-module"></span><h2>mhcflurry.predict_command module<a class="headerlink" href="#module-mhcflurry.predict_command" title="Permalink to this headline">¶</a></h2>
<p>Run MHCflurry predictor on specified peptides.</p>
<p>By default, the presentation predictor is used, and predictions for
MHC I binding affinity, antigen processing, and the composite presentation score
are returned. If you just want binding affinity predictions, pass
–affinity-only.</p>
<p>Examples:</p>
<p>Write a CSV file containing the contents of INPUT.csv plus additional columns
giving MHCflurry predictions:</p>
<p>$ mhcflurry-predict INPUT.csv –out RESULT.csv</p>
<p>The input CSV file is expected to contain columns “allele”, “peptide”, and,
optionally, “n_flank”, and “c_flank”.</p>
<p>If <code class="xref py py-obj docutils literal notranslate"><span class="pre">--out</span></code> is not specified, results are written to stdout.</p>
<p>You can also run on alleles and peptides specified on the commandline, in
which case predictions are written for <em>all combinations</em> of alleles and
peptides:</p>
<p>$ mhcflurry-predict –alleles HLA-A0201 H-2Kb –peptides SIINFEKL DENDREKLLL</p>
<p>Instead of individual alleles (in a CSV or on the command line), you can also
give a comma separated list of alleles giving a sample genotype. In this case,
the tightest binding affinity across the alleles for the sample will be
returned. For example:</p>
<p>$ mhcflurry-predict –peptides SIINFEKL DENDREKLLL     –alleles         HLA-A*02:01,HLA-A*03:01,HLA-B*57:01,HLA-B*45:01,HLA-C*02:01,HLA-C*07:02         HLA-A*01:01,HLA-A*02:06,HLA-B*44:02,HLA-B*07:02,HLA-C*01:01,HLA-C*03:01</p>
<p>will give the tightest predicted affinities across alleles for each of the two
genotypes specified for each peptide.</p>
<dl class="function">
<dt id="mhcflurry.predict_command.run">
<code class="sig-prename descclassname">mhcflurry.predict_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/predict_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.predict_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.predict_scan_command">
<span id="mhcflurry-predict-scan-command-module"></span><h2>mhcflurry.predict_scan_command module<a class="headerlink" href="#module-mhcflurry.predict_scan_command" title="Permalink to this headline">¶</a></h2>
<p>Scan protein sequences using the MHCflurry presentation predictor.</p>
<p>By default, sub-sequences (peptides) with affinity percentile ranks less than
2.0 are returned. You can also specify –results-all to return predictions for
all peptides, or –results-best to return the top peptide for each sequence.</p>
<p>Examples:</p>
<p>Scan a set of sequences in a FASTA file for binders to any alleles in a MHC I
genotype:</p>
<p>$ mhcflurry-predict-scan     test/data/example.fasta     –alleles HLA-A*02:01,HLA-A*03:01,HLA-B*57:01,HLA-B*45:01,HLA-C*02:01,HLA-C*07:02</p>
<p>Instead of a FASTA, you can also pass a CSV that has “sequence_id” and “sequence”
columns.</p>
<p>You can also specify multiple MHC I genotypes to scan as space-separated
arguments to the –alleles option:</p>
<p>$ mhcflurry-predict-scan     test/data/example.fasta     –alleles         HLA-A*02:01,HLA-A*03:01,HLA-B*57:01,HLA-B*45:01,HLA-C*02:02,HLA-C*07:02         HLA-A*01:01,HLA-A*02:06,HLA-B*44:02,HLA-B*07:02,HLA-C*01:02,HLA-C*03:01</p>
<p>If <code class="xref py py-obj docutils literal notranslate"><span class="pre">--out</span></code> is not specified, results are written to standard out.</p>
<p>You can also specify sequences on the commandline:</p>
<p>mhcflurry-predict-scan     –sequences MGYINVFAFPFTIYSLLLCRMNSRNYIAQVDVVNFNLT     –alleles HLA-A*02:01,HLA-A*03:01,HLA-B*57:01,HLA-B*45:01,HLA-C*02:02,HLA-C*07:02</p>
<dl class="function">
<dt id="mhcflurry.predict_scan_command.run">
<code class="sig-prename descclassname">mhcflurry.predict_scan_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/predict_scan_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.predict_scan_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.random_negative_peptides">
<span id="mhcflurry-random-negative-peptides-module"></span><h2>mhcflurry.random_negative_peptides module<a class="headerlink" href="#module-mhcflurry.random_negative_peptides" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.random_negative_peptides.</code><code class="sig-name descname">RandomNegativePeptides</code><span class="sig-paren">(</span><em class="sig-param">**hyperparameters</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/random_negative_peptides.html#RandomNegativePeptides"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Generate random negative (peptide, allele) pairs. These are used during
model training, where they are resampled at each epoch.</p>
<dl class="attribute">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides.hyperparameter_defaults">
<code class="sig-name descname">hyperparameter_defaults</code><em class="property"> = &lt;mhcflurry.hyperparameters.HyperparameterDefaults object&gt;</em><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.hyperparameter_defaults" title="Permalink to this definition">¶</a></dt>
<dd><p>Hyperperameters for random negative peptides.</p>
<dl class="simple">
<dt>Number of random negatives will be:</dt><dd><p>random_negative_rate * (num measurements) + random_negative_constant</p>
</dd>
</dl>
<p>where the exact meaning of (num measurements) depends on the particular
random_negative_method in use.</p>
<p>If random_negative_match_distribution is True, then the amino acid
frequencies of the training data peptides are used to generate the
random peptides.</p>
<dl class="simple">
<dt>Valid values for random_negative_method are:</dt><dd><dl class="simple">
<dt>“by_length”: used for allele-specific prediction. See description in</dt><dd><p><a class="reference internal" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_length" title="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RandomNegativePeptides.plan_by_length</span></code></a> method.</p>
</dd>
<dt>“by_allele”: used for pan-allele prediction. See</dt><dd><p><a class="reference internal" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele" title="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RandomNegativePeptides.plan_by_allele</span></code></a> method.</p>
</dd>
<dt>“by_allele_equalize_nonbinders”: used for pan-allele prediction. See</dt><dd><p><a class="reference internal" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele_equalize_nonbinders" title="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele_equalize_nonbinders"><code class="xref py py-obj docutils literal notranslate"><span class="pre">RandomNegativePeptides.plan_by_allele_equalize_nonbinders</span></code></a> method.</p>
</dd>
<dt>“recommended”: the default. Use by_length if the predictor is allele-</dt><dd><p>specific and by_allele if it’s pan-allele.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan">
<code class="sig-name descname">plan</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">affinities</em>, <em class="sig-param">alleles=None</em>, <em class="sig-param">inequalities=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/random_negative_peptides.html#RandomNegativePeptides.plan"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the number of random negatives for each allele and peptide
length. Call this once after instantiating the object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>peptides</strong><span class="classifier">list of string</span></dt><dd></dd>
<dt><strong>affinities</strong><span class="classifier">list of float</span></dt><dd></dd>
<dt><strong>alleles</strong><span class="classifier">list of string, optional</span></dt><dd></dd>
<dt><strong>inequalities</strong><span class="classifier">list of string (“&gt;”, “&lt;”, or “=”), optional</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame indicating number of random negatives for each length</dt><dd></dd>
<dt>and allele.</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_length">
<code class="sig-name descname">plan_by_length</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">df_all</em>, <em class="sig-param">df_binders=None</em>, <em class="sig-param">df_nonbinders=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/random_negative_peptides.html#RandomNegativePeptides.plan_by_length"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_length" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a random negative plan using the “by_length” policy.</p>
<p>Parameters are as in the <a class="reference internal" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan" title="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plan</span></code></a> method. No return value.</p>
<p>Used for allele-specific predictors. Does not work well for pan-allele.</p>
<p>Different numbers of random negatives per length. Alleles are sampled
proportionally to the number of times they are used in the training
data.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele">
<code class="sig-name descname">plan_by_allele</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">df_all</em>, <em class="sig-param">df_binders=None</em>, <em class="sig-param">df_nonbinders=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/random_negative_peptides.html#RandomNegativePeptides.plan_by_allele"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a random negative plan using the “by_allele” policy.</p>
<p>Parameters are as in the <a class="reference internal" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan" title="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plan</span></code></a> method. No return value.</p>
<p>For each allele, a particular number of random negatives are used
for all lengths. Across alleles, the number of random negatives
varies; within an allele, the number of random negatives for each
length is a constant</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele_equalize_nonbinders">
<code class="sig-name descname">plan_by_allele_equalize_nonbinders</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">df_all</em>, <em class="sig-param">df_binders</em>, <em class="sig-param">df_nonbinders</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/random_negative_peptides.html#RandomNegativePeptides.plan_by_allele_equalize_nonbinders"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele_equalize_nonbinders" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a random negative plan using the
“by_allele_equalize_nonbinders” policy.</p>
<p>Parameters are as in the <a class="reference internal" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan" title="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plan</span></code></a> method. No return value.</p>
<p>Requires that the random_negative_binder_threshold hyperparameter is set.</p>
<p>In a first step, the number of random negatives selected by the
“by_allele” method are added (see <a class="reference internal" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele" title="mhcflurry.random_negative_peptides.RandomNegativePeptides.plan_by_allele"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plan_by_allele</span></code></a>). Then, the total
number of non-binders are calculated for each allele and length. This
total includes non-binder measurements in the training data plus the
random negative peptides added in the first step. In a second step,
additional random negative peptides are added so that for each allele,
all peptide lengths have the same total number of non-binders.</p>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides.get_alleles">
<code class="sig-name descname">get_alleles</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/random_negative_peptides.html#RandomNegativePeptides.get_alleles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.get_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the list of alleles corresponding to each random negative peptide
as returned by <a class="reference internal" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.get_peptides" title="mhcflurry.random_negative_peptides.RandomNegativePeptides.get_peptides"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_peptides</span></code></a>. This does NOT change and can be safely
called once and reused.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides.get_peptides">
<code class="sig-name descname">get_peptides</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/random_negative_peptides.html#RandomNegativePeptides.get_peptides"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.get_peptides" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the list of random negative peptides. This will be different each
time the method is called.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>list of string</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="mhcflurry.random_negative_peptides.RandomNegativePeptides.get_total_count">
<code class="sig-name descname">get_total_count</code><span class="sig-paren">(</span><em class="sig-param">self</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/random_negative_peptides.html#RandomNegativePeptides.get_total_count"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.random_negative_peptides.RandomNegativePeptides.get_total_count" title="Permalink to this definition">¶</a></dt>
<dd><p>Total number of planned random negative peptides.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><dl class="simple">
<dt>int</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.regression_target">
<span id="mhcflurry-regression-target-module"></span><h2>mhcflurry.regression_target module<a class="headerlink" href="#module-mhcflurry.regression_target" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="mhcflurry.regression_target.from_ic50">
<code class="sig-prename descclassname">mhcflurry.regression_target.</code><code class="sig-name descname">from_ic50</code><span class="sig-paren">(</span><em class="sig-param">ic50</em>, <em class="sig-param">max_ic50=50000.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/regression_target.html#from_ic50"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.regression_target.from_ic50" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert ic50s to regression targets in the range [0.0, 1.0].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ic50</strong><span class="classifier">numpy.array of float</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of float</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.regression_target.to_ic50">
<code class="sig-prename descclassname">mhcflurry.regression_target.</code><code class="sig-name descname">to_ic50</code><span class="sig-paren">(</span><em class="sig-param">x</em>, <em class="sig-param">max_ic50=50000.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/regression_target.html#to_ic50"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.regression_target.to_ic50" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert regression targets in the range [0.0, 1.0] to ic50s in the range
[0, 50000.0].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>x</strong><span class="classifier">numpy.array of float</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>numpy.array of float</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.scoring">
<span id="mhcflurry-scoring-module"></span><h2>mhcflurry.scoring module<a class="headerlink" href="#module-mhcflurry.scoring" title="Permalink to this headline">¶</a></h2>
<p>Measures of prediction accuracy</p>
<dl class="function">
<dt id="mhcflurry.scoring.make_scores">
<code class="sig-prename descclassname">mhcflurry.scoring.</code><code class="sig-name descname">make_scores</code><span class="sig-paren">(</span><em class="sig-param">ic50_y</em>, <em class="sig-param">ic50_y_pred</em>, <em class="sig-param">sample_weight=None</em>, <em class="sig-param">threshold_nm=500</em>, <em class="sig-param">max_ic50=50000</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/scoring.html#make_scores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.scoring.make_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate AUC, F1, and Kendall Tau scores.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ic50_y</strong><span class="classifier">float list</span></dt><dd><p>true IC50s (i.e. affinities)</p>
</dd>
<dt><strong>ic50_y_pred</strong><span class="classifier">float list</span></dt><dd><p>predicted IC50s</p>
</dd>
<dt><strong>sample_weight</strong><span class="classifier">float list [optional]</span></dt><dd></dd>
<dt><strong>threshold_nm</strong><span class="classifier">float [optional]</span></dt><dd></dd>
<dt><strong>max_ic50</strong><span class="classifier">float [optional]</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>dict with entries “auc”, “f1”, “tau”</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.select_allele_specific_models_command">
<span id="mhcflurry-select-allele-specific-models-command-module"></span><h2>mhcflurry.select_allele_specific_models_command module<a class="headerlink" href="#module-mhcflurry.select_allele_specific_models_command" title="Permalink to this headline">¶</a></h2>
<p>Model select class1 single allele models.</p>
<dl class="function">
<dt id="mhcflurry.select_allele_specific_models_command.run">
<code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="mhcflurry.select_allele_specific_models_command.ScrambledPredictor">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">ScrambledPredictor</code><span class="sig-paren">(</span><em class="sig-param">predictor</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#ScrambledPredictor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.ScrambledPredictor" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.ScrambledPredictor.predict">
<code class="sig-name descname">predict</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">peptides</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#ScrambledPredictor.predict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.ScrambledPredictor.predict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="mhcflurry.select_allele_specific_models_command.model_select">
<code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">model_select</code><span class="sig-paren">(</span><em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#model_select"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.model_select" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.select_allele_specific_models_command.cache_encoding">
<code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">cache_encoding</code><span class="sig-paren">(</span><em class="sig-param">predictor</em>, <em class="sig-param">peptides</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#cache_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.cache_encoding" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="class">
<dt id="mhcflurry.select_allele_specific_models_command.ScoreFunction">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">ScoreFunction</code><span class="sig-paren">(</span><em class="sig-param">function</em>, <em class="sig-param">summary=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#ScoreFunction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.ScoreFunction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Thin wrapper over a score function (Class1AffinityPredictor -&gt; float).
Used to keep a summary string associated with the function.</p>
</dd></dl>

<dl class="class">
<dt id="mhcflurry.select_allele_specific_models_command.CombinedModelSelector">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">CombinedModelSelector</code><span class="sig-paren">(</span><em class="sig-param">model_selectors</em>, <em class="sig-param">weights=None</em>, <em class="sig-param">min_contribution_percent=1.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#CombinedModelSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.CombinedModelSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Model selector that computes a weighted average over other model selectors.</p>
<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.CombinedModelSelector.usable_for_allele">
<code class="sig-name descname">usable_for_allele</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#CombinedModelSelector.usable_for_allele"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.CombinedModelSelector.usable_for_allele" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.CombinedModelSelector.plan_summary">
<code class="sig-name descname">plan_summary</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#CombinedModelSelector.plan_summary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.CombinedModelSelector.plan_summary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.CombinedModelSelector.score_function">
<code class="sig-name descname">score_function</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em>, <em class="sig-param">dry_run=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#CombinedModelSelector.score_function"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.CombinedModelSelector.score_function" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.select_allele_specific_models_command.ConsensusModelSelector">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">ConsensusModelSelector</code><span class="sig-paren">(</span><em class="sig-param">predictor</em>, <em class="sig-param">num_peptides_per_length=10000</em>, <em class="sig-param">multiply_score_by_value=10.0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#ConsensusModelSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.ConsensusModelSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Model selector that scores sub-ensembles based on their Kendall tau
consistency with the full ensemble over a set of random peptides.</p>
<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.ConsensusModelSelector.usable_for_allele">
<code class="sig-name descname">usable_for_allele</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#ConsensusModelSelector.usable_for_allele"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.ConsensusModelSelector.usable_for_allele" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.ConsensusModelSelector.max_absolute_value">
<code class="sig-name descname">max_absolute_value</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#ConsensusModelSelector.max_absolute_value"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.ConsensusModelSelector.max_absolute_value" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.ConsensusModelSelector.plan_summary">
<code class="sig-name descname">plan_summary</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#ConsensusModelSelector.plan_summary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.ConsensusModelSelector.plan_summary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.ConsensusModelSelector.score_function">
<code class="sig-name descname">score_function</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#ConsensusModelSelector.score_function"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.ConsensusModelSelector.score_function" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.select_allele_specific_models_command.MSEModelSelector">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">MSEModelSelector</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">predictor</em>, <em class="sig-param">min_measurements=1</em>, <em class="sig-param">multiply_score_by_data_size=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MSEModelSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MSEModelSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Model selector that uses mean-squared error to score models. Inequalities
are supported.</p>
<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MSEModelSelector.usable_for_allele">
<code class="sig-name descname">usable_for_allele</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MSEModelSelector.usable_for_allele"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MSEModelSelector.usable_for_allele" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MSEModelSelector.max_absolute_value">
<code class="sig-name descname">max_absolute_value</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MSEModelSelector.max_absolute_value"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MSEModelSelector.max_absolute_value" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MSEModelSelector.plan_summary">
<code class="sig-name descname">plan_summary</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MSEModelSelector.plan_summary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MSEModelSelector.plan_summary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MSEModelSelector.score_function">
<code class="sig-name descname">score_function</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MSEModelSelector.score_function"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MSEModelSelector.score_function" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="mhcflurry.select_allele_specific_models_command.MassSpecModelSelector">
<em class="property">class </em><code class="sig-prename descclassname">mhcflurry.select_allele_specific_models_command.</code><code class="sig-name descname">MassSpecModelSelector</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">predictor</em>, <em class="sig-param">decoys_per_length=0</em>, <em class="sig-param">min_measurements=100</em>, <em class="sig-param">multiply_score_by_data_size=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MassSpecModelSelector"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MassSpecModelSelector" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Model selector that uses PPV of differentiating decoys from hits from
mass-spec experiments.</p>
<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.ppv">
<em class="property">static </em><code class="sig-name descname">ppv</code><span class="sig-paren">(</span><em class="sig-param">y_true</em>, <em class="sig-param">predictions</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MassSpecModelSelector.ppv"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.ppv" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.usable_for_allele">
<code class="sig-name descname">usable_for_allele</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MassSpecModelSelector.usable_for_allele"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.usable_for_allele" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.max_absolute_value">
<code class="sig-name descname">max_absolute_value</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MassSpecModelSelector.max_absolute_value"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.max_absolute_value" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.plan_summary">
<code class="sig-name descname">plan_summary</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MassSpecModelSelector.plan_summary"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.plan_summary" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.score_function">
<code class="sig-name descname">score_function</code><span class="sig-paren">(</span><em class="sig-param">self</em>, <em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_allele_specific_models_command.html#MassSpecModelSelector.score_function"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_allele_specific_models_command.MassSpecModelSelector.score_function" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</div>
<div class="section" id="module-mhcflurry.select_pan_allele_models_command">
<span id="mhcflurry-select-pan-allele-models-command-module"></span><h2>mhcflurry.select_pan_allele_models_command module<a class="headerlink" href="#module-mhcflurry.select_pan_allele_models_command" title="Permalink to this headline">¶</a></h2>
<p>Model select class1 pan-allele models.</p>
<p>APPROACH: For each training fold, we select at least min and at most max models
(where min and max are set by the –{min/max}-models-per-fold argument) using a
step-up (forward) selection procedure. The final ensemble is the union of all
selected models across all folds.</p>
<dl class="function">
<dt id="mhcflurry.select_pan_allele_models_command.mse">
<code class="sig-prename descclassname">mhcflurry.select_pan_allele_models_command.</code><code class="sig-name descname">mse</code><span class="sig-paren">(</span><em class="sig-param">predictions</em>, <em class="sig-param">actual</em>, <em class="sig-param">inequalities=None</em>, <em class="sig-param">affinities_are_already_01_transformed=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_pan_allele_models_command.html#mse"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_pan_allele_models_command.mse" title="Permalink to this definition">¶</a></dt>
<dd><p>Mean squared error of predictions vs. actual</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>predictions</strong><span class="classifier">list of float</span></dt><dd></dd>
<dt><strong>actual</strong><span class="classifier">list of float</span></dt><dd></dd>
<dt><strong>inequalities</strong><span class="classifier">list of string (“&gt;”, “&lt;”, or “=”)</span></dt><dd></dd>
<dt><strong>affinities_are_already_01_transformed</strong><span class="classifier">boolean</span></dt><dd><p>Predictions and actual are taken to be nanomolar affinities if
affinities_are_already_01_transformed is False, otherwise 0-1 values.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>float</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.select_pan_allele_models_command.run">
<code class="sig-prename descclassname">mhcflurry.select_pan_allele_models_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_pan_allele_models_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_pan_allele_models_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.select_pan_allele_models_command.do_model_select_task">
<code class="sig-prename descclassname">mhcflurry.select_pan_allele_models_command.</code><code class="sig-name descname">do_model_select_task</code><span class="sig-paren">(</span><em class="sig-param">item</em>, <em class="sig-param">constant_data={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_pan_allele_models_command.html#do_model_select_task"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_pan_allele_models_command.do_model_select_task" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.select_pan_allele_models_command.model_select">
<code class="sig-prename descclassname">mhcflurry.select_pan_allele_models_command.</code><code class="sig-name descname">model_select</code><span class="sig-paren">(</span><em class="sig-param">fold_num</em>, <em class="sig-param">models</em>, <em class="sig-param">min_models</em>, <em class="sig-param">max_models</em>, <em class="sig-param">constant_data={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_pan_allele_models_command.html#model_select"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_pan_allele_models_command.model_select" title="Permalink to this definition">¶</a></dt>
<dd><p>Model select for a fold.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>fold_num</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>models</strong><span class="classifier">list of Class1NeuralNetwork</span></dt><dd></dd>
<dt><strong>min_models</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>max_models</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>constant_data</strong><span class="classifier">dict</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>dict with keys ‘fold_num’, ‘selected_indices’, ‘summary’</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.select_processing_models_command">
<span id="mhcflurry-select-processing-models-command-module"></span><h2>mhcflurry.select_processing_models_command module<a class="headerlink" href="#module-mhcflurry.select_processing_models_command" title="Permalink to this headline">¶</a></h2>
<p>Model select antigen processing models.</p>
<p>APPROACH: For each training fold, we select at least min and at most max models
(where min and max are set by the –{min/max}-models-per-fold argument) using a
step-up (forward) selection procedure. The final ensemble is the union of all
selected models across all folds. AUC is used as the metric.</p>
<dl class="function">
<dt id="mhcflurry.select_processing_models_command.run">
<code class="sig-prename descclassname">mhcflurry.select_processing_models_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_processing_models_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_processing_models_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.select_processing_models_command.do_model_select_task">
<code class="sig-prename descclassname">mhcflurry.select_processing_models_command.</code><code class="sig-name descname">do_model_select_task</code><span class="sig-paren">(</span><em class="sig-param">item</em>, <em class="sig-param">constant_data={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_processing_models_command.html#do_model_select_task"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_processing_models_command.do_model_select_task" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.select_processing_models_command.model_select">
<code class="sig-prename descclassname">mhcflurry.select_processing_models_command.</code><code class="sig-name descname">model_select</code><span class="sig-paren">(</span><em class="sig-param">fold_num</em>, <em class="sig-param">models</em>, <em class="sig-param">min_models</em>, <em class="sig-param">max_models</em>, <em class="sig-param">constant_data={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/select_processing_models_command.html#model_select"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.select_processing_models_command.model_select" title="Permalink to this definition">¶</a></dt>
<dd><p>Model select for a fold.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>fold_num</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>models</strong><span class="classifier">list of Class1NeuralNetwork</span></dt><dd></dd>
<dt><strong>min_models</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>max_models</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>constant_data</strong><span class="classifier">dict</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>dict with keys ‘fold_num’, ‘selected_indices’, ‘summary’</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.testing_utils">
<span id="mhcflurry-testing-utils-module"></span><h2>mhcflurry.testing_utils module<a class="headerlink" href="#module-mhcflurry.testing_utils" title="Permalink to this headline">¶</a></h2>
<p>Utilities used in MHCflurry unit tests.</p>
<dl class="function">
<dt id="mhcflurry.testing_utils.startup">
<code class="sig-prename descclassname">mhcflurry.testing_utils.</code><code class="sig-name descname">startup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/testing_utils.html#startup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.testing_utils.startup" title="Permalink to this definition">¶</a></dt>
<dd><p>Configure Keras backend for running unit tests.</p>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.testing_utils.cleanup">
<code class="sig-prename descclassname">mhcflurry.testing_utils.</code><code class="sig-name descname">cleanup</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/testing_utils.html#cleanup"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.testing_utils.cleanup" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear tensorflow session and other process-wide resources.</p>
</dd></dl>

</div>
<div class="section" id="module-mhcflurry.train_allele_specific_models_command">
<span id="mhcflurry-train-allele-specific-models-command-module"></span><h2>mhcflurry.train_allele_specific_models_command module<a class="headerlink" href="#module-mhcflurry.train_allele_specific_models_command" title="Permalink to this headline">¶</a></h2>
<p>Train Class1 single allele models.</p>
<dl class="function">
<dt id="mhcflurry.train_allele_specific_models_command.run">
<code class="sig-prename descclassname">mhcflurry.train_allele_specific_models_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_allele_specific_models_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_allele_specific_models_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_allele_specific_models_command.alleles_by_similarity">
<code class="sig-prename descclassname">mhcflurry.train_allele_specific_models_command.</code><code class="sig-name descname">alleles_by_similarity</code><span class="sig-paren">(</span><em class="sig-param">allele</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_allele_specific_models_command.html#alleles_by_similarity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_allele_specific_models_command.alleles_by_similarity" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_allele_specific_models_command.train_model">
<code class="sig-prename descclassname">mhcflurry.train_allele_specific_models_command.</code><code class="sig-name descname">train_model</code><span class="sig-paren">(</span><em class="sig-param">n_models</em>, <em class="sig-param">allele_num</em>, <em class="sig-param">n_alleles</em>, <em class="sig-param">hyperparameter_set_num</em>, <em class="sig-param">num_hyperparameter_sets</em>, <em class="sig-param">allele</em>, <em class="sig-param">hyperparameters</em>, <em class="sig-param">verbose</em>, <em class="sig-param">progress_print_interval</em>, <em class="sig-param">predictor</em>, <em class="sig-param">save_to</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_allele_specific_models_command.html#train_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_allele_specific_models_command.train_model" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_allele_specific_models_command.subselect_df_held_out">
<code class="sig-prename descclassname">mhcflurry.train_allele_specific_models_command.</code><code class="sig-name descname">subselect_df_held_out</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">recriprocal_held_out_fraction=10</em>, <em class="sig-param">seed=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_allele_specific_models_command.html#subselect_df_held_out"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_allele_specific_models_command.subselect_df_held_out" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.train_pan_allele_models_command">
<span id="mhcflurry-train-pan-allele-models-command-module"></span><h2>mhcflurry.train_pan_allele_models_command module<a class="headerlink" href="#module-mhcflurry.train_pan_allele_models_command" title="Permalink to this headline">¶</a></h2>
<p>Train Class1 pan-allele models.</p>
<dl class="function">
<dt id="mhcflurry.train_pan_allele_models_command.assign_folds">
<code class="sig-prename descclassname">mhcflurry.train_pan_allele_models_command.</code><code class="sig-name descname">assign_folds</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">num_folds</em>, <em class="sig-param">held_out_fraction</em>, <em class="sig-param">held_out_max</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_pan_allele_models_command.html#assign_folds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_pan_allele_models_command.assign_folds" title="Permalink to this definition">¶</a></dt>
<dd><p>Split training data into multple test/train pairs, which we refer to as
folds. Note that a given data point may be assigned to multiple test or
train sets; these folds are NOT a non-overlapping partition as used in cross
validation.</p>
<p>A fold is defined by a boolean value for each data point, indicating whether
it is included in the training data for that fold. If it’s not in the
training data, then it’s in the test data.</p>
<p>Folds are balanced in terms of allele content.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>df</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>training data</p>
</dd>
<dt><strong>num_folds</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>held_out_fraction</strong><span class="classifier">float</span></dt><dd><p>Fraction of data to hold out as test data in each fold</p>
</dd>
<dt><strong>held_out_max</strong></dt><dd><p>For a given allele, do not hold out more than held_out_max number of
data points in any fold.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd><p>index is same as df.index, columns are “fold_0”, … “fold_N” giving
whether the data point is in the training data for the fold</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.train_pan_allele_models_command.pretrain_data_iterator">
<code class="sig-prename descclassname">mhcflurry.train_pan_allele_models_command.</code><code class="sig-name descname">pretrain_data_iterator</code><span class="sig-paren">(</span><em class="sig-param">filename</em>, <em class="sig-param">master_allele_encoding</em>, <em class="sig-param">peptides_per_chunk=1024</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_pan_allele_models_command.html#pretrain_data_iterator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_pan_allele_models_command.pretrain_data_iterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Step through a CSV file giving predictions for a large number of peptides
(rows) and alleles (columns).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>filename</strong><span class="classifier">string</span></dt><dd></dd>
<dt><strong>master_allele_encoding</strong><span class="classifier">AlleleEncoding</span></dt><dd></dd>
<dt><strong>peptides_per_chunk</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>Generator of (AlleleEncoding, EncodableSequences, float affinities) tuples</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.train_pan_allele_models_command.run">
<code class="sig-prename descclassname">mhcflurry.train_pan_allele_models_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_pan_allele_models_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_pan_allele_models_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_pan_allele_models_command.main">
<code class="sig-prename descclassname">mhcflurry.train_pan_allele_models_command.</code><code class="sig-name descname">main</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_pan_allele_models_command.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_pan_allele_models_command.main" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_pan_allele_models_command.initialize_training">
<code class="sig-prename descclassname">mhcflurry.train_pan_allele_models_command.</code><code class="sig-name descname">initialize_training</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_pan_allele_models_command.html#initialize_training"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_pan_allele_models_command.initialize_training" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_pan_allele_models_command.train_models">
<code class="sig-prename descclassname">mhcflurry.train_pan_allele_models_command.</code><code class="sig-name descname">train_models</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_pan_allele_models_command.html#train_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_pan_allele_models_command.train_models" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_pan_allele_models_command.train_model">
<code class="sig-prename descclassname">mhcflurry.train_pan_allele_models_command.</code><code class="sig-name descname">train_model</code><span class="sig-paren">(</span><em class="sig-param">work_item_name</em>, <em class="sig-param">work_item_num</em>, <em class="sig-param">num_work_items</em>, <em class="sig-param">architecture_num</em>, <em class="sig-param">num_architectures</em>, <em class="sig-param">fold_num</em>, <em class="sig-param">num_folds</em>, <em class="sig-param">replicate_num</em>, <em class="sig-param">num_replicates</em>, <em class="sig-param">hyperparameters</em>, <em class="sig-param">pretrain_data_filename</em>, <em class="sig-param">verbose</em>, <em class="sig-param">progress_print_interval</em>, <em class="sig-param">predictor</em>, <em class="sig-param">save_to</em>, <em class="sig-param">constant_data={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_pan_allele_models_command.html#train_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_pan_allele_models_command.train_model" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.train_presentation_models_command">
<span id="mhcflurry-train-presentation-models-command-module"></span><h2>mhcflurry.train_presentation_models_command module<a class="headerlink" href="#module-mhcflurry.train_presentation_models_command" title="Permalink to this headline">¶</a></h2>
<p>Train Class1 presentation models.</p>
<dl class="function">
<dt id="mhcflurry.train_presentation_models_command.run">
<code class="sig-prename descclassname">mhcflurry.train_presentation_models_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_presentation_models_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_presentation_models_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_presentation_models_command.main">
<code class="sig-prename descclassname">mhcflurry.train_presentation_models_command.</code><code class="sig-name descname">main</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_presentation_models_command.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_presentation_models_command.main" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.train_processing_models_command">
<span id="mhcflurry-train-processing-models-command-module"></span><h2>mhcflurry.train_processing_models_command module<a class="headerlink" href="#module-mhcflurry.train_processing_models_command" title="Permalink to this headline">¶</a></h2>
<p>Train Class1 processing models.</p>
<dl class="function">
<dt id="mhcflurry.train_processing_models_command.assign_folds">
<code class="sig-prename descclassname">mhcflurry.train_processing_models_command.</code><code class="sig-name descname">assign_folds</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">num_folds</em>, <em class="sig-param">held_out_samples</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_processing_models_command.html#assign_folds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_processing_models_command.assign_folds" title="Permalink to this definition">¶</a></dt>
<dd><p>Split training data into mulitple test/train pairs, which we refer to as
folds. Note that a given data point may be assigned to multiple test or
train sets; these folds are NOT a non-overlapping partition as used in cross
validation.</p>
<p>A fold is defined by a boolean value for each data point, indicating whether
it is included in the training data for that fold. If it’s not in the
training data, then it’s in the test data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>df</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>training data</p>
</dd>
<dt><strong>num_folds</strong><span class="classifier">int</span></dt><dd></dd>
<dt><strong>held_out_samples</strong><span class="classifier">int</span></dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd><p>index is same as df.index, columns are “fold_0”, … “fold_N” giving
whether the data point is in the training data for the fold</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="mhcflurry.train_processing_models_command.run">
<code class="sig-prename descclassname">mhcflurry.train_processing_models_command.</code><code class="sig-name descname">run</code><span class="sig-paren">(</span><em class="sig-param">argv=['-b', 'html', '-v', '-d', '_build/doctrees', '.', '_build/html']</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_processing_models_command.html#run"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_processing_models_command.run" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_processing_models_command.main">
<code class="sig-prename descclassname">mhcflurry.train_processing_models_command.</code><code class="sig-name descname">main</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_processing_models_command.html#main"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_processing_models_command.main" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_processing_models_command.initialize_training">
<code class="sig-prename descclassname">mhcflurry.train_processing_models_command.</code><code class="sig-name descname">initialize_training</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_processing_models_command.html#initialize_training"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_processing_models_command.initialize_training" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_processing_models_command.train_models">
<code class="sig-prename descclassname">mhcflurry.train_processing_models_command.</code><code class="sig-name descname">train_models</code><span class="sig-paren">(</span><em class="sig-param">args</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_processing_models_command.html#train_models"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_processing_models_command.train_models" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="mhcflurry.train_processing_models_command.train_model">
<code class="sig-prename descclassname">mhcflurry.train_processing_models_command.</code><code class="sig-name descname">train_model</code><span class="sig-paren">(</span><em class="sig-param">work_item_name</em>, <em class="sig-param">work_item_num</em>, <em class="sig-param">num_work_items</em>, <em class="sig-param">architecture_num</em>, <em class="sig-param">num_architectures</em>, <em class="sig-param">fold_num</em>, <em class="sig-param">num_folds</em>, <em class="sig-param">replicate_num</em>, <em class="sig-param">num_replicates</em>, <em class="sig-param">hyperparameters</em>, <em class="sig-param">verbose</em>, <em class="sig-param">progress_print_interval</em>, <em class="sig-param">predictor</em>, <em class="sig-param">save_to</em>, <em class="sig-param">constant_data={}</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/mhcflurry/train_processing_models_command.html#train_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#mhcflurry.train_processing_models_command.train_model" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-mhcflurry.version">
<span id="mhcflurry-version-module"></span><h2>mhcflurry.version module<a class="headerlink" href="#module-mhcflurry.version" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="commandline_tools.html" class="btn btn-neutral float-left" title="Command-line reference" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright Timothy O&#39;Donnell
      <span class="lastupdated">
        Last updated on Mar 23, 2020.
      </span>

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>